<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>2 Download book and data | The Epidemiologist R Handbook</title>
<meta name="author" content="the handbook team">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.2"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/header-attrs-2.7/header-attrs.js"></script><script src="libs/jquery-3.5.1/jquery-3.5.1.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.5.3/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.5.3/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.2.4/tabs.js"></script><script src="libs/bs3compat-0.2.4/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="libs/htmlwidgets-1.5.3/htmlwidgets.js"></script><link href="libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="libs/datatables-binding-0.15/datatables.js"></script><link href="libs/dt-core-1.10.20/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="libs/dt-core-1.10.20/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="libs/dt-core-1.10.20/js/jquery.dataTables.min.js"></script><link href="libs/nouislider-7.0.10/jquery.nouislider.min.css" rel="stylesheet">
<script src="libs/nouislider-7.0.10/jquery.nouislider.min.js"></script><link href="libs/selectize-0.12.0/selectize.bootstrap3.css" rel="stylesheet">
<script src="libs/selectize-0.12.0/selectize.min.js"></script><link href="libs/crosstalk-1.1.1/css/crosstalk.css" rel="stylesheet">
<script src="libs/crosstalk-1.1.1/js/crosstalk.min.js"></script><script src="libs/kePrint-0.0.1/kePrint.js"></script><link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet">
<link href="libs/bsTable-3.3.7/bootstrapTable.min.css" rel="stylesheet">
<script src="libs/bsTable-3.3.7/bootstrapTable.js"></script><link href="libs/tabwid-1.0.0/tabwid.css" rel="stylesheet">
<link href="libs/leaflet-1.3.1/leaflet.css" rel="stylesheet">
<script src="libs/leaflet-1.3.1/leaflet.js"></script><link href="libs/leafletfix-1.0.0/leafletfix.css" rel="stylesheet">
<script src="libs/proj4-2.6.2/proj4.min.js"></script><script src="libs/Proj4Leaflet-1.0.1/proj4leaflet.js"></script><link href="libs/rstudio_leaflet-1.3.1/rstudio_leaflet.css" rel="stylesheet">
<script src="libs/leaflet-binding-2.0.4.1/leaflet.js"></script><script src="libs/leaflet-providers-1.9.0/leaflet-providers_1.9.0.js"></script><script src="libs/leaflet-providers-plugin-2.0.4.1/leaflet-providers-plugin.js"></script><script src="libs/viz-1.8.2/viz.js"></script><link href="libs/DiagrammeR-styles-0.2/styles.css" rel="stylesheet">
<script src="libs/grViz-binding-1.0.6.1/grViz.js"></script><script src="libs/d3-4.5.0/d3.min.js"></script><script src="libs/sankey-1/sankey.js"></script><script src="libs/sankeyNetwork-binding-0.4/sankeyNetwork.js"></script><script src="libs/plotly-binding-4.9.3/plotly.js"></script><script src="libs/typedarray-0.1/typedarray.min.js"></script><link href="libs/plotly-htmlwidgets-css-1.57.1/plotly-htmlwidgets.css" rel="stylesheet">
<script src="libs/plotly-main-1.57.1/plotly-latest.min.js"></script><link href="libs/vis-4.20.1/vis.css" rel="stylesheet">
<script src="libs/vis-4.20.1/vis.min.js"></script><script src="libs/visNetwork-binding-2.0.9/visNetwork.js"></script><link href="libs/font-awesome-4.7.0/css/font-awesome.min.css" rel="stylesheet">
<script src="libs/plotly-basic-1.57.1/plotly-basic-1.57.1.min.js"></script><script src="libs/plotly-cartesian-1.57.1/plotly-cartesian-1.57.1.min.js"></script><script src="https://cdn.jsdelivr.net/autocomplete.js/0/autocomplete.jquery.min.js"></script><script src="https://cdn.jsdelivr.net/npm/mark.js@8.11.1/dist/mark.min.js"></script><!-- CSS --><link rel="stylesheet" href="style_bs4.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">The Epidemiologist R Handbook</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"></a></li>
<li class="book-part">About this book</li>
<li><a class="" href="editorial-and-technical-notes.html"><span class="header-section-number">1</span> Editorial and technical notes</a></li>
<li><a class="active" href="download-book-and-data.html"><span class="header-section-number">2</span> Download book and data</a></li>
<li class="book-part">Basics</li>
<li><a class="" href="r-basics.html"><span class="header-section-number">3</span> R Basics</a></li>
<li><a class="" href="transition-to-r.html"><span class="header-section-number">4</span> Transition to R</a></li>
<li><a class="" href="import-and-export.html"><span class="header-section-number">5</span> Import and export</a></li>
<li><a class="" href="r-projects.html"><span class="header-section-number">6</span> R projects</a></li>
<li><a class="" href="suggested-packages-1.html"><span class="header-section-number">7</span> Suggested packages</a></li>
<li class="book-part">Data Management</li>
<li><a class="" href="cleaning-data-and-core-functions.html"><span class="header-section-number">8</span> Cleaning data and core functions</a></li>
<li><a class="" href="working-with-dates.html"><span class="header-section-number">9</span> Working with dates</a></li>
<li><a class="" href="factors.html"><span class="header-section-number">10</span> Factors</a></li>
<li><a class="" href="pivoting-data.html"><span class="header-section-number">11</span> Pivoting data</a></li>
<li><a class="" href="grouping-data.html"><span class="header-section-number">12</span> Grouping data</a></li>
<li><a class="" href="joining-data.html"><span class="header-section-number">13</span> Joining data</a></li>
<li><a class="" href="characters-and-strings.html"><span class="header-section-number">14</span> Characters and strings</a></li>
<li><a class="" href="de-duplication.html"><span class="header-section-number">15</span> De-duplication</a></li>
<li><a class="" href="iteration-loops-and-lists.html"><span class="header-section-number">16</span> Iteration, loops, and lists</a></li>
<li class="book-part">Analysis</li>
<li><a class="" href="descriptive-tables.html"><span class="header-section-number">17</span> Descriptive tables</a></li>
<li><a class="" href="simple-statistical-tests.html"><span class="header-section-number">18</span> Simple statistical tests</a></li>
<li><a class="" href="univariate-and-multivariable-regression.html"><span class="header-section-number">19</span> Univariate and multivariable regression</a></li>
<li><a class="" href="missing-data.html"><span class="header-section-number">20</span> Missing data</a></li>
<li><a class="" href="standardization.html"><span class="header-section-number">21</span> Standardization</a></li>
<li><a class="" href="moving-averages.html"><span class="header-section-number">22</span> Moving averages</a></li>
<li><a class="" href="time-series-and-outbreak-detection.html"><span class="header-section-number">23</span> Time series and outbreak detection</a></li>
<li><a class="" href="epidemic-modeling.html"><span class="header-section-number">24</span> Epidemic modeling</a></li>
<li><a class="" href="survey-analysis.html"><span class="header-section-number">25</span> Survey analysis</a></li>
<li><a class="" href="survival-analysis.html"><span class="header-section-number">26</span> Survival analysis</a></li>
<li><a class="" href="gis-basics.html"><span class="header-section-number">27</span> GIS basics</a></li>
<li class="book-part">Data Visualization</li>
<li><a class="" href="ggplot-tips.html"><span class="header-section-number">28</span> ggplot tips</a></li>
<li><a class="" href="epidemic-curves.html"><span class="header-section-number">29</span> Epidemic curves</a></li>
<li><a class="" href="plot-continuous-data.html"><span class="header-section-number">30</span> Plot continuous data</a></li>
<li><a class="" href="plot-categorical-data.html"><span class="header-section-number">31</span> Plot categorical data</a></li>
<li><a class="" href="tables-for-presentation.html"><span class="header-section-number">32</span> Tables for presentation</a></li>
<li><a class="" href="age-pyramids-and-likert-scales.html"><span class="header-section-number">33</span> Age pyramids and Likert-scales</a></li>
<li><a class="" href="diagrams-and-charts.html"><span class="header-section-number">34</span> Diagrams and charts</a></li>
<li><a class="" href="combinations-analysis.html"><span class="header-section-number">35</span> Combinations analysis</a></li>
<li><a class="" href="heat-tiles.html"><span class="header-section-number">36</span> Heat tiles</a></li>
<li><a class="" href="transmission-chains.html"><span class="header-section-number">37</span> Transmission chains</a></li>
<li><a class="" href="phylogenetic-trees.html"><span class="header-section-number">38</span> Phylogenetic trees</a></li>
<li><a class="" href="first-we-add-gender.html"><span class="header-section-number">39</span> First we add gender:</a></li>
<li><a class="" href="interactive-plots.html"><span class="header-section-number">40</span> Interactive plots</a></li>
<li class="book-part">Reports and dashboards</li>
<li><a class="" href="r-markdown-reports.html"><span class="header-section-number">41</span> R Markdown reports</a></li>
<li><a class="" href="organizing-routine-reports.html"><span class="header-section-number">42</span> Organizing routine reports</a></li>
<li><a class="" href="dashboards-with-r-markdown.html"><span class="header-section-number">43</span> Dashboards with R Markdown</a></li>
<li><a class="" href="dashboards-with-shiny.html"><span class="header-section-number">44</span> Dashboards with Shiny</a></li>
<li class="book-part">Miscellaneous</li>
<li><a class="" href="writing-functions.html"><span class="header-section-number">45</span> Writing functions</a></li>
<li><a class="" href="version-control-and-collaboration-with-github.html"><span class="header-section-number">46</span> Version control and collaboration with Github</a></li>
<li><a class="" href="directory-interactions.html"><span class="header-section-number">47</span> Directory interactions</a></li>
<li><a class="" href="errors-help.html"><span class="header-section-number">48</span> Errors &amp; help</a></li>
<li><a class="" href="relational-databases.html"><span class="header-section-number">49</span> Relational databases</a></li>
<li><a class="" href="r-on-network-drives.html"><span class="header-section-number">50</span> R on network drives</a></li>
<li><a class="" href="advanced-rstudio.html"><span class="header-section-number">51</span> Advanced RStudio</a></li>
</ul>

        <div class="book-extra">
          
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="download-book-and-data" class="section level1" number="2">
<h1>
<span class="header-section-number">2</span> Download book and data<a class="anchor" aria-label="anchor" href="#download-book-and-data"><i class="fas fa-link"></i></a>
</h1>
<div id="download-offline-handbook" class="section level2" number="2.1">
<h2>
<span class="header-section-number">2.1</span> Download offline handbook<a class="anchor" aria-label="anchor" href="#download-offline-handbook"><i class="fas fa-link"></i></a>
</h2>
<p>You can download the offline version of this handbook as an HTML file. Once downloaded, you can view the file in your web browser even if you no longer have internet access.</p>
<p>Note: The file is large (&gt;40MB). When you open it, it may take several minutes for the images and the Table of Contents to load. Also, the offline handbook displays as one long page, so please search for specific content using with Ctrl+f (Cmd-f).</p>
<p><strong>There are two ways you can download the handbook:</strong></p>
<div id="option-1-download-via-our-r-package" class="section level3 unnumbered">
<h3>Option 1: Download via our R package<a class="anchor" aria-label="anchor" href="#option-1-download-via-our-r-package"><i class="fas fa-link"></i></a>
</h3>
<p>First, run the following code to install our R package <strong>handbookfiles</strong> from the Github repository <em>epirhandbook</em>. This package in not on CRAN, so use the special function <code>p_install_gh()</code> to install it from Github.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">pacman</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pacman/man/p_install_gh.html">p_install_gh</a></span><span class="op">(</span><span class="st">"epirhandbook/handbookfiles"</span><span class="op">)</span></code></pre></div>
<p>Now, run the function <code>download_book()</code> to download the handbook to your computer. By default a pop-up window will appear, allowing you to select a save location. Alternatively, you can provide a file path to <code>save_loc =</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">handbookfiles</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/handbookfiles/man/download_book.html">download_book</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
</div>
<div id="option-2-manual-download-from-github" class="section level3 unnumbered">
<h3>Option 2: Manual download from Github<a class="anchor" aria-label="anchor" href="#option-2-manual-download-from-github"><i class="fas fa-link"></i></a>
</h3>
<p>If the approach above does not work for you, follow these instructions to download the file firstly from our Github repository.</p>
<ol style="list-style-type: decimal">
<li>Click this <a href="https://github.com/nsbatra/Epi_R_handbook/tree/master/offline_long">link to go the html file</a> in our Github repository’s “offline_long” folder.</li>
</ol>
<div class="inline-figure"><img src="images/download_offline.png" width="75%" height="50%" style="display: block; margin: auto;"></div>
<ol start="2" style="list-style-type: decimal">
<li>Click the “Download” button. A window will open with HTML source code.</li>
</ol>
<div class="inline-figure"><img src="images/download_offline_button.png" width="75%" height="50%" style="display: block; margin: auto;"></div>
<ol start="3" style="list-style-type: decimal">
<li>“Save As” the webpage, via right-click (windows) or Cmd-s (mac) - if given the option, set the file type as “Webpage, Complete”</li>
</ol>
<div class="inline-figure"><img src="images/download_offline_saveas.png" width="75%" height="50%" style="display: block; margin: auto;"></div>
</div>
</div>
<div id="download-data-to-follow-along" class="section level2" number="2.2">
<h2>
<span class="header-section-number">2.2</span> Download data to follow along<a class="anchor" aria-label="anchor" href="#download-data-to-follow-along"><i class="fas fa-link"></i></a>
</h2>
<div id="option-1-use-our-handbook_files-package" class="section level3" number="2.2.1">
<h3>
<span class="header-section-number">2.2.1</span> Option 1: Use our <strong>handbook_files</strong> package<a class="anchor" aria-label="anchor" href="#option-1-use-our-handbook_files-package"><i class="fas fa-link"></i></a>
</h3>
<p>UNDER CONSTRUCTION</p>
<p>You can use our R package <strong>handbook_files</strong> to save the data files to your computer. The package will do the work of downloading the files from our Github repository.</p>
</div>
<div id="option-2-manually-download-from-github" class="section level3" number="2.2.2">
<h3>
<span class="header-section-number">2.2.2</span> Option 2: Manually download from Github<a class="anchor" aria-label="anchor" href="#option-2-manually-download-from-github"><i class="fas fa-link"></i></a>
</h3>
<p>Alternatively, all the data in this handbook can be manually downloaded from the <strong><a href="https://github.com/nsbatra/Epi_R_handbook/tree/master/data">“data” folder</a></strong> of our Github repository. When you arrive at this folder, click the file you want to download.</p>
<p>To “follow along” with any of the handbook examples that use the <code>linelist</code> dataset, download and load either the “linelist_raw.xlsx” (for the Cleaning page), or “linelist_cleaned.rds” (for any other page).</p>
<p>Here are detailed instructions for downloading the data files from Github:</p>
<div id="raw-linelist-.xlsx" class="section level4 unnumbered">
<h4>“Raw” linelist (.xlsx)<a class="anchor" aria-label="anchor" href="#raw-linelist-.xlsx"><i class="fas fa-link"></i></a>
</h4>
<p>Use the “linelist_raw.xlsx” to follow-along with the <a href="cleaning-data-and-core-functions.html#cleaning-data-and-core-functions">Cleaning data and core functions</a> page. To download the <em>raw</em> linelist (uncleaned, with messy data), use the following appraoch.<br><em>Note: Use “linelist_cleaned.rds” to follow-along with any other page that uses the case linelist.</em></p>
<ol style="list-style-type: decimal">
<li>Go to the <strong><a href="https://github.com/nsbatra/Epi_R_handbook/tree/master/data/case_linelists">“data” folder</a></strong> of our Github repository<br>
</li>
<li>Click on “linelist_raw.xlsx”<br>
</li>
<li>Click the “Download” button as shown below<br>
</li>
<li>Save the file on your computer, and import it into R with the <code>import()</code> function from <strong>rio</strong>, as described in the page on <a href="import-and-export.html#import-and-export">Import and export</a>.</li>
</ol>
<div class="inline-figure"><img src="images/download_xlsx.png" width="75%" height="50%" style="display: block; margin: auto;"></div>
</div>
<div id="clean-linelist-.rds" class="section level4 unnumbered">
<h4>Clean linelist (.rds)<a class="anchor" aria-label="anchor" href="#clean-linelist-.rds"><i class="fas fa-link"></i></a>
</h4>
<p>Use the “linelist_cleaned.rds” to follow-along with any page that uses the case linelist (other than the data cleaning page). A .rds file is an R-specific file type that preserves column classes. This ensures you will have minimal cleaning to do after importing.</p>
<ol style="list-style-type: decimal">
<li>Go to the <strong><a href="https://github.com/nsbatra/Epi_R_handbook/tree/master/data/case_linelists">“data” folder</a></strong> of our Github repository<br>
</li>
<li>Click on “linelist_cleaned.rds”<br>
</li>
<li>Click the “Download” button as shown below<br>
</li>
<li>Save the file on your computer, and import it into R with the <code>import()</code> function from <strong>rio</strong>, as described in the page on <a href="import-and-export.html#import-and-export">Import and export</a>.</li>
</ol>
<div class="inline-figure"><img src="images/download_rds.png" width="75%" height="50%" style="display: block; margin: auto;"></div>
</div>
<div id="csv-files" class="section level4 unnumbered">
<h4>.csv files<a class="anchor" aria-label="anchor" href="#csv-files"><i class="fas fa-link"></i></a>
</h4>
<p>It is easy to download a relatively small CSV file directly from Github into R with an R command.</p>
<ol style="list-style-type: decimal">
<li>Go to the <strong><a href="https://github.com/nsbatra/Epi_R_handbook/tree/master/data">“data” folder</a></strong> of our Github repository<br>
</li>
<li>Open the appropriate subfolder and click on the .csv file of interest<br>
</li>
<li>Click on the “Raw” button (you will then see the “raw” csv data, as shown below)<br>
</li>
<li>Copy the URL (web address)<br>
</li>
<li>Use the URL in the <code>import()</code> R command, as shown below</li>
</ol>
<div class="inline-figure"><img src="images/download_csv_raw.png" width="75%" height="50%" style="display: block; margin: auto;"></div>
<p>Provide the URL to <code>import()</code> from <strong>rio</strong> as shown below.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Example of importing the country demographics csv file</span>
<span class="va">country_demographics</span> <span class="op">&lt;-</span> <span class="fu">import</span><span class="op">(</span><span class="st">"https://raw.githubusercontent.com/nsbatra/Epi_R_handbook/master/data/country_demographics.csv"</span><span class="op">)</span>

<span class="co"># Example of importing the likert-scale data csv file</span>
<span class="va">likert_data</span> <span class="op">&lt;-</span> <span class="fu">import</span><span class="op">(</span><span class="st">"https://raw.githubusercontent.com/nsbatra/Epi_R_handbook/master/data/likert_data.csv"</span><span class="op">)</span></code></pre></div>
<p>If this the above approach not work, you can click the “Raw” button, right-click on the webpage to “Save As” the data as a .txt file to your computer. This file can then be imported into R with <code>import()</code>.</p>
</div>
<div id="shapefiles" class="section level4 unnumbered">
<h4>Shapefiles<a class="anchor" aria-label="anchor" href="#shapefiles"><i class="fas fa-link"></i></a>
</h4>
<p>Shapefiles have many sub-component files, each with a different file extention. One file will have the “.shp” extension, but others may have “.dbf”, “.prj”, etc.</p>
<p>The <a href="gis-basics.html#gis-basics">GIS basics</a> page provides links to the <em>Humanitarian Data Exchange</em> website where you can download the shapefiles directly. For example, the health facility points can be downloaded <a href="https://data.humdata.org/dataset/hotosm_sierra_leone_health_facilities">here</a>. Download the points shapefile (shp) “.zip” folder. Once saved to your computer, “unzip” the folder. You will see several files with different extensions (e.g. “.shp”, “.prj”, “.shx”) - all these must be saved in the same folder. As described in the <a href="gis-basics.html#gis-basics">GIS basics</a> page, provide the filepath and name of the “.shp” file to <code>st_read()</code> from the <strong>sf</strong> package.</p>
<p>Alternatively, you can download the shapefiles from the R Handbook Github “data” folder (see the “gis” sub-folder). However, be aware that you will need to download <em>each</em> sub-file individually to your computer. In Github, click on each file individually and download them by clicking on the “Download” button. Below, you can see how the shapefile “sle_adm3” consists of many files - each of which must be downloaded from Github.</p>
<div class="inline-figure"><img src="images/download_shp.png" width="918" height="50%" style="display: block; margin: auto;"></div>
</div>
<div id="phylogenetic-trees-newick-files-saved-as-.txt" class="section level4 unnumbered">
<h4>Phylogenetic trees (newick files, saved as .txt)<a class="anchor" aria-label="anchor" href="#phylogenetic-trees-newick-files-saved-as-.txt"><i class="fas fa-link"></i></a>
</h4>
<p>See the page on <a href="phylogenetic-trees.html#phylogenetic-trees">Phylogenetic trees</a>. In Github “data” folder, navigate to the “phylo” subfolder where you can click on the .txt files. Click “Raw”, and right-click to download as for the HTML handbook file - save the .txt file on your computer. Then import it with <code>read.tree()</code> from the <strong>ape</strong> package.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html">read.tree</a></span><span class="op">(</span><span class="st">"Shigella_tree.txt"</span><span class="op">)</span></code></pre></div>
</div>
<div id="other-files" class="section level4 unnumbered">
<h4>Other files<a class="anchor" aria-label="anchor" href="#other-files"><i class="fas fa-link"></i></a>
</h4>
<p>Click on the file in the Gitub “data” folder, and use the “Download” button, as described above for .xlsx files.</p>
</div>
</div>
</div>
<div id="background-information" class="section level2" number="2.3">
<h2>
<span class="header-section-number">2.3</span> Background information<a class="anchor" aria-label="anchor" href="#background-information"><i class="fas fa-link"></i></a>
</h2>
<p>Below are more details about some of the data:</p>
<ul>
<li>The case linelist
<ul>
<li>A <em>simulated</em> Ebola outbreak, expanded by the handbook team from the <code>ebola_sim</code> practice dataset in the <strong>outbreaks</strong> package</li>
</ul>
</li>
<li>Aggregated counts
<ul>
<li>A simulated dataset of malaria counts by age, day, and facility in a fictional region, which can be downloaded from the data folder noted above as “district_count_data.xlsx”</li>
</ul>
</li>
<li>Time series and outbreak detection
<ul>
<li>Campylobacter cases reported in Germany 2002-2011. Available from <strong>surveillance</strong> package</li>
<li>Climate data (temperature in degrees celsius and rain fail in millimetres) in Germany 2002-2011. Downloaded from the EU Copernicus satellite reanalysis dataset using the <strong>ecmwfr</strong> package as explained in the time series page.</li>
</ul>
</li>
<li>GIS page shapefiles
<ul>
<li>Downloaded from the Humanitarian Data Exchange (HDX) - see links in the GIS basics page</li>
</ul>
</li>
<li>Phylogenetic tree data
<ul>
<li>Newick file of phylogenetic tree constructed from whole genome sequencing of 299 Shigella sonnei samples and corresponding sample data.</li>
<li>The Belgian samples and resulting data are kindly provided by the Belgian NRC for Salmonella and Shigella in the scope of a project conducted by an ECDC EUPHEM Fellow, and will also be published in a manuscript.<br>
</li>
<li>The international data are openly available on public databases (ncbi) and have been previously published.</li>
</ul>
</li>
</ul>
</div>
</div>



<div class="chapter-nav">
<div class="prev"><a href="editorial-and-technical-notes.html"><span class="header-section-number">1</span> Editorial and technical notes</a></div>
<div class="next"><a href="r-basics.html"><span class="header-section-number">3</span> R Basics</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#download-book-and-data"><span class="header-section-number">2</span> Download book and data</a></li>
<li>
<a class="nav-link" href="#download-offline-handbook"><span class="header-section-number">2.1</span> Download offline handbook</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#option-1-download-via-our-r-package">Option 1: Download via our R package</a></li>
<li><a class="nav-link" href="#option-2-manual-download-from-github">Option 2: Manual download from Github</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#download-data-to-follow-along"><span class="header-section-number">2.2</span> Download data to follow along</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#option-1-use-our-handbook_files-package"><span class="header-section-number">2.2.1</span> Option 1: Use our handbook_files package</a></li>
<li><a class="nav-link" href="#option-2-manually-download-from-github"><span class="header-section-number">2.2.2</span> Option 2: Manually download from Github</a></li>
</ul>
</li>
<li><a class="nav-link" href="#background-information"><span class="header-section-number">2.3</span> Background information</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
          
        </ul>
</div>
    </nav>
</div>
</div>
  

  

</div>
 <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>The Epidemiologist R Handbook</strong>" was written by the handbook team. It was last built on 2021-04-20.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>
</html>
