<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>15 Phylogenetic trees | The Epidemiologist R Handbook</title>
<meta name="author" content="the handbook team">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.2"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/header-attrs-2.8/header-attrs.js"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.2.5.1/tabs.js"></script><script src="libs/bs3compat-0.2.5.1/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="libs/htmlwidgets-1.5.3/htmlwidgets.js"></script><link href="libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="libs/datatables-binding-0.18/datatables.js"></script><link href="libs/dt-core-1.10.20/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="libs/dt-core-1.10.20/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="libs/dt-core-1.10.20/js/jquery.dataTables.min.js"></script><link href="libs/nouislider-7.0.10/jquery.nouislider.min.css" rel="stylesheet">
<script src="libs/nouislider-7.0.10/jquery.nouislider.min.js"></script><link href="libs/selectize-0.12.0/selectize.bootstrap3.css" rel="stylesheet">
<script src="libs/selectize-0.12.0/selectize.min.js"></script><link href="libs/crosstalk-1.1.1/css/crosstalk.css" rel="stylesheet">
<script src="libs/crosstalk-1.1.1/js/crosstalk.min.js"></script><script src="libs/kePrint-0.0.1/kePrint.js"></script><link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet">
<link href="libs/leaflet-1.3.1/leaflet.css" rel="stylesheet">
<script src="libs/leaflet-1.3.1/leaflet.js"></script><link href="libs/leafletfix-1.0.0/leafletfix.css" rel="stylesheet">
<script src="libs/proj4-2.6.2/proj4.min.js"></script><script src="libs/Proj4Leaflet-1.0.1/proj4leaflet.js"></script><link href="libs/rstudio_leaflet-1.3.1/rstudio_leaflet.css" rel="stylesheet">
<script src="libs/leaflet-binding-2.0.4.1/leaflet.js"></script><script src="libs/leaflet-providers-1.9.0/leaflet-providers_1.9.0.js"></script><script src="libs/leaflet-providers-plugin-2.0.4.1/leaflet-providers-plugin.js"></script><link href="libs/tabwid-1.0.0/tabwid.css" rel="stylesheet">
<script src="libs/viz-1.8.2/viz.js"></script><link href="libs/DiagrammeR-styles-0.2/styles.css" rel="stylesheet">
<script src="libs/grViz-binding-1.0.6.1/grViz.js"></script><script src="libs/d3-4.5.0/d3.min.js"></script><script src="libs/sankey-1/sankey.js"></script><script src="libs/sankeyNetwork-binding-0.4/sankeyNetwork.js"></script><script src="libs/plotly-binding-4.9.3/plotly.js"></script><script src="libs/typedarray-0.1/typedarray.min.js"></script><link href="libs/plotly-htmlwidgets-css-1.57.1/plotly-htmlwidgets.css" rel="stylesheet">
<script src="libs/plotly-main-1.57.1/plotly-latest.min.js"></script><link href="libs/vis-4.20.1/vis.css" rel="stylesheet">
<script src="libs/vis-4.20.1/vis.min.js"></script><script src="libs/visNetwork-binding-2.0.9/visNetwork.js"></script><link href="libs/font-awesome-4.7.0/css/font-awesome.min.css" rel="stylesheet">
<script src="libs/plotly-basic-1.57.1/plotly-basic-1.57.1.min.js"></script><script src="libs/plotly-cartesian-1.57.1/plotly-cartesian-1.57.1.min.js"></script><script src="libs/proj4js-2.3.15/proj4.js"></script><link href="libs/highcharts-8.1.2/css/motion.css" rel="stylesheet">
<script src="libs/highcharts-8.1.2/highcharts.js"></script><script src="libs/highcharts-8.1.2/highcharts-3d.js"></script><script src="libs/highcharts-8.1.2/highcharts-more.js"></script><script src="libs/highcharts-8.1.2/modules/stock.js"></script><script src="libs/highcharts-8.1.2/modules/map.js"></script><script src="libs/highcharts-8.1.2/modules/annotations.js"></script><script src="libs/highcharts-8.1.2/modules/data.js"></script><script src="libs/highcharts-8.1.2/modules/drilldown.js"></script><script src="libs/highcharts-8.1.2/modules/item-series.js"></script><script src="libs/highcharts-8.1.2/modules/offline-exporting.js"></script><script src="libs/highcharts-8.1.2/modules/overlapping-datalabels.js"></script><script src="libs/highcharts-8.1.2/modules/exporting.js"></script><script src="libs/highcharts-8.1.2/modules/export-data.js"></script><script src="libs/highcharts-8.1.2/modules/funnel.js"></script><script src="libs/highcharts-8.1.2/modules/heatmap.js"></script><script src="libs/highcharts-8.1.2/modules/treemap.js"></script><script src="libs/highcharts-8.1.2/modules/sankey.js"></script><script src="libs/highcharts-8.1.2/modules/dependency-wheel.js"></script><script src="libs/highcharts-8.1.2/modules/organization.js"></script><script src="libs/highcharts-8.1.2/modules/solid-gauge.js"></script><script src="libs/highcharts-8.1.2/modules/streamgraph.js"></script><script src="libs/highcharts-8.1.2/modules/sunburst.js"></script><script src="libs/highcharts-8.1.2/modules/vector.js"></script><script src="libs/highcharts-8.1.2/modules/wordcloud.js"></script><script src="libs/highcharts-8.1.2/modules/xrange.js"></script><script src="libs/highcharts-8.1.2/modules/tilemap.js"></script><script src="libs/highcharts-8.1.2/modules/venn.js"></script><script src="libs/highcharts-8.1.2/modules/gantt.js"></script><script src="libs/highcharts-8.1.2/modules/timeline.js"></script><script src="libs/highcharts-8.1.2/modules/parallel-coordinates.js"></script><script src="libs/highcharts-8.1.2/modules/bullet.js"></script><script src="libs/highcharts-8.1.2/modules/coloraxis.js"></script><script src="libs/highcharts-8.1.2/modules/dumbbell.js"></script><script src="libs/highcharts-8.1.2/modules/lollipop.js"></script><script src="libs/highcharts-8.1.2/modules/series-label.js"></script><script src="libs/highcharts-8.1.2/plugins/motion.js"></script><script src="libs/highcharts-8.1.2/custom/reset.js"></script><script src="libs/highcharts-8.1.2/modules/boost.js"></script><script src="libs/highchart-binding-0.8.2/highchart.js"></script><script src="https://cdn.jsdelivr.net/autocomplete.js/0/autocomplete.jquery.min.js"></script><script src="https://cdn.jsdelivr.net/npm/mark.js@8.11.1/dist/mark.min.js"></script><!-- CSS --><link rel="stylesheet" href="style_bs4.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">The Epidemiologist R Handbook</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"></a></li>
<li><a class="" href="epidemic-modeling.html"><span class="header-section-number">1</span> Epidemic modeling</a></li>
<li><a class="" href="contact-tracing.html"><span class="header-section-number">2</span> Contact tracing</a></li>
<li><a class="" href="survey-analysis.html"><span class="header-section-number">3</span> Survey analysis</a></li>
<li><a class="" href="survival-analysis.html"><span class="header-section-number">4</span> Survival analysis</a></li>
<li><a class="" href="gis-basics.html"><span class="header-section-number">5</span> GIS basics</a></li>
<li class="book-part">Data Visualization</li>
<li><a class="" href="tables-for-presentation.html"><span class="header-section-number">6</span> Tables for presentation</a></li>
<li><a class="" href="ggplot-basics.html"><span class="header-section-number">7</span> ggplot basics</a></li>
<li><a class="" href="ggplot-tips.html"><span class="header-section-number">8</span> ggplot tips</a></li>
<li><a class="" href="epidemic-curves.html"><span class="header-section-number">9</span> Epidemic curves</a></li>
<li><a class="" href="demographic-pyramids-and-likert-scales.html"><span class="header-section-number">10</span> Demographic pyramids and Likert-scales</a></li>
<li><a class="" href="heat-plots.html"><span class="header-section-number">11</span> Heat plots</a></li>
<li><a class="" href="diagrams-and-charts.html"><span class="header-section-number">12</span> Diagrams and charts</a></li>
<li><a class="" href="combinations-analysis.html"><span class="header-section-number">13</span> Combinations analysis</a></li>
<li><a class="" href="transmission-chains.html"><span class="header-section-number">14</span> Transmission chains</a></li>
<li><a class="active" href="phylogenetic-trees.html"><span class="header-section-number">15</span> Phylogenetic trees</a></li>
<li><a class="" href="interactive-plots.html"><span class="header-section-number">16</span> Interactive plots</a></li>
<li class="book-part">Reports and dashboards</li>
<li><a class="" href="reports-with-r-markdown.html"><span class="header-section-number">17</span> Reports with R Markdown</a></li>
<li><a class="" href="organizing-routine-reports.html"><span class="header-section-number">18</span> Organizing routine reports</a></li>
<li><a class="" href="dashboards-with-r-markdown.html"><span class="header-section-number">19</span> Dashboards with R Markdown</a></li>
<li><a class="" href="dashboards-with-shiny.html"><span class="header-section-number">20</span> Dashboards with Shiny</a></li>
<li class="book-part">Miscellaneous</li>
<li><a class="" href="writing-functions.html"><span class="header-section-number">21</span> Writing functions</a></li>
<li><a class="" href="directory-interactions.html"><span class="header-section-number">22</span> Directory interactions</a></li>
<li><a class="" href="version-control-and-collaboration-with-git-and-github.html"><span class="header-section-number">23</span> Version control and collaboration with Git and Github</a></li>
<li><a class="" href="common-errors.html"><span class="header-section-number">24</span> Common errors</a></li>
<li><a class="" href="getting-help.html"><span class="header-section-number">25</span> Getting help</a></li>
<li><a class="" href="r-on-network-drives.html"><span class="header-section-number">26</span> R on network drives</a></li>
<li><a class="" href="data-table.html"><span class="header-section-number">27</span> Data Table</a></li>
</ul>

        <div class="book-extra">
          
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="phylogenetic-trees" class="section level1" number="15">
<h1>
<span class="header-section-number">15</span> Phylogenetic trees<a class="anchor" aria-label="anchor" href="#phylogenetic-trees"><i class="fas fa-link"></i></a>
</h1>
<!-- ======================================================= -->
<div id="overview-5" class="section level2" number="15.1">
<h2>
<span class="header-section-number">15.1</span> Overview<a class="anchor" aria-label="anchor" href="#overview-5"><i class="fas fa-link"></i></a>
</h2>
<p><strong>Phylogenetic trees</strong> are used to visualize and describe the relatedness and evolution of organisms based on the sequence of their genetic code.</p>
<p>They can be constructed from genetic sequences using distance-based methods (such as neighbor-joining method) or character-based methods (such as maximum likelihood and Bayesian Markov Chain Monte Carlo method). Next-generation sequencing (NGS) has become more affordable and is becoming more widely used in public health to describe pathogens causing infectious diseases. Portable sequencing devices decrease the turn around time and hold promises to make data available for the support of outbreak investigation in real-time. NGS data can be used to identify the origin or source of an outbreak strain and its propagation, as well as determine presence of antimicrobial resistance genes. To visualize the genetic relatedness between samples a phylogenetic tree is constructed.</p>
<p>In this page we will learn how to use the <strong>ggtree</strong> package, which allows for combined visualization of phylogenetic trees with additional sample data in form of a dataframe. This will enable us to observe patterns and improve understanding of the outbreak dynamic.</p>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_overview_graph-1.png" width="80%" style="display: block; margin: auto;"></div>
<!-- ======================================================= -->
</div>
<div id="preparation-16" class="section level2" number="15.2">
<h2>
<span class="header-section-number">15.2</span> Preparation<a class="anchor" aria-label="anchor" href="#preparation-16"><i class="fas fa-link"></i></a>
</h2>
<div id="load-packages-12" class="section level3 unnumbered">
<h3>Load packages<a class="anchor" aria-label="anchor" href="#load-packages-12"><i class="fas fa-link"></i></a>
</h3>
<p>This code chunk shows the loading of required packages. In this handbook we emphasize <code>p_load()</code> from <strong>pacman</strong>, which installs the package if necessary <em>and</em> loads it for use. You can also load installed packages with <code><a href="https://rdrr.io/r/base/library.html">library()</a></code> from <strong>base</strong> R. See the page on [R basics] for more information on R packages.</p>
<div class="sourceCode" id="cb581"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">pacman</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pacman/man/p_load.html">p_load</a></span><span class="op">(</span>
  <span class="va">rio</span>,             <span class="co"># import/export</span>
  <span class="va">here</span>,            <span class="co"># relative file paths</span>
  <span class="va">tidyverse</span>,       <span class="co"># general data management and visualization</span>
  <span class="va">ape</span>,             <span class="co"># to import and export phylogenetic files</span>
  <span class="va">ggtree</span>,          <span class="co"># to visualize phylogenetic files</span>
  <span class="va">treeio</span>,          <span class="co"># to visualize phylogenetic files</span>
  <span class="va">ggnewscale</span><span class="op">)</span>      <span class="co"># to add additional layers of color schemes</span></code></pre></div>
</div>
<div id="import-data-9" class="section level3 unnumbered">
<h3>Import data<a class="anchor" aria-label="anchor" href="#import-data-9"><i class="fas fa-link"></i></a>
</h3>
<p>The data for this page can be downloaded with the instructions on the [Download handbook and data] page.</p>
<p>There are several different formats in which a phylogenetic tree can be stored (eg. Newick, NEXUS, Phylip). A common one is the Newick file format (.nwk), which is the standard for representing trees in computer-readable form. This means an entire tree can be expressed in a string format such as “((t2:0.04,t1:0.34):0.89,(t5:0.37,(t4:0.03,t3:0.67):0.9):0.59);”, listing all nodes and tips and their relationship (branch length) to each other.</p>
<p>Note: It is important to understand that the phylogenetic tree file in itself does not contain sequencing data, but is merely the result of the genetic distances between the sequences. We therefore cannot extract sequencing data from a tree file.</p>
<p>First, we use the <code>read.tree()</code> function from <strong>ape</strong> package to import a Newick phylogenetic tree file in .txt format, and store it in a list object of class “phylo”. If necessary, use the <code>here()</code> function from the <strong>here</strong> package to specify the relative file path.</p>
<p>Note: In this case the newick tree is saved as a .txt file for easier handling and downloading from Github.</p>
<div class="sourceCode" id="cb582"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html">read.tree</a></span><span class="op">(</span><span class="st">"Shigella_tree.txt"</span><span class="op">)</span></code></pre></div>
<p>We inspect our tree object and see it contains 299 tips (or samples) and 236 nodes.</p>
<div class="sourceCode" id="cb583"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tree</span></code></pre></div>
<pre><code>## 
## Phylogenetic tree with 299 tips and 236 internal nodes.
## 
## Tip labels:
##   SRR5006072, SRR4192106, S18BD07865, S18BD00489, S17BD08906, S17BD05939, ...
## Node labels:
##   17, 29, 100, 67, 100, 100, ...
## 
## Rooted; includes branch lengths.</code></pre>
<p>Second, we import a table stored as a .csv file with additional information for each sequenced sample, such as gender, country of origin and attributes for antimicrobial resistance, using the <code>import()</code> function from the <strong>rio</strong> package:</p>
<div class="sourceCode" id="cb585"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sample_data</span> <span class="op">&lt;-</span> <span class="fu">import</span><span class="op">(</span><span class="st">"sample_data_Shigella_tree.csv"</span><span class="op">)</span></code></pre></div>
<p>Below are the first 50 rows of the data:</p>
<div id="htmlwidget-3f7087ea140e354d1b53" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-3f7087ea140e354d1b53">{"x":{"filter":"none","data":[["S17BD05944","S15BD07413","S18BD07247","S19BD07384","S18BD07338","S18BD02657","S19BD06884","S16BD02574","S16BD04069","S17BD08170","S15BD07250","S17BD07935","S17BD00384","S19BD06034","S17BD08948","S17BD04542","S17BD07399","S17BD08233","S19BD07381","S17BD07961","S18BD09176","S19BD07393","S18BD02580","S17BD02526","S17BD06833","S18BD01024","S18BD03411","S18BD09515","S17BD07332","S18BD06598","S17BD02164","S19BD06386","S19BD06283","S19BD06285","S19BD05221","S18BD06749","S17BD02088","S17BD07692","S14BD03573","S17BD01135","S19BD06689","S15BD06969","S18BD01106","S17BD00525","S19BD07036","S18BD06401","S18BD06486","S17BD06426","S17BD05916","S18BD07708"],["Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei"],["Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium"],["Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe"],[null,"Georgia",null,"Morocco",null,null,null,"Thailand",null,null,null,null,null,null,null,"Iran",null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,"Morocco","France","Turkey","Egypt","Tunisia","Congo",null,null,null,null,null,null,null,null,null,null,"Egypt",null,null],["2017","2015","2018","2019","2018","2018","2019","2016","2016","2017","2015","2017","2017","2019","2017","2017","2017","2017","2019","2017","2018","2019","2018","2017","2017","2018","2018","2018","2017","2018","2017","2019","2019","2019","2019","2018","2017","2017","2014","2017","2019","2015","2018","2017","2019","2018","2018","2017","2017","2018"],["Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes"],["NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL"],[null,"Male","Female","Female","Male","Male","Male","Male","Male","Male","Female",null,"Male","Male","Female","Male","Male","Female","Female","Male","Male","Male","Male","Female","Male","Female","Female","Male","Female","Male","Female","Female","Female","Female","Female","Female","Male","Male",null,"female","Male","Female","Male","Male","Male","Male","Male","Male","Male","Female"],[null,null,null,null,null,"gyrA D87Y","gyrA D87Y","gyrA D87Y","gyrA D87Y",null,null,"gyrA D87Y",null,null,"gyrA D87Y",null,null,null,null,null,null,null,null,null,null,"gyrA D87Y","gyrA S83L",null,null,null,null,null,null,null,null,null,null,"gyrA D87Y","gyrA D87Y",null,null,null,"gyrA D87Y",null,null,null,null,null,null,null],[null,null,null,null,null,"mphA","mphA",null,null,null,null,"mphA",null,null,"mphA",null,null,null,null,null,null,null,null,null,null,"mphA","mphA",null,null,null,null,null,null,null,"mphA/ermB",null,null,"mphA",null,null,null,null,"mphA",null,null,null,null,null,null,null],[8,0,16,8,8,64,64,0,0,64,0,64,8,8,64,8,8,8,8,8,8,16,8,8,8,64,64,8,8,8,8,8,8,2,64,8,8,64,0,8,8,0,64,8,4,8,8,8,8,8],[0.25,0,0.03,0.02,0.02,0.5,0.5,0,0,8,0,0.5,0.02,0.02,0.5,0.25,0.02,0.02,0.02,0.02,0.02,0.12,0.02,0.02,0.02,0.5,0.25,0.03,0.02,0.02,0.02,0.02,0.02,0.02,0.02,0.02,0.02,1,0,0.02,0.02,0,0.5,0.02,0.02,4,0.02,0.02,0.25,0.02]],"container":"<table class=\"white-space: nowrap\">\n  <thead>\n    <tr>\n      <th>Sample_ID<\/th>\n      <th>serotype<\/th>\n      <th>Country<\/th>\n      <th>Continent<\/th>\n      <th>Travel_history<\/th>\n      <th>Year<\/th>\n      <th>Belgium<\/th>\n      <th>Source<\/th>\n      <th>Gender<\/th>\n      <th>gyrA_mutations<\/th>\n      <th>macrolide_resistance_genes<\/th>\n      <th>MIC_AZM<\/th>\n      <th>MIC_CIP<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"pageLength":5,"scrollX":true,"columnDefs":[{"className":"dt-right","targets":[11,12]}],"order":[],"autoWidth":false,"orderClasses":false,"lengthMenu":[5,10,25,50,100]}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="clean-and-inspect" class="section level3 unnumbered">
<h3>Clean and inspect<a class="anchor" aria-label="anchor" href="#clean-and-inspect"><i class="fas fa-link"></i></a>
</h3>
<p>We clean and inspect our data: In order to assign the correct sample data to the phylogenetic tree, the values in the column <code>Sample_ID</code> in the <code>sample_data</code> data frame need to match the <code>tip.labels</code> values in the <code>tree</code> file:</p>
<p>We check the formatting of the <code>tip.labels</code> in the <code>tree</code> file by looking at the first 6 entries using with <code><a href="https://rdrr.io/r/utils/head.html">head()</a></code> from <strong>base</strong> R.</p>
<div class="sourceCode" id="cb586"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">tree</span><span class="op">$</span><span class="va">tip.label</span><span class="op">)</span> </code></pre></div>
<pre><code>## [1] "SRR5006072" "SRR4192106" "S18BD07865" "S18BD00489" "S17BD08906" "S17BD05939"</code></pre>
<p>We also make sure the first column in our <code>sample_data</code> data frame is <code>Sample_ID</code>. We look at the column names of our dataframe using <code><a href="https://rdrr.io/r/base/colnames.html">colnames()</a></code> from <strong>base</strong> R.</p>
<div class="sourceCode" id="cb588"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">sample_data</span><span class="op">)</span>   </code></pre></div>
<pre><code>##  [1] "Sample_ID"                  "serotype"                   "Country"                    "Continent"                  "Travel_history"            
##  [6] "Year"                       "Belgium"                    "Source"                     "Gender"                     "gyrA_mutations"            
## [11] "macrolide_resistance_genes" "MIC_AZM"                    "MIC_CIP"</code></pre>
<p>We look at the <code>Sample_IDs</code> in the data frame to make sure the formatting is the same than in the <code>tip.label</code> (eg. letters are all capitals, no extra underscores <code>_</code> between letters and numbers, etc.)</p>
<div class="sourceCode" id="cb590"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">sample_data</span><span class="op">$</span><span class="va">Sample_ID</span><span class="op">)</span> <span class="co"># we again inspect only the first 6 using head()</span></code></pre></div>
<pre><code>## [1] "S17BD05944" "S15BD07413" "S18BD07247" "S19BD07384" "S18BD07338" "S18BD02657"</code></pre>
<p>We can also compare if all samples are present in the <code>tree</code> file and vice versa by generating a logical vector of TRUE or FALSE where they do or do not match. These are not printed here, for simplicity.</p>
<div class="sourceCode" id="cb592"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sample_data</span><span class="op">$</span><span class="va">Sample_ID</span> <span class="op">%in%</span> <span class="va">tree</span><span class="op">$</span><span class="va">tip.label</span>

<span class="va">tree</span><span class="op">$</span><span class="va">tip.label</span> <span class="op">%in%</span> <span class="va">sample_data</span><span class="op">$</span><span class="va">Sample_ID</span></code></pre></div>
<p>We can use these vectors to show any sample IDs that are not on the tree (there are none).</p>
<div class="sourceCode" id="cb593"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sample_data</span><span class="op">$</span><span class="va">Sample_ID</span><span class="op">[</span><span class="op">!</span><span class="va">tree</span><span class="op">$</span><span class="va">tip.label</span> <span class="op">%in%</span> <span class="va">sample_data</span><span class="op">$</span><span class="va">Sample_ID</span><span class="op">]</span></code></pre></div>
<pre><code>## character(0)</code></pre>
<p>Upon inspection we can see that the format of <code>Sample_ID</code> in the dataframe corresponds to the format of sample names at the <code>tip.labels</code>. These do not have to be sorted in the same order to be matched.</p>
<p>We are ready to go!</p>
<!-- ======================================================= -->
</div>
</div>
<div id="simple-tree-visualization" class="section level2" number="15.3">
<h2>
<span class="header-section-number">15.3</span> Simple tree visualization<a class="anchor" aria-label="anchor" href="#simple-tree-visualization"><i class="fas fa-link"></i></a>
</h2>
<div id="different-tree-layouts" class="section level3 unnumbered">
<h3>Different tree layouts<a class="anchor" aria-label="anchor" href="#different-tree-layouts"><i class="fas fa-link"></i></a>
</h3>
<p><strong>ggtree</strong> offers many different layout formats and some may be more suitable for your specific purpose than others. Below are a few demonstrations. For other options see this <a href="http://yulab-smu.top/treedata-book/chapter4.html">online book</a>.</p>
<p>Here are some example tree layouts:</p>
<div class="sourceCode" id="cb595"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ggtree</span><span class="op">(</span><span class="va">tree</span><span class="op">)</span>                                            <span class="co"># simple linear tree</span>
<span class="fu">ggtree</span><span class="op">(</span><span class="va">tree</span>,  branch.length <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span>                   <span class="co"># simple linear tree with all tips aligned</span>
<span class="fu">ggtree</span><span class="op">(</span><span class="va">tree</span>, layout<span class="op">=</span><span class="st">"circular"</span><span class="op">)</span>                         <span class="co"># simple circular tree</span>
<span class="fu">ggtree</span><span class="op">(</span><span class="va">tree</span>, layout<span class="op">=</span><span class="st">"circular"</span>, branch.length <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="co"># simple circular tree with all tips aligned</span></code></pre></div>
<p><img src="_main_files/figure-html/phylogenetic_trees_example_formats-1.png" width="50%"><img src="_main_files/figure-html/phylogenetic_trees_example_formats-2.png" width="50%"><img src="_main_files/figure-html/phylogenetic_trees_example_formats-3.png" width="50%"><img src="_main_files/figure-html/phylogenetic_trees_example_formats-4.png" width="50%"></p>
</div>
<div id="simple-tree-plus-sample-data" class="section level3 unnumbered">
<h3>Simple tree plus sample data<a class="anchor" aria-label="anchor" href="#simple-tree-plus-sample-data"><i class="fas fa-link"></i></a>
</h3>
<p>The <strong>%&lt;+%</strong> operator is used to connect the <code>sample_data</code> data frame to the <code>tree</code> file.
The most easy annotation of your tree is the addition of the sample names at the tips, as well as coloring of tip points and if desired the branches:</p>
<p>Here is an example of a circular tree:</p>
<div class="sourceCode" id="cb596"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ggtree</span><span class="op">(</span><span class="va">tree</span>, layout <span class="op">=</span> <span class="st">"circular"</span>, branch.length <span class="op">=</span> <span class="st">'none'</span><span class="op">)</span> <span class="op">%&lt;+%</span> <span class="va">sample_data</span> <span class="op">+</span> <span class="co"># %&lt;+% adds dataframe with sample data to tree</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/AsIs.html">I</a></span><span class="op">(</span><span class="va">Belgium</span><span class="op">)</span><span class="op">)</span><span class="op">+</span>                       <span class="co"># color the branches according to a variable in your dataframe</span>
  <span class="fu">scale_color_manual</span><span class="op">(</span>
    name <span class="op">=</span> <span class="st">"Sample Origin"</span>,                      <span class="co"># name of your color scheme (will show up in the legend like this)</span>
    breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Yes"</span>, <span class="st">"No"</span><span class="op">)</span>,                     <span class="co"># the different options in your variable</span>
    labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"NRCSS Belgium"</span>, <span class="st">"Other"</span><span class="op">)</span>,        <span class="co"># how you want the different options named in your legend, allows for formatting</span>
    values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"black"</span><span class="op">)</span>,                  <span class="co"># the color you want to assign to the variable </span>
    na.value <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span> <span class="op">+</span>                        <span class="co"># color NA values in black as well</span>
  <span class="fu">new_scale_color</span><span class="op">(</span><span class="op">)</span><span class="op">+</span>                             <span class="co"># allows to add an additional color scheme for another variable</span>
    <span class="fu">geom_tippoint</span><span class="op">(</span>
      mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">Continent</span><span class="op">)</span>,          <span class="co"># tip color by continent. You may change shape adding "shape = "</span>
      size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span><span class="op">+</span>                               <span class="co"># define the size of the point at the tip</span>
  <span class="fu">scale_color_brewer</span><span class="op">(</span>
    name <span class="op">=</span> <span class="st">"Continent"</span>,                    <span class="co"># name of your color scheme (will show up in the legend like this)</span>
    palette <span class="op">=</span> <span class="st">"Set1"</span>,                      <span class="co"># we choose a set of colors coming with the brewer package</span>
    na.value <span class="op">=</span> <span class="st">"grey"</span><span class="op">)</span> <span class="op">+</span>                    <span class="co"># for the NA values we choose the color grey</span>
  <span class="fu">geom_tiplab</span><span class="op">(</span>                             <span class="co"># adds name of sample to tip of its branch </span>
    color <span class="op">=</span> <span class="st">'black'</span>,                       <span class="co"># (add as many text lines as you wish with + , but you may need to adjust offset value to place them next to each other)</span>
    offset <span class="op">=</span> <span class="fl">1</span>,
    size <span class="op">=</span> <span class="fl">1</span>,
    geom <span class="op">=</span> <span class="st">"text"</span>,
    align <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">+</span>    
  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Phylogenetic tree of Shigella sonnei"</span><span class="op">)</span><span class="op">+</span>       <span class="co"># title of your graph</span>
  <span class="fu">theme</span><span class="op">(</span>
    axis.title.x <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>, <span class="co"># removes x-axis title</span>
    axis.title.y <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>, <span class="co"># removes y-axis title</span>
    legend.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>    <span class="co"># defines font size and format of the legend title</span>
      face <span class="op">=</span> <span class="st">"bold"</span>,
      size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span>,   
    legend.text<span class="op">=</span><span class="fu">element_text</span><span class="op">(</span>       <span class="co"># defines font size and format of the legend text</span>
      face <span class="op">=</span> <span class="st">"bold"</span>,
      size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>,  
    plot.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>      <span class="co"># defines font size and format of the plot title</span>
      size <span class="op">=</span> <span class="fl">12</span>,
      face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span>,  
    legend.position <span class="op">=</span> <span class="st">"bottom"</span>,     <span class="co"># defines placement of the legend</span>
    legend.box <span class="op">=</span> <span class="st">"vertical"</span>,        <span class="co"># defines placement of the legend</span>
    legend.margin <span class="op">=</span> <span class="fu">margin</span><span class="op">(</span><span class="op">)</span><span class="op">)</span>   </code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_adding_sampledata-1.png" width="672" style="display: block; margin: auto;"></div>
<p>You can export your tree plot with <code>ggsave()</code> as you would any other ggplot object. Written this way, <code>ggsave()</code> saves the last image produced to the file path you specify. Remember that you can use <code>here()</code> and relative file paths to easily save in subfolders, etc.</p>
<div class="sourceCode" id="cb597"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ggsave</span><span class="op">(</span><span class="st">"example_tree_circular_1.png"</span>, width <span class="op">=</span> <span class="fl">12</span>, height <span class="op">=</span> <span class="fl">14</span><span class="op">)</span></code></pre></div>
<!-- ======================================================= -->
</div>
</div>
<div id="tree-manipulation" class="section level2" number="15.4">
<h2>
<span class="header-section-number">15.4</span> Tree manipulation<a class="anchor" aria-label="anchor" href="#tree-manipulation"><i class="fas fa-link"></i></a>
</h2>
<p>Sometimes you may have a very large phylogenetic tree and you are only interested in one part of the tree. For example, if you produced a tree including historical or international samples to get a large overview of where your dataset might fit in the bigger picture. But then to look closer at your data you want to inspect only that portion of the bigger tree.</p>
<p>Since the phylogenetic tree file is just the output of sequencing data analysis, we can not manipulate the order of the nodes and branches in the file itself. These have already been determined in previous analysis from the raw NGS data. We are able though to zoom into parts, hide parts and even subset part of the tree.</p>
<div id="zoom-in" class="section level3 unnumbered">
<h3>Zoom in<a class="anchor" aria-label="anchor" href="#zoom-in"><i class="fas fa-link"></i></a>
</h3>
<p>If you don’t want to “cut” your tree, but only inspect part of it more closely you can zoom in to view a specific part.</p>
<p>First, we plot the entire tree in linear format and add numeric labels to each node in the tree.</p>
<div class="sourceCode" id="cb598"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">ggtree</span><span class="op">(</span><span class="va">tree</span>,<span class="op">)</span> <span class="op">%&lt;+%</span> <span class="va">sample_data</span> <span class="op">+</span>
  <span class="fu">geom_tiplab</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span> <span class="op">+</span>                <span class="co"># labels the tips of all branches with the sample name in the tree file</span>
  <span class="fu">geom_text2</span><span class="op">(</span>
    mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>subset <span class="op">=</span> <span class="op">!</span><span class="va">isTip</span>,
                  label <span class="op">=</span> <span class="va">node</span><span class="op">)</span>,
    size <span class="op">=</span> <span class="fl">5</span>,
    color <span class="op">=</span> <span class="st">"darkred"</span>,
    hjust <span class="op">=</span> <span class="fl">1</span>,
    vjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>                            <span class="co"># labels all the nodes in the tree</span>

<span class="va">p</span>  <span class="co"># print</span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_zoom_in-1.png" width="50%" style="display: block; margin: auto;"></div>
<p>To zoom in to one particular branch (sticking out to the right), use <code>viewClade()</code> on the ggtree object <code>p</code> and provide the node number to get a closer look:</p>
<div class="sourceCode" id="cb599"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">viewClade</span><span class="op">(</span><span class="va">p</span>, node <span class="op">=</span> <span class="fl">452</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_zoom_in_452-1.png" width="50%" style="display: block; margin: auto;"></div>
</div>
<div id="collapsing-branches" class="section level3 unnumbered">
<h3>Collapsing branches<a class="anchor" aria-label="anchor" href="#collapsing-branches"><i class="fas fa-link"></i></a>
</h3>
<p>However, we may want to ignore this branch and can collapse it at that same node (node nr. 452) using <code>collapse()</code>. This tree is defined as <code>p_collapsed</code>.</p>
<div class="sourceCode" id="cb600"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p_collapsed</span> <span class="op">&lt;-</span> <span class="fu">collapse</span><span class="op">(</span><span class="va">p</span>, node <span class="op">=</span> <span class="fl">452</span><span class="op">)</span>
<span class="va">p_collapsed</span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_collapse_452-1.png" width="50%" style="display: block; margin: auto;"></div>
<p>For clarity, when we print <code>p_collapsed</code>, we add a <code>geom_point2()</code> (a blue diamond) at the node of the collapsed branch.</p>
<div class="sourceCode" id="cb601"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p_collapsed</span> <span class="op">+</span> 
<span class="fu">geom_point2</span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>subset <span class="op">=</span> <span class="op">(</span><span class="va">node</span> <span class="op">==</span> <span class="fl">452</span><span class="op">)</span><span class="op">)</span>,  <span class="co"># we assign a symbol to the collapsed node</span>
            size <span class="op">=</span> <span class="fl">5</span>,                     <span class="co"># define the size of the symbol</span>
            shape <span class="op">=</span> <span class="fl">23</span>,                   <span class="co"># define the shape of the symbol</span>
            fill <span class="op">=</span> <span class="st">"steelblue"</span><span class="op">)</span>           <span class="co"># define the color of the symbol</span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/unnamed-chunk-468-1.png" width="672"></div>
</div>
<div id="subsetting-a-tree" class="section level3 unnumbered">
<h3>Subsetting a tree<a class="anchor" aria-label="anchor" href="#subsetting-a-tree"><i class="fas fa-link"></i></a>
</h3>
<p>If we want to make a more permanent change and create a new, reduced tree to work with we can subset part of it with <code>tree_subset()</code>. Then you can save it as new newick tree file or .txt file.</p>
<p>First, we inspect the tree nodes and tip labels in order to decide what to subset.</p>
<div class="sourceCode" id="cb602"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ggtree</span><span class="op">(</span>
  <span class="va">tree</span>,
  branch.length <span class="op">=</span> <span class="st">'none'</span>,
  layout <span class="op">=</span> <span class="st">'circular'</span><span class="op">)</span> <span class="op">%&lt;+%</span> <span class="va">sample_data</span> <span class="op">+</span>               <span class="co"># we add the asmple data using the %&lt;+% operator</span>
  <span class="fu">geom_tiplab</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">+</span>                                <span class="co"># label tips of all branches with sample name in tree file</span>
  <span class="fu">geom_text2</span><span class="op">(</span>
    mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>subset <span class="op">=</span> <span class="op">!</span><span class="va">isTip</span>, label <span class="op">=</span> <span class="va">node</span><span class="op">)</span>,
    size <span class="op">=</span> <span class="fl">3</span>,
    color <span class="op">=</span> <span class="st">"darkred"</span><span class="op">)</span> <span class="op">+</span>                                <span class="co"># labels all the nodes in the tree</span>
 <span class="fu">theme</span><span class="op">(</span>
   legend.position <span class="op">=</span> <span class="st">"none"</span>,                            <span class="co"># removes the legend all together</span>
   axis.title.x <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,
   axis.title.y <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,
   plot.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span>, face<span class="op">=</span><span class="st">"bold"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_subsetting-1.png" width="50%" style="display: block; margin: auto;"></div>
<p>Now, say we have decided to subset the tree at node 528 (keep only tips within this branch after node 528) and we save it as a new <code>sub_tree1</code> object:</p>
<div class="sourceCode" id="cb603"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sub_tree1</span> <span class="op">&lt;-</span> <span class="fu">tree_subset</span><span class="op">(</span>
  <span class="va">tree</span>,
  node <span class="op">=</span> <span class="fl">528</span><span class="op">)</span>                                            <span class="co"># we subset the tree at node 528</span></code></pre></div>
<p>Lets have a look at the subset tree 1:</p>
<div class="sourceCode" id="cb604"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ggtree</span><span class="op">(</span><span class="va">sub_tree1</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_tiplab</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Subset tree 1"</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/unnamed-chunk-470-1.png" width="672"></div>
<p>You can also subset based on one particular sample, specifying how many nodes “backwards” you want to include. Let’s subset the same part of the tree based on a sample, in this case S17BD07692, going back 9 nodes and we save it as a new <code>sub_tree2</code> object:</p>
<div class="sourceCode" id="cb605"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sub_tree2</span> <span class="op">&lt;-</span> <span class="fu">tree_subset</span><span class="op">(</span>
  <span class="va">tree</span>,
  <span class="st">"S17BD07692"</span>,
  levels_back <span class="op">=</span> <span class="fl">9</span><span class="op">)</span> <span class="co"># levels back defines how many nodes backwards from the sample tip you want to go</span></code></pre></div>
<p>Lets have a look at the subset tree 2:</p>
<div class="sourceCode" id="cb606"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ggtree</span><span class="op">(</span><span class="va">sub_tree2</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_tiplab</span><span class="op">(</span>size <span class="op">=</span><span class="fl">3</span><span class="op">)</span>  <span class="op">+</span>
  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Subset tree 2"</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/unnamed-chunk-472-1.png" width="672"></div>
<p>You can also save your new tree either as a Newick type or even a text file using the <code>write.tree()</code> function from <strong>ape</strong> package:</p>
<div class="sourceCode" id="cb607"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># to save in .nwk format</span>
<span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/write.tree.html">write.tree</a></span><span class="op">(</span><span class="va">sub_tree2</span>, file<span class="op">=</span><span class="st">'data/phylo/Shigella_subtree_2.nwk'</span><span class="op">)</span>

<span class="co"># to save in .txt format</span>
<span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/write.tree.html">write.tree</a></span><span class="op">(</span><span class="va">sub_tree2</span>, file<span class="op">=</span><span class="st">'data/phylo/Shigella_subtree_2.txt'</span><span class="op">)</span></code></pre></div>
</div>
<div id="rotating-nodes-in-a-tree" class="section level3 unnumbered">
<h3>Rotating nodes in a tree<a class="anchor" aria-label="anchor" href="#rotating-nodes-in-a-tree"><i class="fas fa-link"></i></a>
</h3>
<p>As mentioned before we cannot change the order of tips or nodes in the tree, as this is based on their genetic relatedness and is not subject to visual manipulation. But we can rote branches around nodes if that eases our visualization.</p>
<p>First, we plot our new subset tree 2 with node labels to choose the node we want to manipulate and store it an a ggtree plot object <code>p</code>.</p>
<div class="sourceCode" id="cb608"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">ggtree</span><span class="op">(</span><span class="va">sub_tree2</span><span class="op">)</span> <span class="op">+</span>  
  <span class="fu">geom_tiplab</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_text2</span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>subset<span class="op">=</span><span class="op">!</span><span class="va">isTip</span>, label<span class="op">=</span><span class="va">node</span><span class="op">)</span>, <span class="co"># labels all the nodes in the tree</span>
             size <span class="op">=</span> <span class="fl">5</span>,
             color <span class="op">=</span> <span class="st">"darkred"</span>, 
             hjust <span class="op">=</span> <span class="fl">1</span>, 
             vjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> 
<span class="va">p</span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_rotating_1-1.png" width="50%" style="display: block; margin: auto;"></div>
<p>We can then manipulate nodes by applying <strong>ggtree::rotate()</strong> or <strong>ggtree::flip()</strong>:
Note: to illustrate which nodes we are manipulating we first apply the <strong>geom_hilight()</strong> function from <strong>ggtree</strong> to highlight the samples in the nodes we are interested in and store that ggtree plot object in a new object <code>p1</code>.</p>
<div class="sourceCode" id="cb609"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p1</span> <span class="op">&lt;-</span> <span class="va">p</span> <span class="op">+</span> <span class="fu">geom_hilight</span><span class="op">(</span>  <span class="co"># highlights node 39 in blue, "extend =" allows us to define the length of the color block</span>
  node <span class="op">=</span> <span class="fl">39</span>,
  fill <span class="op">=</span> <span class="st">"steelblue"</span>,
  extend <span class="op">=</span> <span class="fl">0.0017</span><span class="op">)</span> <span class="op">+</span>  
<span class="fu">geom_hilight</span><span class="op">(</span>            <span class="co"># highlights the node 37 in yellow</span>
  node <span class="op">=</span> <span class="fl">37</span>,
  fill <span class="op">=</span> <span class="st">"yellow"</span>,
  extend <span class="op">=</span> <span class="fl">0.0017</span><span class="op">)</span> <span class="op">+</span>               
<span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Original tree"</span><span class="op">)</span>


<span class="va">p1</span> <span class="co"># print</span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_rotating_2-1.png" width="50%" style="display: block; margin: auto;"></div>
<p>Now we can rotate node 37 in object <code>p1</code> so that the samples on node 38 move to the top. We store the rotated tree in a new object <code>p2</code>.</p>
<div class="sourceCode" id="cb610"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu">rotate</span><span class="op">(</span><span class="va">p1</span>, <span class="fl">37</span><span class="op">)</span> <span class="op">+</span> 
      <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Rotated Node 37"</span><span class="op">)</span>


<span class="va">p2</span>   <span class="co"># print</span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/unnamed-chunk-473-1.png" width="672"></div>
<p>Or we can use the flip command to rotate node 36 in object <code>p1</code> and switch node 37 to the top and node 39 to the bottom. We store the flipped tree in a new object <code>p3</code>.</p>
<div class="sourceCode" id="cb611"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p3</span> <span class="op">&lt;-</span>  <span class="fu">flip</span><span class="op">(</span><span class="va">p1</span>, <span class="fl">39</span>, <span class="fl">37</span><span class="op">)</span> <span class="op">+</span>
      <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Rotated Node 36"</span><span class="op">)</span>


<span class="va">p3</span>   <span class="co"># print</span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/unnamed-chunk-474-1.png" width="672"></div>
</div>
<div id="example-subtree-with-sample-data-annotation" class="section level3 unnumbered">
<h3>Example subtree with sample data annotation<a class="anchor" aria-label="anchor" href="#example-subtree-with-sample-data-annotation"><i class="fas fa-link"></i></a>
</h3>
<p>Lets say we are investigating the cluster of cases with clonal expansion which occurred in 2017 and 2018 at node 39 in our sub-tree. We add the year of strain isolation as well as travel history and color by country to see origin of other closely related strains:</p>
<div class="sourceCode" id="cb612"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ggtree</span><span class="op">(</span><span class="va">sub_tree2</span><span class="op">)</span> <span class="op">%&lt;+%</span> <span class="va">sample_data</span> <span class="op">+</span>     <span class="co"># we use th %&lt;+% operator to link to the sample_data</span>
  <span class="fu">geom_tiplab</span><span class="op">(</span>                          <span class="co"># labels the tips of all branches with the sample name in the tree file</span>
    size <span class="op">=</span> <span class="fl">2.5</span>,
    offset <span class="op">=</span> <span class="fl">0.001</span>,
    align <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">theme_tree2</span><span class="op">(</span><span class="op">)</span><span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.window.html">xlim</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.015</span><span class="op">)</span><span class="op">+</span>                       <span class="co"># set the x-axis limits of our tree</span>
  <span class="fu">geom_tippoint</span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>color<span class="op">=</span><span class="va">Country</span><span class="op">)</span>,     <span class="co"># color the tip point by continent</span>
                size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span><span class="op">+</span> 
  <span class="fu">scale_color_brewer</span><span class="op">(</span>
    name <span class="op">=</span> <span class="st">"Country"</span>, 
    palette <span class="op">=</span> <span class="st">"Set1"</span>, 
    na.value <span class="op">=</span> <span class="st">"grey"</span><span class="op">)</span><span class="op">+</span>
  <span class="fu">geom_tiplab</span><span class="op">(</span>                          <span class="co"># add isolation year as a text label at the tips</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">Year</span><span class="op">)</span>,
    color <span class="op">=</span> <span class="st">'blue'</span>,
    offset <span class="op">=</span> <span class="fl">0.0045</span>,
    size <span class="op">=</span> <span class="fl">3</span>,
    linetype <span class="op">=</span> <span class="st">"blank"</span> ,
    geom <span class="op">=</span> <span class="st">"text"</span>,
    align <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">+</span> 
  <span class="fu">geom_tiplab</span><span class="op">(</span>                          <span class="co"># add travel history as a text label at the tips, in red color</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">Travel_history</span><span class="op">)</span>,
    color <span class="op">=</span> <span class="st">'red'</span>,
    offset <span class="op">=</span> <span class="fl">0.006</span>,
    size <span class="op">=</span> <span class="fl">3</span>,
    linetype <span class="op">=</span> <span class="st">"blank"</span>,
    geom <span class="op">=</span> <span class="st">"text"</span>,
    align <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">+</span> 
  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Phylogenetic tree of Belgian S. sonnei strains with travel history"</span><span class="op">)</span><span class="op">+</span>  <span class="co"># add plot title</span>
  <span class="fu">xlab</span><span class="op">(</span><span class="st">"genetic distance (0.001 = 4 nucleotides difference)"</span><span class="op">)</span><span class="op">+</span>                    <span class="co"># add a label to the x-axis </span>
  <span class="fu">theme</span><span class="op">(</span>
    axis.title.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>,
    axis.title.y <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,
    legend.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>face <span class="op">=</span> <span class="st">"bold"</span>, size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span>,
    legend.text <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>face <span class="op">=</span> <span class="st">"bold"</span>, size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>,
    plot.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span>, face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_inspect_subset_example-1.png" width="80%" style="display: block; margin: auto;"></div>
<p>Our observation points towards an import event of strains from Asia, which then circulated in Belgium over the years and seem to have caused our latest outbreak.</p>
<!-- ======================================================= -->
</div>
</div>
<div id="more-complex-trees-adding-heatmaps-of-sample-data" class="section level2 unnumbered">
<h2>More complex trees: adding heatmaps of sample data<a class="anchor" aria-label="anchor" href="#more-complex-trees-adding-heatmaps-of-sample-data"><i class="fas fa-link"></i></a>
</h2>
<p>We can add more complex information, such as categorical presence of antimicrobial resistance genes and numeric values for actually measured resistance to antimicrobials in form of a heatmap using the <strong>ggtree::gheatmap()</strong> function.</p>
<p>First we need to plot our tree (this can be either linear or circular) and store it in a new ggtree plot object <code>p</code>: We will use the sub_tree from part 3.)</p>
<div class="sourceCode" id="cb613"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">ggtree</span><span class="op">(</span><span class="va">sub_tree2</span>, branch.length<span class="op">=</span><span class="st">'none'</span>, layout<span class="op">=</span><span class="st">'circular'</span><span class="op">)</span> <span class="op">%&lt;+%</span> <span class="va">sample_data</span> <span class="op">+</span>
  <span class="fu">geom_tiplab</span><span class="op">(</span>size <span class="op">=</span><span class="fl">3</span><span class="op">)</span> <span class="op">+</span> 
 <span class="fu">theme</span><span class="op">(</span>
   legend.position <span class="op">=</span> <span class="st">"none"</span>,
    axis.title.x <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,
    axis.title.y <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,
    plot.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>
      size <span class="op">=</span> <span class="fl">12</span>,
      face <span class="op">=</span> <span class="st">"bold"</span>,
      hjust <span class="op">=</span> <span class="fl">0.5</span>,
      vjust <span class="op">=</span> <span class="op">-</span><span class="fl">15</span><span class="op">)</span><span class="op">)</span>
<span class="va">p</span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_sampledata_heatmap-1.png" width="60%" style="display: block; margin: auto;"></div>
<p>Second, we prepare our data. To visualize different variables with new color schemes, we subset our dataframe to the desired variable. It is important to add the <code>Sample_ID</code> as rownames otherwise it cannot match the data to the tree <code>tip.labels</code>:</p>
<p>In our example we want to look at gender and mutations that could confer resistance to Ciprofloxacin, an important first line antibiotic used to treat Shigella infections.</p>
<p>We create a dataframe for gender:</p>
<div class="sourceCode" id="cb614"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gender</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="st">"gender"</span> <span class="op">=</span> <span class="va">sample_data</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Gender"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">gender</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">sample_data</span><span class="op">$</span><span class="va">Sample_ID</span></code></pre></div>
<p>We create a dataframe for mutations in the gyrA gene, which confer Ciprofloxacin resistance:</p>
<div class="sourceCode" id="cb615"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cipR</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="st">"cipR"</span> <span class="op">=</span> <span class="va">sample_data</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"gyrA_mutations"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">cipR</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">sample_data</span><span class="op">$</span><span class="va">Sample_ID</span></code></pre></div>
<p>We create a dataframe for the measured minimum inhibitory concentration (MIC) for Ciprofloxacin from the laboratory:</p>
<div class="sourceCode" id="cb616"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">MIC_Cip</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="st">"mic_cip"</span> <span class="op">=</span> <span class="va">sample_data</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"MIC_CIP"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">MIC_Cip</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">sample_data</span><span class="op">$</span><span class="va">Sample_ID</span></code></pre></div>
<p>We create a first plot adding a binary heatmap for gender to the phylogenetic tree and storing it in a new ggtree plot object <code>h1</code>:</p>
<div class="sourceCode" id="cb617"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">h1</span> <span class="op">&lt;-</span>  <span class="fu">gheatmap</span><span class="op">(</span><span class="va">p</span>, <span class="va">gender</span>,                                 <span class="co"># we add a heatmap layer of the gender dataframe to our tree plot</span>
                offset <span class="op">=</span> <span class="fl">10</span>,                               <span class="co"># offset shifts the heatmap to the right,</span>
                width <span class="op">=</span> <span class="fl">0.10</span>,                              <span class="co"># width defines the width of the heatmap column,</span>
                color <span class="op">=</span> <span class="cn">NULL</span>,                              <span class="co"># color defines the boarder of the heatmap columns</span>
         colnames <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span>                               <span class="co"># hides column names for the heatmap</span>
  <span class="fu">scale_fill_manual</span><span class="op">(</span>name <span class="op">=</span> <span class="st">"Gender"</span>,                       <span class="co"># define the coloring scheme and legend for gender</span>
                    values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#00d1b1"</span>, <span class="st">"purple"</span><span class="op">)</span>,
                    breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Male"</span>, <span class="st">"Female"</span><span class="op">)</span>,
                    labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Male"</span>, <span class="st">"Female"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
   <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span>,
        legend.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span>,
        legend.text <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>,
        legend.box <span class="op">=</span> <span class="st">"vertical"</span>, legend.margin <span class="op">=</span> <span class="fu">margin</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## Scale for 'y' is already present. Adding another scale for 'y', which will replace the existing scale.</code></pre>
<pre><code>## Scale for 'fill' is already present. Adding another scale for 'fill', which will replace the existing scale.</code></pre>
<div class="sourceCode" id="cb620"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">h1</span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_sampledata_heatmap_gender-1.png" width="70%" style="display: block; margin: auto;"></div>
<p>Then we add information on mutations in the gyrA gene, which confer resistance to Ciprofloxacin:</p>
<p>Note: The presence of chromosomal point mutations in WGS data was prior determined using the PointFinder tool developed by Zankari et al. (see reference in the additional references section)</p>
<p>First, we assign a new color scheme to our existing plot object <code>h1</code> and store it in a now object <code>h2</code>. This enables us to define and change the colors for our second variable in the heatmap.</p>
<div class="sourceCode" id="cb621"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">h2</span> <span class="op">&lt;-</span> <span class="va">h1</span> <span class="op">+</span> <span class="fu">new_scale_fill</span><span class="op">(</span><span class="op">)</span> </code></pre></div>
<p>Then we add the second heatmap layer to <code>h2</code> and store the combined plots in a new object <code>h3</code>:</p>
<div class="sourceCode" id="cb622"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">h3</span> <span class="op">&lt;-</span> <span class="fu">gheatmap</span><span class="op">(</span><span class="va">h2</span>, <span class="va">cipR</span>,         <span class="co"># adds the second row of heatmap describing Ciprofloxacin resistance mutations</span>
               offset <span class="op">=</span> <span class="fl">12</span>, 
               width <span class="op">=</span> <span class="fl">0.10</span>, 
               colnames <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">scale_fill_manual</span><span class="op">(</span>name <span class="op">=</span> <span class="st">"Ciprofloxacin resistance \n conferring mutation"</span>,
                    values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#fe9698"</span>,<span class="st">"#ea0c92"</span><span class="op">)</span>,
                    breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span> <span class="st">"gyrA D87Y"</span>, <span class="st">"gyrA S83L"</span><span class="op">)</span>,
                    labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span> <span class="st">"gyrA d87y"</span>, <span class="st">"gyrA s83l"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
   <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span>,
        legend.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span>,
        legend.text <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>,
        legend.box <span class="op">=</span> <span class="st">"vertical"</span>, legend.margin <span class="op">=</span> <span class="fu">margin</span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">+</span>
  <span class="fu">guides</span><span class="op">(</span>fill <span class="op">=</span> <span class="fu">guide_legend</span><span class="op">(</span>nrow <span class="op">=</span> <span class="fl">2</span>,byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## Scale for 'y' is already present. Adding another scale for 'y', which will replace the existing scale.</code></pre>
<pre><code>## Scale for 'fill' is already present. Adding another scale for 'fill', which will replace the existing scale.</code></pre>
<div class="sourceCode" id="cb625"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">h3</span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_sampledata_heatmap_cip_genes-1.png" width="80%" style="display: block; margin: auto;"></div>
<p>We repeat the above process, by first adding a new color scale layer to our existing object <code>h3</code>, and then adding the continuous data on the minimum inhibitory concentration (MIC) of Ciprofloxacin for each strain to the resulting object <code>h4</code> to produce the final object <code>h5</code>:</p>
<div class="sourceCode" id="cb626"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># First we add the new coloring scheme:</span>
<span class="va">h4</span> <span class="op">&lt;-</span> <span class="va">h3</span> <span class="op">+</span> <span class="fu">new_scale_fill</span><span class="op">(</span><span class="op">)</span>

<span class="co"># then we combine the two into a new plot:</span>
<span class="va">h5</span> <span class="op">&lt;-</span> <span class="fu">gheatmap</span><span class="op">(</span><span class="va">h4</span>, <span class="va">MIC_Cip</span>,  
               offset <span class="op">=</span> <span class="fl">14</span>, 
               width <span class="op">=</span> <span class="fl">0.10</span>,
                colnames <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">+</span>
  <span class="fu">scale_fill_continuous</span><span class="op">(</span>name <span class="op">=</span> <span class="st">"MIC for Ciprofloxacin"</span>,  <span class="co"># here we define a gradient color scheme for the continuous variable of MIC</span>
                      low <span class="op">=</span> <span class="st">"yellow"</span>, high <span class="op">=</span> <span class="st">"red"</span>,
                      breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.50</span>, <span class="fl">1.00</span><span class="op">)</span>,
                      na.value <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span> <span class="op">+</span>
   <span class="fu">guides</span><span class="op">(</span>fill <span class="op">=</span> <span class="fu">guide_colourbar</span><span class="op">(</span>barwidth <span class="op">=</span> <span class="fl">5</span>, barheight <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">+</span>
   <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span>,
        legend.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span>,
        legend.text <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>,
        legend.box <span class="op">=</span> <span class="st">"vertical"</span>, legend.margin <span class="op">=</span> <span class="fu">margin</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## Scale for 'y' is already present. Adding another scale for 'y', which will replace the existing scale.</code></pre>
<pre><code>## Scale for 'fill' is already present. Adding another scale for 'fill', which will replace the existing scale.</code></pre>
<div class="sourceCode" id="cb629"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">h5</span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_sampledata_heatmap_cip_MIC-1.png" width="90%" style="display: block; margin: auto;"></div>
<p>We can do the same exercise for a linear tree:</p>
<div class="sourceCode" id="cb630"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">ggtree</span><span class="op">(</span><span class="va">sub_tree2</span><span class="op">)</span> <span class="op">%&lt;+%</span> <span class="va">sample_data</span> <span class="op">+</span>
  <span class="fu">geom_tiplab</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span> <span class="co"># labels the tips</span>
  <span class="fu">theme_tree2</span><span class="op">(</span><span class="op">)</span><span class="op">+</span>
  <span class="fu">xlab</span><span class="op">(</span><span class="st">"genetic distance (0.001 = 4 nucleotides difference)"</span><span class="op">)</span><span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.window.html">xlim</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.015</span><span class="op">)</span><span class="op">+</span>
 <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span>,
      axis.title.y <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,
      plot.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span>, 
                                face <span class="op">=</span> <span class="st">"bold"</span>,
                                hjust <span class="op">=</span> <span class="fl">0.5</span>,
                                vjust <span class="op">=</span> <span class="op">-</span><span class="fl">15</span><span class="op">)</span><span class="op">)</span>
<span class="va">p</span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_sampledata_heatmap_linear_1-1.png" width="80%" style="display: block; margin: auto;"></div>
<p>First we add gender:</p>
<div class="sourceCode" id="cb631"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">h1</span> <span class="op">&lt;-</span>  <span class="fu">gheatmap</span><span class="op">(</span><span class="va">p</span>, <span class="va">gender</span>, 
                offset <span class="op">=</span> <span class="fl">0.003</span>,
                width <span class="op">=</span> <span class="fl">0.1</span>, 
                color<span class="op">=</span><span class="st">"black"</span>, 
         colnames <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">+</span>
  <span class="fu">scale_fill_manual</span><span class="op">(</span>name <span class="op">=</span> <span class="st">"Gender"</span>,
                    values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#00d1b1"</span>, <span class="st">"purple"</span><span class="op">)</span>,
                    breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Male"</span>, <span class="st">"Female"</span><span class="op">)</span>,
                    labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Male"</span>, <span class="st">"Female"</span><span class="op">)</span><span class="op">)</span><span class="op">+</span>
   <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span>,
        legend.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span>,
        legend.text <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>,
        legend.box <span class="op">=</span> <span class="st">"vertical"</span>, legend.margin <span class="op">=</span> <span class="fu">margin</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## Scale for 'y' is already present. Adding another scale for 'y', which will replace the existing scale.</code></pre>
<pre><code>## Scale for 'fill' is already present. Adding another scale for 'fill', which will replace the existing scale.</code></pre>
<div class="sourceCode" id="cb634"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">h1</span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_sampledata_heatmap_linear_2-1.png" width="80%" style="display: block; margin: auto;"></div>
<p>Then we add Ciprofloxacin resistance mutations after adding another color scheme layer:</p>
<div class="sourceCode" id="cb635"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">h2</span> <span class="op">&lt;-</span> <span class="va">h1</span> <span class="op">+</span> <span class="fu">new_scale_fill</span><span class="op">(</span><span class="op">)</span>
<span class="va">h3</span> <span class="op">&lt;-</span> <span class="fu">gheatmap</span><span class="op">(</span><span class="va">h2</span>, <span class="va">cipR</span>,   
               offset <span class="op">=</span> <span class="fl">0.004</span>, 
               width <span class="op">=</span> <span class="fl">0.1</span>,
               color <span class="op">=</span> <span class="st">"black"</span>,
                colnames <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">+</span>
  <span class="fu">scale_fill_manual</span><span class="op">(</span>name <span class="op">=</span> <span class="st">"Ciprofloxacin resistance \n conferring mutation"</span>,
                    values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#fe9698"</span>,<span class="st">"#ea0c92"</span><span class="op">)</span>,
                    breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span> <span class="st">"gyrA D87Y"</span>, <span class="st">"gyrA S83L"</span><span class="op">)</span>,
                    labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span> <span class="st">"gyrA d87y"</span>, <span class="st">"gyrA s83l"</span><span class="op">)</span><span class="op">)</span><span class="op">+</span>
   <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span>,
        legend.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span>,
        legend.text <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>,
        legend.box <span class="op">=</span> <span class="st">"vertical"</span>, legend.margin <span class="op">=</span> <span class="fu">margin</span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">+</span>
  <span class="fu">guides</span><span class="op">(</span>fill <span class="op">=</span> <span class="fu">guide_legend</span><span class="op">(</span>nrow <span class="op">=</span> <span class="fl">2</span>,byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## Scale for 'y' is already present. Adding another scale for 'y', which will replace the existing scale.</code></pre>
<pre><code>## Scale for 'fill' is already present. Adding another scale for 'fill', which will replace the existing scale.</code></pre>
<div class="sourceCode" id="cb638"><pre class="downlit sourceCode r">
<code class="sourceCode R"> <span class="va">h3</span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_sampledata_heatmap_linear_3-1.png" width="80%" style="display: block; margin: auto;"></div>
<p>Then we add the minimum inhibitory concentration determined by the laboratory (MIC):</p>
<div class="sourceCode" id="cb639"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">h4</span> <span class="op">&lt;-</span> <span class="va">h3</span> <span class="op">+</span> <span class="fu">new_scale_fill</span><span class="op">(</span><span class="op">)</span>
<span class="va">h5</span> <span class="op">&lt;-</span> <span class="fu">gheatmap</span><span class="op">(</span><span class="va">h4</span>, <span class="va">MIC_Cip</span>, 
               offset <span class="op">=</span> <span class="fl">0.005</span>,  
               width <span class="op">=</span> <span class="fl">0.1</span>,
               color <span class="op">=</span> <span class="st">"black"</span>, 
                colnames <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">+</span>
  <span class="fu">scale_fill_continuous</span><span class="op">(</span>name <span class="op">=</span> <span class="st">"MIC for Ciprofloxacin"</span>,
                      low <span class="op">=</span> <span class="st">"yellow"</span>, high <span class="op">=</span> <span class="st">"red"</span>,
                      breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0.50</span>,<span class="fl">1.00</span><span class="op">)</span>,
                      na.value <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span><span class="op">+</span>
   <span class="fu">guides</span><span class="op">(</span>fill <span class="op">=</span> <span class="fu">guide_colourbar</span><span class="op">(</span>barwidth <span class="op">=</span> <span class="fl">5</span>, barheight <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">+</span>
   <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span>,
        legend.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>,
        legend.text <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>,
        legend.box <span class="op">=</span> <span class="st">"horizontal"</span>, legend.margin <span class="op">=</span> <span class="fu">margin</span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">+</span>
  <span class="fu">guides</span><span class="op">(</span>shape <span class="op">=</span> <span class="fu">guide_legend</span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## Scale for 'y' is already present. Adding another scale for 'y', which will replace the existing scale.</code></pre>
<pre><code>## Scale for 'fill' is already present. Adding another scale for 'fill', which will replace the existing scale.</code></pre>
<div class="sourceCode" id="cb642"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">h5</span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_sampledata_heatmap_linear_4-1.png" width="80%" style="display: block; margin: auto;"></div>
<!-- ======================================================= -->
</div>
<div id="resources-14" class="section level2" number="15.5">
<h2>
<span class="header-section-number">15.5</span> Resources<a class="anchor" aria-label="anchor" href="#resources-14"><i class="fas fa-link"></i></a>
</h2>
<p><a href="http://hydrodictyon.eeb.uconn.edu/eebedia/index.php/Ggtree#" class="uri">http://hydrodictyon.eeb.uconn.edu/eebedia/index.php/Ggtree#</a> Clade_Colors
<a href="https://bioconductor.riken.jp/packages/3.2/bioc/vignettes/ggtree/inst/doc/treeManipulation.html" class="uri">https://bioconductor.riken.jp/packages/3.2/bioc/vignettes/ggtree/inst/doc/treeManipulation.html</a>
<a href="https://guangchuangyu.github.io/ggtree-book/chapter-ggtree.html" class="uri">https://guangchuangyu.github.io/ggtree-book/chapter-ggtree.html</a>
<a href="https://bioconductor.riken.jp/packages/3.8/bioc/vignettes/ggtree/inst/doc/treeManipulation.html" class="uri">https://bioconductor.riken.jp/packages/3.8/bioc/vignettes/ggtree/inst/doc/treeManipulation.html</a></p>
<p>Ea Zankari, Rosa Allesøe, Katrine G Joensen, Lina M Cavaco, Ole Lund, Frank M Aarestrup, PointFinder: a novel web tool for WGS-based detection of antimicrobial resistance associated with chromosomal point mutations in bacterial pathogens, Journal of Antimicrobial Chemotherapy, Volume 72, Issue 10, October 2017, Pages 2764–2768, <a href="https://doi.org/10.1093/jac/dkx217" class="uri">https://doi.org/10.1093/jac/dkx217</a></p>

</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="transmission-chains.html"><span class="header-section-number">14</span> Transmission chains</a></div>
<div class="next"><a href="interactive-plots.html"><span class="header-section-number">16</span> Interactive plots</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#phylogenetic-trees"><span class="header-section-number">15</span> Phylogenetic trees</a></li>
<li><a class="nav-link" href="#overview-5"><span class="header-section-number">15.1</span> Overview</a></li>
<li>
<a class="nav-link" href="#preparation-16"><span class="header-section-number">15.2</span> Preparation</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#load-packages-12">Load packages</a></li>
<li><a class="nav-link" href="#import-data-9">Import data</a></li>
<li><a class="nav-link" href="#clean-and-inspect">Clean and inspect</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#simple-tree-visualization"><span class="header-section-number">15.3</span> Simple tree visualization</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#different-tree-layouts">Different tree layouts</a></li>
<li><a class="nav-link" href="#simple-tree-plus-sample-data">Simple tree plus sample data</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#tree-manipulation"><span class="header-section-number">15.4</span> Tree manipulation</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#zoom-in">Zoom in</a></li>
<li><a class="nav-link" href="#collapsing-branches">Collapsing branches</a></li>
<li><a class="nav-link" href="#subsetting-a-tree">Subsetting a tree</a></li>
<li><a class="nav-link" href="#rotating-nodes-in-a-tree">Rotating nodes in a tree</a></li>
<li><a class="nav-link" href="#example-subtree-with-sample-data-annotation">Example subtree with sample data annotation</a></li>
</ul>
</li>
<li><a class="nav-link" href="#more-complex-trees-adding-heatmaps-of-sample-data">More complex trees: adding heatmaps of sample data</a></li>
<li><a class="nav-link" href="#resources-14"><span class="header-section-number">15.5</span> Resources</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
          
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>The Epidemiologist R Handbook</strong>" was written by the handbook team. It was last built on 2021-05-31.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
