<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>19 Univariate and multivariable regression | The Epidemiologist R Handbook</title>
<meta name="author" content="the handbook team">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.2"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/header-attrs-2.9/header-attrs.js"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.2.5.1/tabs.js"></script><script src="libs/bs3compat-0.2.5.1/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="libs/htmlwidgets-1.5.3/htmlwidgets.js"></script><link href="libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="libs/datatables-binding-0.18/datatables.js"></script><link href="libs/dt-core-1.10.20/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="libs/dt-core-1.10.20/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="libs/dt-core-1.10.20/js/jquery.dataTables.min.js"></script><link href="libs/nouislider-7.0.10/jquery.nouislider.min.css" rel="stylesheet">
<script src="libs/nouislider-7.0.10/jquery.nouislider.min.js"></script><link href="libs/selectize-0.12.0/selectize.bootstrap3.css" rel="stylesheet">
<script src="libs/selectize-0.12.0/selectize.min.js"></script><link href="libs/crosstalk-1.1.1/css/crosstalk.css" rel="stylesheet">
<script src="libs/crosstalk-1.1.1/js/crosstalk.min.js"></script><script src="libs/kePrint-0.0.1/kePrint.js"></script><link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet">
<link href="libs/tabwid-1.0.0/tabwid.css" rel="stylesheet">
<link href="libs/leaflet-1.3.1/leaflet.css" rel="stylesheet">
<script src="libs/leaflet-1.3.1/leaflet.js"></script><link href="libs/leafletfix-1.0.0/leafletfix.css" rel="stylesheet">
<script src="libs/proj4-2.6.2/proj4.min.js"></script><script src="libs/Proj4Leaflet-1.0.1/proj4leaflet.js"></script><link href="libs/rstudio_leaflet-1.3.1/rstudio_leaflet.css" rel="stylesheet">
<script src="libs/leaflet-binding-2.0.4.1/leaflet.js"></script><script src="libs/leaflet-providers-1.9.0/leaflet-providers_1.9.0.js"></script><script src="libs/leaflet-providers-plugin-2.0.4.1/leaflet-providers-plugin.js"></script><script src="libs/viz-1.8.2/viz.js"></script><link href="libs/DiagrammeR-styles-0.2/styles.css" rel="stylesheet">
<script src="libs/grViz-binding-1.0.6.1/grViz.js"></script><script src="libs/d3-4.5.0/d3.min.js"></script><script src="libs/sankey-1/sankey.js"></script><script src="libs/sankeyNetwork-binding-0.4/sankeyNetwork.js"></script><script src="libs/plotly-binding-4.9.3/plotly.js"></script><script src="libs/typedarray-0.1/typedarray.min.js"></script><link href="libs/plotly-htmlwidgets-css-1.57.1/plotly-htmlwidgets.css" rel="stylesheet">
<script src="libs/plotly-main-1.57.1/plotly-latest.min.js"></script><link href="libs/vis-4.20.1/vis.css" rel="stylesheet">
<script src="libs/vis-4.20.1/vis.min.js"></script><script src="libs/visNetwork-binding-2.0.9/visNetwork.js"></script><link href="libs/font-awesome-4.7.0/css/font-awesome.min.css" rel="stylesheet">
<script src="libs/plotly-basic-1.57.1/plotly-basic-1.57.1.min.js"></script><script src="libs/plotly-cartesian-1.57.1/plotly-cartesian-1.57.1.min.js"></script><script src="libs/proj4js-2.3.15/proj4.js"></script><link href="libs/highcharts-8.1.2/css/motion.css" rel="stylesheet">
<script src="libs/highcharts-8.1.2/highcharts.js"></script><script src="libs/highcharts-8.1.2/highcharts-3d.js"></script><script src="libs/highcharts-8.1.2/highcharts-more.js"></script><script src="libs/highcharts-8.1.2/modules/stock.js"></script><script src="libs/highcharts-8.1.2/modules/map.js"></script><script src="libs/highcharts-8.1.2/modules/annotations.js"></script><script src="libs/highcharts-8.1.2/modules/data.js"></script><script src="libs/highcharts-8.1.2/modules/drilldown.js"></script><script src="libs/highcharts-8.1.2/modules/item-series.js"></script><script src="libs/highcharts-8.1.2/modules/offline-exporting.js"></script><script src="libs/highcharts-8.1.2/modules/overlapping-datalabels.js"></script><script src="libs/highcharts-8.1.2/modules/exporting.js"></script><script src="libs/highcharts-8.1.2/modules/export-data.js"></script><script src="libs/highcharts-8.1.2/modules/funnel.js"></script><script src="libs/highcharts-8.1.2/modules/heatmap.js"></script><script src="libs/highcharts-8.1.2/modules/treemap.js"></script><script src="libs/highcharts-8.1.2/modules/sankey.js"></script><script src="libs/highcharts-8.1.2/modules/dependency-wheel.js"></script><script src="libs/highcharts-8.1.2/modules/organization.js"></script><script src="libs/highcharts-8.1.2/modules/solid-gauge.js"></script><script src="libs/highcharts-8.1.2/modules/streamgraph.js"></script><script src="libs/highcharts-8.1.2/modules/sunburst.js"></script><script src="libs/highcharts-8.1.2/modules/vector.js"></script><script src="libs/highcharts-8.1.2/modules/wordcloud.js"></script><script src="libs/highcharts-8.1.2/modules/xrange.js"></script><script src="libs/highcharts-8.1.2/modules/tilemap.js"></script><script src="libs/highcharts-8.1.2/modules/venn.js"></script><script src="libs/highcharts-8.1.2/modules/gantt.js"></script><script src="libs/highcharts-8.1.2/modules/timeline.js"></script><script src="libs/highcharts-8.1.2/modules/parallel-coordinates.js"></script><script src="libs/highcharts-8.1.2/modules/bullet.js"></script><script src="libs/highcharts-8.1.2/modules/coloraxis.js"></script><script src="libs/highcharts-8.1.2/modules/dumbbell.js"></script><script src="libs/highcharts-8.1.2/modules/lollipop.js"></script><script src="libs/highcharts-8.1.2/modules/series-label.js"></script><script src="libs/highcharts-8.1.2/plugins/motion.js"></script><script src="libs/highcharts-8.1.2/custom/reset.js"></script><script src="libs/highcharts-8.1.2/modules/boost.js"></script><script src="libs/highchart-binding-0.8.2/highchart.js"></script><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-QXDW878QLX"></script><script>
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());

      gtag('config', 'G-QXDW878QLX');
    </script><script src="https://cdn.jsdelivr.net/autocomplete.js/0/autocomplete.jquery.min.js"></script><script src="https://cdn.jsdelivr.net/npm/mark.js@8.11.1/dist/mark.min.js"></script><!-- CSS --><link rel="stylesheet" href="style_bs4.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">The Epidemiologist R Handbook</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"></a></li>
<li class="book-part">About this book</li>
<li><a class="" href="editorial-and-technical-notes.html"><span class="header-section-number">1</span> Editorial and technical notes</a></li>
<li><a class="" href="download-handbook-and-data.html"><span class="header-section-number">2</span> Download handbook and data</a></li>
<li class="book-part">Basics</li>
<li><a class="" href="r-basics.html"><span class="header-section-number">3</span> R Basics</a></li>
<li><a class="" href="transition-to-r.html"><span class="header-section-number">4</span> Transition to R</a></li>
<li><a class="" href="suggested-packages-1.html"><span class="header-section-number">5</span> Suggested packages</a></li>
<li><a class="" href="r-projects.html"><span class="header-section-number">6</span> R projects</a></li>
<li><a class="" href="import-and-export.html"><span class="header-section-number">7</span> Import and export</a></li>
<li class="book-part">Data Management</li>
<li><a class="" href="cleaning-data-and-core-functions.html"><span class="header-section-number">8</span> Cleaning data and core functions</a></li>
<li><a class="" href="working-with-dates.html"><span class="header-section-number">9</span> Working with dates</a></li>
<li><a class="" href="characters-and-strings.html"><span class="header-section-number">10</span> Characters and strings</a></li>
<li><a class="" href="factors.html"><span class="header-section-number">11</span> Factors</a></li>
<li><a class="" href="pivoting-data.html"><span class="header-section-number">12</span> Pivoting data</a></li>
<li><a class="" href="grouping-data.html"><span class="header-section-number">13</span> Grouping data</a></li>
<li><a class="" href="joining-data.html"><span class="header-section-number">14</span> Joining data</a></li>
<li><a class="" href="de-duplication.html"><span class="header-section-number">15</span> De-duplication</a></li>
<li><a class="" href="iteration-loops-and-lists.html"><span class="header-section-number">16</span> Iteration, loops, and lists</a></li>
<li class="book-part">Analysis</li>
<li><a class="" href="descriptive-tables.html"><span class="header-section-number">17</span> Descriptive tables</a></li>
<li><a class="" href="simple-statistical-tests.html"><span class="header-section-number">18</span> Simple statistical tests</a></li>
<li><a class="active" href="univariate-and-multivariable-regression.html"><span class="header-section-number">19</span> Univariate and multivariable regression</a></li>
<li><a class="" href="missing-data.html"><span class="header-section-number">20</span> Missing data</a></li>
<li><a class="" href="standardised-rates.html"><span class="header-section-number">21</span> Standardised rates</a></li>
<li><a class="" href="moving-averages.html"><span class="header-section-number">22</span> Moving averages</a></li>
<li><a class="" href="time-series-and-outbreak-detection.html"><span class="header-section-number">23</span> Time series and outbreak detection</a></li>
<li><a class="" href="epidemic-modeling.html"><span class="header-section-number">24</span> Epidemic modeling</a></li>
<li><a class="" href="contact-tracing-1.html"><span class="header-section-number">25</span> Contact tracing</a></li>
<li><a class="" href="survey-analysis.html"><span class="header-section-number">26</span> Survey analysis</a></li>
<li><a class="" href="survival-analysis.html"><span class="header-section-number">27</span> Survival analysis</a></li>
<li><a class="" href="gis-basics.html"><span class="header-section-number">28</span> GIS basics</a></li>
<li class="book-part">Data Visualization</li>
<li><a class="" href="tables-for-presentation.html"><span class="header-section-number">29</span> Tables for presentation</a></li>
<li><a class="" href="ggplot-basics.html"><span class="header-section-number">30</span> ggplot basics</a></li>
<li><a class="" href="ggplot-tips.html"><span class="header-section-number">31</span> ggplot tips</a></li>
<li><a class="" href="epidemic-curves.html"><span class="header-section-number">32</span> Epidemic curves</a></li>
<li><a class="" href="demographic-pyramids-and-likert-scales.html"><span class="header-section-number">33</span> Demographic pyramids and Likert-scales</a></li>
<li><a class="" href="heat-plots.html"><span class="header-section-number">34</span> Heat plots</a></li>
<li><a class="" href="diagrams-and-charts.html"><span class="header-section-number">35</span> Diagrams and charts</a></li>
<li><a class="" href="combinations-analysis.html"><span class="header-section-number">36</span> Combinations analysis</a></li>
<li><a class="" href="transmission-chains.html"><span class="header-section-number">37</span> Transmission chains</a></li>
<li><a class="" href="phylogenetic-trees-1.html"><span class="header-section-number">38</span> Phylogenetic trees</a></li>
<li><a class="" href="interactive-plots.html"><span class="header-section-number">39</span> Interactive plots</a></li>
<li class="book-part">Reports and dashboards</li>
<li><a class="" href="reports-with-r-markdown.html"><span class="header-section-number">40</span> Reports with R Markdown</a></li>
<li><a class="" href="organizing-routine-reports.html"><span class="header-section-number">41</span> Organizing routine reports</a></li>
<li><a class="" href="dashboards-with-r-markdown.html"><span class="header-section-number">42</span> Dashboards with R Markdown</a></li>
<li><a class="" href="dashboards-with-shiny.html"><span class="header-section-number">43</span> Dashboards with Shiny</a></li>
<li class="book-part">Miscellaneous</li>
<li><a class="" href="writing-functions-1.html"><span class="header-section-number">44</span> Writing functions</a></li>
<li><a class="" href="directory-interactions.html"><span class="header-section-number">45</span> Directory interactions</a></li>
<li><a class="" href="version-control-and-collaboration-with-git-and-github.html"><span class="header-section-number">46</span> Version control and collaboration with Git and Github</a></li>
<li><a class="" href="common-errors.html"><span class="header-section-number">47</span> Common errors</a></li>
<li><a class="" href="getting-help.html"><span class="header-section-number">48</span> Getting help</a></li>
<li><a class="" href="r-on-network-drives.html"><span class="header-section-number">49</span> R on network drives</a></li>
<li><a class="" href="data-table.html"><span class="header-section-number">50</span> Data Table</a></li>
</ul>

        <div class="book-extra">
          
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="univariate-and-multivariable-regression" class="section level1" number="19">
<h1>
<span class="header-section-number">19</span> Univariate and multivariable regression<a class="anchor" aria-label="anchor" href="#univariate-and-multivariable-regression"><i class="fas fa-link"></i></a>
</h1>
<!-- ======================================================= -->
<p>This page demonstrates the use of <strong>base</strong> R regression functions such as <code><a href="https://rdrr.io/r/stats/glm.html">glm()</a></code> and the <strong>gtsummary</strong> package to
look at associations between variables (e.g. odds ratios, risk ratios and hazard
ratios). It also uses functions like <code>tidy()</code> from the <strong>broom</strong> package to clean-up regression outputs.</p>
<ol style="list-style-type: decimal">
<li>Univariate: two-by-two tables</li>
<li>Stratified: mantel-haenszel estimates<br>
</li>
<li>Multivariable: variable selection, model selection, final table</li>
<li>Forest plots</li>
</ol>
<p>For Cox proportional hazard regression, see the <a href="survival-analysis.html#survival-analysis">Survival analysis</a> page.</p>
<p><span style="color: black;"><strong><em>NOTE:</em></strong> We use the term <em>multivariable</em> to refer to a regression with multiple explanatory variables. In this sense a <em>multivariate</em> model would be a regression with several outcomes - see this <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3518362/">editorial</a> for detail </span></p>
<!-- ======================================================= -->
<div id="preparation-10" class="section level2" number="19.1">
<h2>
<span class="header-section-number">19.1</span> Preparation<a class="anchor" aria-label="anchor" href="#preparation-10"><i class="fas fa-link"></i></a>
</h2>
<div id="load-packages-12" class="section level3 unnumbered">
<h3>Load packages<a class="anchor" aria-label="anchor" href="#load-packages-12"><i class="fas fa-link"></i></a>
</h3>
<p>This code chunk shows the loading of packages required for the analyses. In this handbook we emphasize <code><a href="https://rdrr.io/pkg/pacman/man/p_load.html">p_load()</a></code> from <strong>pacman</strong>, which installs the package if necessary <em>and</em> loads it for use. You can also load installed packages with <code><a href="https://rdrr.io/r/base/library.html">library()</a></code> from <strong>base</strong> R. See the page on <a href="r-basics.html#r-basics">R basics</a> for more information on R packages.</p>
<div class="sourceCode" id="cb958"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">pacman</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pacman/man/p_load.html">p_load</a></span><span class="op">(</span>
  <span class="va">rio</span>,          <span class="co"># File import</span>
  <span class="va">here</span>,         <span class="co"># File locator</span>
  <span class="va">tidyverse</span>,    <span class="co"># data management + ggplot2 graphics, </span>
  <span class="va">stringr</span>,      <span class="co"># manipulate text strings </span>
  <span class="va">purrr</span>,        <span class="co"># loop over objects in a tidy way</span>
  <span class="va">gtsummary</span>,    <span class="co"># summary statistics and tests </span>
  <span class="va">broom</span>,        <span class="co"># tidy up results from regressions</span>
  <span class="va">lmtest</span>,       <span class="co"># likelihood-ratio tests</span>
  <span class="va">parameters</span>,   <span class="co"># alternative to tidy up results from regressions</span>
  <span class="va">see</span>          <span class="co"># alternative to visualise forest plots</span>
  <span class="op">)</span></code></pre></div>
</div>
<div id="import-data-12" class="section level3 unnumbered">
<h3>Import data<a class="anchor" aria-label="anchor" href="#import-data-12"><i class="fas fa-link"></i></a>
</h3>
<p>We import the dataset of cases from a simulated Ebola epidemic. If you want to follow along, <a href="https://github.com/appliedepi/epirhandbook_eng/raw/master/data/case_linelists/linelist_cleaned.rds" class="download-button">click to download the “clean” linelist</a> (as .rds file). Import your data with the <code><a href="https://rdrr.io/pkg/rio/man/import.html">import()</a></code> function from the <strong>rio</strong> package (it accepts many file types like .xlsx, .rds, .csv - see the <a href="import-and-export.html#import-and-export">Import and export</a> page for details).</p>
<div class="sourceCode" id="cb959"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># import the linelist</span>
<span class="va">linelist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rio/man/import.html">import</a></span><span class="op">(</span><span class="st">"linelist_cleaned.rds"</span><span class="op">)</span></code></pre></div>
<p>The first 50 rows of the linelist are displayed below.</p>
<div id="htmlwidget-8ff0494df73ff3a8112a" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-8ff0494df73ff3a8112a">{"x":{"filter":"none","data":[["5fe599","8689b7","11f8ea","b8812a","893f25","be99c8","07e3e8","369449","f393b4","1389ca","2978ac","57a565","fc15ef","2eaa9a","bbfa93","c97dd9","f50e8a","3a7673","7f5a01","ddddee","99e8fa","567136","9371a9","bc2adf","403057","8bd1e8","f327be","42e1a9","90e5fe","959170","8ebf6e","e56412","6d788e","a47529","67be4e","da8ecb","148f18","2cb9a5","f5c142","70a9fe","3ad520","062638","c76676","baacc1","497372","23e499","38cc4a","3789ee","c71dcd","6b70f0"],[4,4,2,3,3,3,4,4,4,4,4,4,6,5,6,9,10,8,7,6,7,6,8,6,10,8,6,12,5,8,7,9,11,5,8,5,6,11,7,9,7,8,9,12,13,9,8,10,8,7],["2014-05-08",null,null,"2014-05-04","2014-05-18","2014-05-03","2014-05-22","2014-05-28",null,null,"2014-05-30","2014-05-28","2014-06-14","2014-06-07","2014-06-09",null,null,null,"2014-06-23","2014-06-18","2014-06-24",null,null,"2014-07-03",null,"2014-07-10","2014-06-14",null,"2014-06-18","2014-06-29","2014-07-02","2014-07-12","2014-07-12","2014-06-13","2014-07-15","2014-06-20",null,null,"2014-07-20",null,"2014-07-12","2014-07-19","2014-07-18","2014-07-18","2014-07-27",null,"2014-07-19","2014-07-26","2014-07-24",null],["2014-05-13","2014-05-13","2014-05-16","2014-05-18","2014-05-21","2014-05-22","2014-05-27","2014-06-02","2014-06-05","2014-06-05","2014-06-06","2014-06-13","2014-06-16","2014-06-17","2014-06-18","2014-06-19","2014-06-22","2014-06-23","2014-06-25","2014-06-26","2014-06-28","2014-07-02","2014-07-08","2014-07-09","2014-07-09","2014-07-10","2014-07-12","2014-07-12","2014-07-13","2014-07-13","2014-07-14","2014-07-15","2014-07-16","2014-07-17","2014-07-17","2014-07-18","2014-07-19","2014-07-22","2014-07-22","2014-07-24","2014-07-24","2014-07-25","2014-07-25","2014-07-27","2014-07-29","2014-07-30",null,"2014-08-01","2014-08-02","2014-08-03"],["2014-05-15","2014-05-14","2014-05-18","2014-05-20","2014-05-22","2014-05-23","2014-05-29","2014-06-03","2014-06-06","2014-06-07","2014-06-08","2014-06-15","2014-06-17","2014-06-17","2014-06-20","2014-06-19","2014-06-23","2014-06-24","2014-06-27","2014-06-28","2014-06-29","2014-07-03","2014-07-09","2014-07-09","2014-07-11","2014-07-11","2014-07-13","2014-07-14","2014-07-14","2014-07-13","2014-07-14","2014-07-17","2014-07-17","2014-07-18","2014-07-19","2014-07-20","2014-07-20","2014-07-22","2014-07-24","2014-07-26","2014-07-24","2014-07-27","2014-07-25","2014-07-27","2014-07-31","2014-08-01","2014-08-03","2014-08-02","2014-08-02","2014-08-04"],[null,"2014-05-18","2014-05-30",null,"2014-05-29","2014-05-24","2014-06-01","2014-06-07","2014-06-18","2014-06-09","2014-06-15",null,"2014-07-09",null,"2014-06-30","2014-07-11","2014-07-01","2014-06-25","2014-07-06","2014-07-02","2014-07-09","2014-07-07","2014-07-20",null,"2014-07-22","2014-07-16","2014-07-14","2014-07-20","2014-07-16","2014-07-19","2014-07-27","2014-07-19",null,"2014-07-26","2014-08-14","2014-08-01","2014-07-23","2014-08-28","2014-07-28","2014-07-19",null,"2014-08-03",null,null,null,"2014-08-06","2014-08-21","2014-09-13","2014-08-04",null],[null,"Recover","Recover",null,"Recover","Recover","Recover","Death","Recover","Death","Death","Death","Recover","Recover",null,"Recover",null,null,"Death","Death","Recover",null,null,null,"Death",null,"Death","Death",null,"Death","Recover","Death","Recover","Death","Recover",null,"Death","Recover","Recover","Death",null,null,"Death","Death","Death","Death","Recover",null,"Death","Death"],["m","f","m","f","m","f","f","f","m","f","m","m","m","f","f","m","f","f","f","f","m","m","f","m","f","m","m","f","m","f","f","f","m","m","f","m","f","f","f","m","f","m","f","m","m","f","m","f","m","m"],[2,3,56,18,3,16,16,0,61,27,12,42,19,7,7,13,35,17,11,11,19,54,14,28,6,3,31,6,67,14,10,21,20,45,1,12,3,15,20,36,7,13,14,3,10,1,0,20,26,14],["years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years"],[2,3,56,18,3,16,16,0,61,27,12,42,19,7,7,13,35,17,11,11,19,54,14,28,6,3,31,6,67,14,10,21,20,45,1,12,3,15,20,36,7,13,14,3,10,1,0,20,26,14],["0-4","0-4","50-69","15-19","0-4","15-19","15-19","0-4","50-69","20-29","10-14","30-49","15-19","5-9","5-9","10-14","30-49","15-19","10-14","10-14","15-19","50-69","10-14","20-29","5-9","0-4","30-49","5-9","50-69","10-14","10-14","20-29","20-29","30-49","0-4","10-14","0-4","15-19","20-29","30-49","5-9","10-14","10-14","0-4","10-14","0-4","0-4","20-29","20-29","10-14"],["0-4","0-4","55-59","15-19","0-4","15-19","15-19","0-4","60-64","25-29","10-14","40-44","15-19","5-9","5-9","10-14","35-39","15-19","10-14","10-14","15-19","50-54","10-14","25-29","5-9","0-4","30-34","5-9","65-69","10-14","10-14","20-24","20-24","45-49","0-4","10-14","0-4","15-19","20-24","35-39","5-9","10-14","10-14","0-4","10-14","0-4","0-4","20-24","25-29","10-14"],["Other","Missing","St. Mark's Maternity Hospital (SMMH)","Port Hospital","Military Hospital","Port Hospital","Missing","Missing","Missing","Missing","Port Hospital","Military Hospital","Missing","Missing","Other","Port Hospital","Port Hospital","Port Hospital","Missing","Other","Port Hospital","Port Hospital","St. Mark's Maternity Hospital (SMMH)","Missing","Other","Missing","St. Mark's Maternity Hospital (SMMH)","Military Hospital","Port Hospital","Central Hospital","Military Hospital","Central Hospital","Missing","Military Hospital","Other","Missing","Missing","Port Hospital","Port Hospital","Port Hospital","Missing","Central Hospital","Military Hospital","Other","Other","Other","Missing","St. Mark's Maternity Hospital (SMMH)","St. Mark's Maternity Hospital (SMMH)","Missing"],[-13.2157351064963,-13.2152339775486,-13.212910703914,-13.2363711169728,-13.2228638912441,-13.222625321098,-13.2331547837254,-13.2320975453153,-13.2225511595637,-13.2572163655863,-13.2206286746001,-13.253989309478,-13.2385127873491,-13.209391925612,-13.2157278814899,-13.2243437095992,-13.2336087079551,-13.21422143145,-13.2339681355349,-13.2535640411465,-13.2250089377786,-13.2160657166043,-13.2680671272333,-13.2266742923612,-13.2160179088168,-13.2482584611565,-13.2156319199566,-13.2142410663192,-13.2614879104088,-13.2452992638476,-13.2630592726116,-13.2343341712241,-13.2199077448676,-13.2227293309912,-13.2343062806506,-13.218781651651,-13.2483677722899,-13.2097478342339,-13.2680867723786,-13.2587535457526,-13.262635786914,-13.2697246824573,-13.2209026809759,-13.2330734719715,-13.2680923666905,-13.2547212675054,-13.2573683214693,-13.2137356012883,-13.2175973322257,-13.2486407324245],[8.46897295100924,8.45171855856465,8.46481700596819,8.4754761613651,8.46082377490923,8.46183062600728,8.46272931462646,8.46144367534271,8.46191259217774,8.47292327643506,8.48401630165138,8.45837125340844,8.47761705512509,8.47570184950483,8.47779946878972,8.47145134147474,8.47804840685363,8.48528034195779,8.46957530395867,8.45957352078114,8.47404889511544,8.48802917129884,8.473437335922,8.48408263734462,8.46242233645879,8.47026822126572,8.46398447480533,8.4641347894342,8.45623094629607,8.48334624336805,8.47493999153642,8.47832062438022,8.4693933891765,8.48480589906514,8.47121232619015,8.48438437371817,8.48466158574339,8.47714159984428,8.46238127010609,8.45568597813113,8.4632880274758,8.47940722413856,8.46353857052336,8.46178968158864,8.47508713872833,8.45825808128071,8.4532568143863,8.4732571907655,8.47911586641933,8.48480340615605],["f547d6",null,null,"f90f5f","11f8ea","aec8ec","893f25","133ee7",null,null,"996f3a","133ee7","37a6f6","9f6884","4802b1",null,null,null,"a75c7f","8e104d","ab634e",null,null,"b799eb",null,"5d9e4d","a15e13",null,"ea3740","beb26e","567136","894024","36e2e7","a2086d","7baf73","eb2277",null,null,"d6584f",null,"312ecf","52ea64","cfd79c","d145b7","174288",null,"53608c","3b096b","f5c142",null],["other",null,null,"other","other","other","other","other",null,null,"other","other","other","other","other",null,null,null,"other","other","other",null,null,"other",null,"other","other",null,"other","funeral","other","funeral","other","other","other","funeral",null,null,"other",null,"other","other","other","other","other",null,"funeral","other","other",null],[27,25,91,41,36,56,47,0,86,69,67,84,68,44,34,66,78,47,53,47,71,86,53,69,38,46,68,37,100,56,50,57,65,72,29,69,37,48,54,71,47,61,47,35,53,16,13,59,69,67],[48,59,238,135,71,116,87,11,226,174,112,186,174,90,91,152,214,137,117,131,150,241,131,161,80,69,188,66,233,142,110,182,164,214,26,157,39,154,133,168,100,125,123,67,134,31,36,125,183,169],[22,22,21,23,23,21,21,22,22,22,22,22,22,21,23,22,23,21,22,23,21,23,21,24,23,22,24,23,20,24,24,20,24,21,22,21,23,22,23,23,23,22,23,22,22,22,23,22,22,22],["no",null,null,"no","no","no",null,"no","no","no","no","no","no","no","no","no","no","no",null,"no","no","no","no","no",null,"no","no","no",null,null,"no","no",null,"no","no",null,null,"no","no",null,"no","no",null,"no","no","no","no","no","no",null],["no",null,null,"no","no","no",null,"no","no","no","no","no","no","no","no","no","yes","no",null,"no","no","no","yes","no",null,"no","no","yes",null,null,"no","no",null,"no","no",null,null,"no","no",null,"no","no",null,"no","yes","no","no","no","no",null],["yes",null,null,"no","yes","yes",null,"yes","yes","yes","yes","yes","yes","yes","yes","yes","yes","yes",null,"yes","yes","yes","yes","yes",null,"yes","yes","yes",null,null,"yes","yes",null,"yes","yes",null,null,"yes","yes",null,"yes","yes",null,"yes","yes","yes","yes","yes","no",null],["no",null,null,"no","no","no",null,"no","no","no","no","no","no","no","no","yes","no","no",null,"no","no","no","no","no",null,"no","no","no",null,null,"no","no",null,"no","no",null,null,"yes","yes",null,"no","no",null,"no","no","no","no","no","no",null],["yes",null,null,"no","yes","yes",null,"yes","yes","no","yes","no","no","no","yes","no","no","no",null,"no","yes","no","no","no",null,"no","no","no",null,null,"no","yes",null,"yes","yes",null,null,"yes","yes",null,"yes","yes",null,"yes","yes","no","yes","yes","yes",null],[36.8,36.9,36.9,36.8,36.9,37.6,37.3,37,36.4,35.9,36.5,36.9,36.5,37.1,36.5,37.3,37,38,38,36,37,36.7,36.9,36.5,37,36.5,37.6,36.6,36.6,36.2,36.4,37.1,37.5,37.5,37.4,36.9,36.4,37.3,37,37.8,36.5,37.5,36.7,37,37.3,36.6,36.5,36.6,37.6,36.8],[null,"09:36","16:48","11:22","12:60","14:13","14:33","09:25","11:16","10:55","16:03","11:14","12:42","11:06","09:10","08:45",null,"15:41","13:34","18:58","12:43","16:33","14:29","07:18","08:11","16:32","16:17","07:32","17:45",null,"13:24","14:43","02:33","11:36","17:28","16:27",null,"20:49",null,"11:38","14:25","13:42","21:22","13:33","19:06","17:14","20:09",null,"10:23","09:09"],[117.1875,71.8184429761563,16.0652496292635,22.4965706447188,71.4144019043841,41.6171224732461,62.0953890870657,0,16.8376536925366,22.7903289734443,53.4119897959184,24.2802636142907,22.4600343506408,54.320987654321,41.0578432556455,28.5664819944598,17.0320552013276,25.0412914912888,38.7172182043977,27.3876813705495,31.5555555555556,14.8069075945662,30.8839811199813,26.6193433895297,59.375,96.6183574879227,19.2394748755093,84.9403122130395,18.4199377406104,27.7722674072605,41.3223140495868,17.2080666586161,24.1671624033314,15.7218971089178,428.994082840237,27.9930220292912,243.261012491782,20.2395007589813,30.527446435638,25.15589569161,47,39.04,31.0661643201798,77.9683671196257,29.5165961238583,166.493236212279,100.308641975309,37.76,20.6037803457852,23.458562375267],[2,1,2,2,1,1,2,1,1,2,2,2,1,0,2,0,1,1,2,2,1,1,1,0,2,1,1,2,1,0,0,2,1,1,2,2,1,0,2,2,0,2,0,0,2,2,null,1,0,1]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th>case_id<\/th>\n      <th>generation<\/th>\n      <th>date_infection<\/th>\n      <th>date_onset<\/th>\n      <th>date_hospitalisation<\/th>\n      <th>date_outcome<\/th>\n      <th>outcome<\/th>\n      <th>gender<\/th>\n      <th>age<\/th>\n      <th>age_unit<\/th>\n      <th>age_years<\/th>\n      <th>age_cat<\/th>\n      <th>age_cat5<\/th>\n      <th>hospital<\/th>\n      <th>lon<\/th>\n      <th>lat<\/th>\n      <th>infector<\/th>\n      <th>source<\/th>\n      <th>wt_kg<\/th>\n      <th>ht_cm<\/th>\n      <th>ct_blood<\/th>\n      <th>fever<\/th>\n      <th>chills<\/th>\n      <th>cough<\/th>\n      <th>aches<\/th>\n      <th>vomit<\/th>\n      <th>temp<\/th>\n      <th>time_admission<\/th>\n      <th>bmi<\/th>\n      <th>days_onset_hosp<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"pageLength":5,"scrollX":true,"columnDefs":[{"className":"dt-right","targets":[1,8,10,14,15,18,19,20,26,28,29]}],"order":[],"autoWidth":false,"orderClasses":false,"lengthMenu":[5,10,25,50,100]}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="clean-data" class="section level3 unnumbered">
<h3>Clean data<a class="anchor" aria-label="anchor" href="#clean-data"><i class="fas fa-link"></i></a>
</h3>
<div id="store-explanatory-variables" class="section level4 unnumbered">
<h4>Store explanatory variables<a class="anchor" aria-label="anchor" href="#store-explanatory-variables"><i class="fas fa-link"></i></a>
</h4>
<p>We store the names of the explanatory columns as a character vector. This will be referenced later.</p>
<div class="sourceCode" id="cb960"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## define variables of interest </span>
<span class="va">explanatory_vars</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"gender"</span>, <span class="st">"fever"</span>, <span class="st">"chills"</span>, <span class="st">"cough"</span>, <span class="st">"aches"</span>, <span class="st">"vomit"</span><span class="op">)</span></code></pre></div>
</div>
<div id="convert-to-1s-and-0s" class="section level4 unnumbered">
<h4>Convert to 1’s and 0’s<a class="anchor" aria-label="anchor" href="#convert-to-1s-and-0s"><i class="fas fa-link"></i></a>
</h4>
<p>Below we convert the explanatory columns from “yes”/“no”, “m”/“f”, and “dead”/“alive” to <strong>1 / 0</strong>, to cooperate with the expectations of logistic regression models. To do this efficiently, used <code>across()</code> from <strong>dplyr</strong> to transform multiple columns at one time. The function we apply to each column is <code>case_when()</code> (also <strong>dplyr</strong>) which applies logic to convert specified values to 1’s and 0’s. See sections on <code>across()</code> and <code>case_when()</code> in the <a href="cleaning-data-and-core-functions.html#clean_across">Cleaning data and core functions page</a>).</p>
<p>Note: the “.” below represents the column that is being processed by <code>across()</code> at that moment.</p>
<div class="sourceCode" id="cb961"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## convert dichotomous variables to 0/1 </span>
<span class="va">linelist</span> <span class="op">&lt;-</span> <span class="va">linelist</span> <span class="op">%&gt;%</span>  
  <span class="fu">mutate</span><span class="op">(</span><span class="fu">across</span><span class="op">(</span>                                      
    .cols <span class="op">=</span> <span class="fu">all_of</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">explanatory_vars</span>, <span class="st">"outcome"</span><span class="op">)</span><span class="op">)</span>,  <span class="co">## for each column listed and "outcome"</span>
    .fns <span class="op">=</span> <span class="op">~</span><span class="fu">case_when</span><span class="op">(</span>                              
      <span class="va">.</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"m"</span>, <span class="st">"yes"</span>, <span class="st">"Death"</span><span class="op">)</span>   <span class="op">~</span> <span class="fl">1</span>,           <span class="co">## recode male, yes and death to 1</span>
      <span class="va">.</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"f"</span>, <span class="st">"no"</span>,  <span class="st">"Recover"</span><span class="op">)</span> <span class="op">~</span> <span class="fl">0</span>,           <span class="co">## female, no and recover to 0</span>
      <span class="cn">TRUE</span>                            <span class="op">~</span> <span class="cn">NA_real_</span><span class="op">)</span>    <span class="co">## otherwise set to missing</span>
    <span class="op">)</span>
  <span class="op">)</span></code></pre></div>
</div>
<div id="drop-rows-with-missing-values" class="section level4 unnumbered">
<h4>Drop rows with missing values<a class="anchor" aria-label="anchor" href="#drop-rows-with-missing-values"><i class="fas fa-link"></i></a>
</h4>
<p>To drop rows with missing values, can use the <strong>tidyr</strong> function <code>drop_na()</code>. However, we only want to do this for rows that are missing values in the columns of interest.</p>
<p>The first thing we must to is make sure our <code>explanatory_vars</code> vector includes the column <code>age</code> (<code>age</code> would have produced an error in the previous <code>case_when()</code> operation, which was only for dichotomous variables). Then we pipe the <code>linelist</code> to <code>drop_na()</code> to remove any rows with missing values in the <code>outcome</code> column or any of the <code>explanatory_vars</code> columns.</p>
<p>Before running the code, the number of rows in the <code>linelist</code> is <code><a href="https://rdrr.io/r/base/nrow.html">nrow(linelist)</a></code>.</p>
<div class="sourceCode" id="cb962"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## add in age_category to the explanatory vars </span>
<span class="va">explanatory_vars</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">explanatory_vars</span>, <span class="st">"age_cat"</span><span class="op">)</span>

<span class="co">## drop rows with missing information for variables of interest </span>
<span class="va">linelist</span> <span class="op">&lt;-</span> <span class="va">linelist</span> <span class="op">%&gt;%</span> 
  <span class="fu">drop_na</span><span class="op">(</span><span class="fu">any_of</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"outcome"</span>, <span class="va">explanatory_vars</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>The number of rows remaining in <code>linelist</code> is <code><a href="https://rdrr.io/r/base/nrow.html">nrow(linelist)</a></code>.</p>
<!-- ======================================================= -->
</div>
</div>
</div>
<div id="univariate" class="section level2" number="19.2">
<h2>
<span class="header-section-number">19.2</span> Univariate<a class="anchor" aria-label="anchor" href="#univariate"><i class="fas fa-link"></i></a>
</h2>
<p>Just like in the page on <a href="https://epirhandbook.com/descriptive-tables.html">Descriptive tables</a>, your use case will determine which R package you use. We present two options for doing univariate analysis:</p>
<ul>
<li>Use functions available in <strong>base</strong> R to quickly print results to the console. Use the <strong>broom</strong> package to tidy up the outputs.<br>
</li>
<li>Use the <strong>gtsummary</strong> package to model and get publication-ready outputs</li>
</ul>
<!-- ======================================================= --><div id="base-r-3" class="section level3 unnumbered">
<h3>
<strong>base</strong> R<a class="anchor" aria-label="anchor" href="#base-r-3"><i class="fas fa-link"></i></a>
</h3>
<div id="linear-regression" class="section level4 unnumbered">
<h4>Linear regression<a class="anchor" aria-label="anchor" href="#linear-regression"><i class="fas fa-link"></i></a>
</h4>
<p>The <strong>base</strong> R function <code><a href="https://rdrr.io/r/stats/lm.html">lm()</a></code> perform linear regression, assessing the relationship between numeric response and explanatory variables that are assumed to have a linear relationship.</p>
<p>Provide the equation as a formula, with the response and explanatory column names separated by a tilde <code><a href="https://rdrr.io/r/base/tilde.html">~</a></code>. Also, specify the dataset to <code>data =</code>. Define the model results as an R object, to use later.</p>
<div class="sourceCode" id="cb963"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lm_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span><span class="op">(</span><span class="va">ht_cm</span> <span class="op">~</span> <span class="va">age</span>, data <span class="op">=</span> <span class="va">linelist</span><span class="op">)</span></code></pre></div>
<p>You can then run <code><a href="https://rdrr.io/r/base/summary.html">summary()</a></code> on the model results to see the coefficients (Estimates), P-value, residuals, and other measures.</p>
<div class="sourceCode" id="cb964"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">lm_results</span><span class="op">)</span></code></pre></div>
<pre><code>## 
## Call:
## lm(formula = ht_cm ~ age, data = linelist)
## 
## Residuals:
##      Min       1Q   Median       3Q      Max 
## -128.579  -15.854    1.177   15.887  175.483 
## 
## Coefficients:
##             Estimate Std. Error t value Pr(&gt;|t|)    
## (Intercept)  69.9051     0.5979   116.9   &lt;2e-16 ***
## age           3.4354     0.0293   117.2   &lt;2e-16 ***
## ---
## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
## 
## Residual standard error: 23.75 on 4165 degrees of freedom
## Multiple R-squared:  0.7675, Adjusted R-squared:  0.7674 
## F-statistic: 1.375e+04 on 1 and 4165 DF,  p-value: &lt; 2.2e-16</code></pre>
<p>Alternatively you can use the <code>tidy()</code> function from the <strong>broom</strong> package to pull
the results in to a table. What the results tell us is that for each year increase in age the height increases
by 3.5 cm and this is statistically significant.</p>
<div class="sourceCode" id="cb966"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">tidy</span><span class="op">(</span><span class="va">lm_results</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 2 x 5
##   term        estimate std.error statistic p.value
##   &lt;chr&gt;          &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;   &lt;dbl&gt;
## 1 (Intercept)    69.9     0.598       117.       0
## 2 age             3.44    0.0293      117.       0</code></pre>
<p>You can then also use this regression to add it to a <strong>ggplot</strong>, to do this we
first pull the points for the observed data and the fitted line in to one data frame
using the <code>augment()</code> function from <strong>broom</strong>.</p>
<div class="sourceCode" id="cb968"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## pull the regression points and observed data in to one dataset</span>
<span class="va">points</span> <span class="op">&lt;-</span> <span class="fu">augment</span><span class="op">(</span><span class="va">lm_results</span><span class="op">)</span>

<span class="co">## plot the data using age as the x-axis </span>
<span class="fu">ggplot</span><span class="op">(</span><span class="va">points</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">age</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="co">## add points for height </span>
  <span class="fu">geom_point</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>y <span class="op">=</span> <span class="va">ht_cm</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="co">## add your regression line </span>
  <span class="fu">geom_line</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>y <span class="op">=</span> <span class="va">.fitted</span><span class="op">)</span>, colour <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/lin_reg_res_plot-1.png" width="672"></div>
<p>It is also possible to add a simple linear regression straight straight in <strong>ggplot</strong>
using the <code>geom_smooth()</code> function.</p>
<div class="sourceCode" id="cb969"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## add your data to a plot </span>
 <span class="fu">ggplot</span><span class="op">(</span><span class="va">linelist</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">age</span>, y <span class="op">=</span> <span class="va">ht_cm</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="co">## show points</span>
  <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="co">## add a linear regression </span>
  <span class="fu">geom_smooth</span><span class="op">(</span>method <span class="op">=</span> <span class="st">"lm"</span>, se <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<pre><code>## `geom_smooth()` using formula 'y ~ x'</code></pre>
<div class="inline-figure"><img src="_main_files/figure-html/geom_smooth-1.png" width="672"></div>
<p>See the Resource section at the end of this chapter for more detailed tutorials.</p>
</div>
<div id="logistic-regression" class="section level4 unnumbered">
<h4>Logistic regression<a class="anchor" aria-label="anchor" href="#logistic-regression"><i class="fas fa-link"></i></a>
</h4>
<p>The function <code><a href="https://rdrr.io/r/stats/glm.html">glm()</a></code> from the <strong>stats</strong> package (part of <strong>base</strong> R) is used to fit Generalized Linear Models (GLM).</p>
<p><code><a href="https://rdrr.io/r/stats/glm.html">glm()</a></code> can be used for univariate and multivariable logistic regression (e.g. to get Odds Ratios). Here are the core parts:</p>
<div class="sourceCode" id="cb971"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># arguments for glm()</span>
<span class="fu"><a href="https://rdrr.io/r/stats/glm.html">glm</a></span><span class="op">(</span><span class="va">formula</span>, <span class="va">family</span>, <span class="va">data</span>, <span class="va">weights</span>, <span class="va">subset</span>, <span class="va">...</span><span class="op">)</span></code></pre></div>
<ul>
<li>
<code>formula =</code> The model is provided to <code><a href="https://rdrr.io/r/stats/glm.html">glm()</a></code> as an equation, with the outcome on the left and explanatory variables on the right of a tilde <code><a href="https://rdrr.io/r/base/tilde.html">~</a></code>.<br>
</li>
<li>
<code>family =</code> This determines the type of model to run. For logistic regression, use <code>family = "binomial"</code>, for poisson use <code>family = "poisson"</code>. Other examples are in the table below.<br>
</li>
<li>
<code>data =</code> Specify your data frame</li>
</ul>
<p>If necessary, you can also specify the link function via the syntax <code>family = familytype(link = "linkfunction"))</code>. You can read more in the documentation about other families and optional arguments such as <code>weights =</code> and <code>subset =</code> (<code><a href="https://rdrr.io/r/stats/glm.html">?glm</a></code>).</p>
<div class="inline-table"><table class="table table-sm">
<thead><tr class="header">
<th>Family</th>
<th>Default link function</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>"binomial"</code></td>
<td><code>(link = "logit")</code></td>
</tr>
<tr class="even">
<td><code>"gaussian"</code></td>
<td><code>(link = "identity")</code></td>
</tr>
<tr class="odd">
<td><code>"Gamma"</code></td>
<td><code>(link = "inverse")</code></td>
</tr>
<tr class="even">
<td><code>"inverse.gaussian"</code></td>
<td><code>(link = "1/mu^2")</code></td>
</tr>
<tr class="odd">
<td><code>"poisson"</code></td>
<td><code>(link = "log")</code></td>
</tr>
<tr class="even">
<td><code>"quasi"</code></td>
<td><code>(link = "identity", variance = "constant")</code></td>
</tr>
<tr class="odd">
<td><code>"quasibinomial"</code></td>
<td><code>(link = "logit")</code></td>
</tr>
<tr class="even">
<td><code>"quasipoisson"</code></td>
<td><code>(link = "log")</code></td>
</tr>
</tbody>
</table></div>
<p>When running <code><a href="https://rdrr.io/r/stats/glm.html">glm()</a></code> it is most common to save the results as a named R object. Then you can print the results to your console using <code><a href="https://rdrr.io/r/base/summary.html">summary()</a></code> as shown below, or perform other operations on the results (e.g. exponentiate).</p>
<p>If you need to run a negative binomial regression you can use the <strong>MASS</strong> package; the <code>glm.nb()</code> uses the same syntax as <code><a href="https://rdrr.io/r/stats/glm.html">glm()</a></code>.
For a walk-through of different regressions, see the <a href="https://stats.idre.ucla.edu/other/dae/">UCLA stats page</a>.</p>
</div>
<div id="univariate-glm" class="section level4 unnumbered">
<h4>Univariate <code>glm()</code><a class="anchor" aria-label="anchor" href="#univariate-glm"><i class="fas fa-link"></i></a>
</h4>
<p>In this example we are assessing the association between different age categories and the outcome of death (coded as 1 in the Preparation section). Below is a univariate model of <code>outcome</code> by <code>age_cat</code>. We save the model output as <code>model</code> and then print it with <code><a href="https://rdrr.io/r/base/summary.html">summary()</a></code> to the console. Note the estimates provided are the <em>log odds</em> and that the baseline level is the first factor level of <code>age_cat</code> (“0-4”).</p>
<div class="sourceCode" id="cb972"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/glm.html">glm</a></span><span class="op">(</span><span class="va">outcome</span> <span class="op">~</span> <span class="va">age_cat</span>, family <span class="op">=</span> <span class="st">"binomial"</span>, data <span class="op">=</span> <span class="va">linelist</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">model</span><span class="op">)</span></code></pre></div>
<pre><code>## 
## Call:
## glm(formula = outcome ~ age_cat, family = "binomial", data = linelist)
## 
## Deviance Residuals: 
##    Min      1Q  Median      3Q     Max  
## -1.339  -1.278   1.024   1.080   1.354  
## 
## Coefficients:
##               Estimate Std. Error z value Pr(&gt;|z|)   
## (Intercept)   0.233738   0.072805   3.210  0.00133 **
## age_cat5-9   -0.062898   0.101733  -0.618  0.53640   
## age_cat10-14  0.138204   0.107186   1.289  0.19726   
## age_cat15-19 -0.005565   0.113343  -0.049  0.96084   
## age_cat20-29  0.027511   0.102133   0.269  0.78765   
## age_cat30-49  0.063764   0.113771   0.560  0.57517   
## age_cat50-69 -0.387889   0.259240  -1.496  0.13459   
## age_cat70+   -0.639203   0.915770  -0.698  0.48518   
## ---
## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
## 
## (Dispersion parameter for binomial family taken to be 1)
## 
##     Null deviance: 5712.4  on 4166  degrees of freedom
## Residual deviance: 5705.1  on 4159  degrees of freedom
## AIC: 5721.1
## 
## Number of Fisher Scoring iterations: 4</code></pre>
<p>To alter the baseline level of a given variable, ensure the column is class Factor and move the desired level to the first position with <code>fct_relevel()</code> (see page on <a href="factors.html#factors">Factors</a>). For example, below we take column <code>age_cat</code> and set “20-29” as the baseline before piping the modified data frame into <code><a href="https://rdrr.io/r/stats/glm.html">glm()</a></code>.</p>
<div class="sourceCode" id="cb974"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">linelist</span> <span class="op">%&gt;%</span> 
  <span class="fu">mutate</span><span class="op">(</span>age_cat <span class="op">=</span> <span class="fu">fct_relevel</span><span class="op">(</span><span class="va">age_cat</span>, <span class="st">"20-29"</span>, after <span class="op">=</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/stats/glm.html">glm</a></span><span class="op">(</span>formula <span class="op">=</span> <span class="va">outcome</span> <span class="op">~</span> <span class="va">age_cat</span>, family <span class="op">=</span> <span class="st">"binomial"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## 
## Call:
## glm(formula = outcome ~ age_cat, family = "binomial", data = .)
## 
## Deviance Residuals: 
##    Min      1Q  Median      3Q     Max  
## -1.339  -1.278   1.024   1.080   1.354  
## 
## Coefficients:
##              Estimate Std. Error z value Pr(&gt;|z|)    
## (Intercept)   0.26125    0.07163   3.647 0.000265 ***
## age_cat0-4   -0.02751    0.10213  -0.269 0.787652    
## age_cat5-9   -0.09041    0.10090  -0.896 0.370220    
## age_cat10-14  0.11069    0.10639   1.040 0.298133    
## age_cat15-19 -0.03308    0.11259  -0.294 0.768934    
## age_cat30-49  0.03625    0.11302   0.321 0.748390    
## age_cat50-69 -0.41540    0.25891  -1.604 0.108625    
## age_cat70+   -0.66671    0.91568  -0.728 0.466546    
## ---
## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
## 
## (Dispersion parameter for binomial family taken to be 1)
## 
##     Null deviance: 5712.4  on 4166  degrees of freedom
## Residual deviance: 5705.1  on 4159  degrees of freedom
## AIC: 5721.1
## 
## Number of Fisher Scoring iterations: 4</code></pre>
</div>
<div id="printing-results" class="section level4 unnumbered">
<h4>Printing results<a class="anchor" aria-label="anchor" href="#printing-results"><i class="fas fa-link"></i></a>
</h4>
<p>For most uses, several modifications must be made to the above outputs. The function <code>tidy()</code> from the package <strong>broom</strong> is convenient for making the model results presentable.</p>
<p>Here we demonstrate how to combine model outputs with a table of counts.</p>
<ol style="list-style-type: decimal">
<li>Get the <em>exponentiated</em> log odds ratio estimates and confidence intervals by passing the model to <code>tidy()</code> and setting <code>exponentiate = TRUE</code> and <code>conf.int = TRUE</code>.</li>
</ol>
<div class="sourceCode" id="cb976"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/glm.html">glm</a></span><span class="op">(</span><span class="va">outcome</span> <span class="op">~</span> <span class="va">age_cat</span>, family <span class="op">=</span> <span class="st">"binomial"</span>, data <span class="op">=</span> <span class="va">linelist</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">tidy</span><span class="op">(</span>exponentiate <span class="op">=</span> <span class="cn">TRUE</span>, conf.int <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">%&gt;%</span>        <span class="co"># exponentiate and produce CIs</span>
  <span class="fu">mutate</span><span class="op">(</span><span class="fu">across</span><span class="op">(</span><span class="fu">where</span><span class="op">(</span><span class="va">is.numeric</span><span class="op">)</span>, <span class="va">round</span>, digits <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span>  <span class="co"># round all numeric columns</span></code></pre></div>
<p>Below is the outputted tibble <code>model</code>:</p>
<div id="htmlwidget-eb3a8196605e19421f2f" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-eb3a8196605e19421f2f">{"x":{"filter":"none","data":[["(Intercept)","age_cat5-9","age_cat10-14","age_cat15-19","age_cat20-29","age_cat30-49","age_cat50-69","age_cat70+"],[1.26,0.94,1.15,0.99,1.03,1.07,0.68,0.53],[0.07,0.1,0.11,0.11,0.1,0.11,0.26,0.92],[3.21,-0.62,1.29,-0.05,0.27,0.56,-1.5,-0.7],[0,0.54,0.2,0.96,0.79,0.58,0.13,0.49],[1.1,0.77,0.93,0.8,0.84,0.85,0.41,0.07],[1.46,1.15,1.42,1.24,1.26,1.33,1.13,3.2]],"container":"<table class=\"white-space: nowrap\">\n  <thead>\n    <tr>\n      <th>term<\/th>\n      <th>estimate<\/th>\n      <th>std.error<\/th>\n      <th>statistic<\/th>\n      <th>p.value<\/th>\n      <th>conf.low<\/th>\n      <th>conf.high<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"pageLength":8,"scrollX":true,"columnDefs":[{"className":"dt-right","targets":[1,2,3,4,5,6]}],"order":[],"autoWidth":false,"orderClasses":false,"lengthMenu":[8,10,25,50,100]}},"evals":[],"jsHooks":[]}</script><ol start="2" style="list-style-type: decimal">
<li>Combine these model results with a table of counts. Below, we create the a counts cross-table with the <code>tabyl()</code> function from <strong>janitor</strong>, as covered in the <a href="descriptive-tables.html#descriptive-tables">Descriptive tables</a> page.</li>
</ol>
<div class="sourceCode" id="cb977"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts_table</span> <span class="op">&lt;-</span> <span class="va">linelist</span> <span class="op">%&gt;%</span> 
  <span class="fu">janitor</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/janitor/man/tabyl.html">tabyl</a></span><span class="op">(</span><span class="va">age_cat</span>, <span class="va">outcome</span><span class="op">)</span></code></pre></div>
<!-- * Group rows by outcome, and get counts by age category   -->
<!-- * Pivot wider so the column are `age_cat`, `0`, and `1`   -->
<!-- * Remove row for `NA` `age_cat`, if applicable, to align with the model results   -->
<!-- ```{r} -->
<!-- counts_table <- linelist %>%  -->
<!--   filter(!is.na(outcome) & !is.na(age_cat)) %>%    # ensure outcome and age_cat are present  -->
<!--   group_by(outcome) %>%                            # get counts of variable of interest grouped by outcome -->
<!--   count(age_cat) %>%   ## gets number or rows by unique outcome-age category combinations   -->
<!--   pivot_wider(names_from = outcome, values_from = n)    ## spread data to wide format (as in cross-tabulation) -->
<!-- ``` -->
<p>Here is what this <code>counts_table</code> data frame looks like:</p>
<div id="htmlwidget-2d0de26b5c1ae2a9e9b4" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-2d0de26b5c1ae2a9e9b4">{"x":{"filter":"none","data":[["0-4","5-9","10-14","15-19","20-29","30-49","50-69","70+"],[338,365,273,238,345,228,35,3],[427,433,396,299,448,307,30,2]],"container":"<table class=\"white-space: nowrap\">\n  <thead>\n    <tr>\n      <th>age_cat<\/th>\n      <th>0<\/th>\n      <th>1<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"pageLength":8,"scrollX":true,"columnDefs":[{"className":"dt-right","targets":[1,2]}],"order":[],"autoWidth":false,"orderClasses":false,"lengthMenu":[8,10,25,50,100]}},"evals":[],"jsHooks":[]}</script><p>Now we can bind the <code>counts_table</code> and the <code>model</code> results together horizontally with <code>bind_cols()</code> (<strong>dplyr</strong>). Remember that with <code>bind_cols()</code> the rows in the two data frames must be aligned perfectly. In this code, because we are binding within a pipe chain, we use <code>.</code> to represent the piped object <code>counts_table</code> as we bind it to <code>model</code>. To finish the process, we use <code>select()</code> to pick the desired columns and their order, and finally apply the <strong>base</strong> R <code><a href="https://rdrr.io/r/base/Round.html">round()</a></code> function across all numeric columns to specify 2 decimal places.</p>
<div class="sourceCode" id="cb978"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">combined</span> <span class="op">&lt;-</span> <span class="va">counts_table</span> <span class="op">%&gt;%</span>           <span class="co"># begin with table of counts</span>
  <span class="fu">bind_cols</span><span class="op">(</span><span class="va">.</span>, <span class="va">model</span><span class="op">)</span> <span class="op">%&gt;%</span>              <span class="co"># combine with the outputs of the regression </span>
  <span class="fu">select</span><span class="op">(</span><span class="va">term</span>, <span class="fl">2</span><span class="op">:</span><span class="fl">3</span>, <span class="va">estimate</span>,          <span class="co"># select and re-order cols</span>
         <span class="va">conf.low</span>, <span class="va">conf.high</span>, <span class="va">p.value</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">mutate</span><span class="op">(</span><span class="fu">across</span><span class="op">(</span><span class="fu">where</span><span class="op">(</span><span class="va">is.numeric</span><span class="op">)</span>, <span class="va">round</span>, digits <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span> <span class="co">## round to 2 decimal places</span></code></pre></div>
<p>Here is what the combined data frame looks like, printed nicely as an image with a function from <strong>flextable</strong>. The <a href="tables-for-presentation.html#tables-for-presentation">Tables for presentation</a> explains how to customize such tables with <strong>flextable</strong>, or or you can use numerous other packages such as <strong>knitr</strong> or <strong>GT</strong>.</p>
<div class="sourceCode" id="cb979"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">combined</span> <span class="op">&lt;-</span> <span class="va">combined</span> <span class="op">%&gt;%</span> 
  <span class="fu">flextable</span><span class="fu">::</span><span class="fu"><a href="https://davidgohel.github.io/flextable/reference/flextable.html">qflextable</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
</div>
<div id="looping-multiple-univariate-models" class="section level4 unnumbered">
<h4>Looping multiple univariate models<a class="anchor" aria-label="anchor" href="#looping-multiple-univariate-models"><i class="fas fa-link"></i></a>
</h4>
<p>Below we present a method using <code><a href="https://rdrr.io/r/stats/glm.html">glm()</a></code> and <code>tidy()</code> for a more simple approach, see the section on <strong>gtsummary</strong>.</p>
<p>To run the models on several exposure variables to produce univariate odds ratios (i.e. not controlling for each other), you can use the approach below. It uses <code>str_c()</code> from <strong>stringr</strong> to create univariate formulas (see <a href="characters-and-strings.html#characters-and-strings">Characters and strings</a>), runs the <code><a href="https://rdrr.io/r/stats/glm.html">glm()</a></code> regression on each formula, passes each <code><a href="https://rdrr.io/r/stats/glm.html">glm()</a></code> output to <code>tidy()</code> and finally collapses all the model outputs together with <code>bind_rows()</code> from <strong>tidyr</strong>. This approach uses <code>map()</code> from the package <strong>purrr</strong> to iterate - see the page on <a href="iteration-loops-and-lists.html#iteration-loops-and-lists">Iteration, loops, and lists</a> for more information on this tool.</p>
<ol style="list-style-type: decimal">
<li><p>Create a vector of column names of the explanatory variables. We already have this as <code>explanatory_vars</code> from the Preparation section of this page.</p></li>
<li><p>Use <code>str_c()</code> to create multiple string formulas, with <code>outcome</code> on the left, and a column name from <code>explanatory_vars</code> on the right. The period <code>.</code> substitutes for the column name in <code>explanatory_vars</code>.</p></li>
</ol>
<div class="sourceCode" id="cb980"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">explanatory_vars</span> <span class="op">%&gt;%</span> <span class="fu">str_c</span><span class="op">(</span><span class="st">"outcome ~ "</span>, <span class="va">.</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "outcome ~ gender"  "outcome ~ fever"   "outcome ~ chills"  "outcome ~ cough"   "outcome ~ aches"   "outcome ~ vomit"   "outcome ~ age_cat"</code></pre>
<ol start="3" style="list-style-type: decimal">
<li><p>Pass these string formulas to <code>map()</code> and set <code>~glm()</code> as the function to apply to each input. Within <code><a href="https://rdrr.io/r/stats/glm.html">glm()</a></code>, set the regression formula as <code><a href="https://rdrr.io/r/stats/formula.html">as.formula(.x)</a></code> where <code>.x</code> will be replaced by the string formula defined in the step above. <code>map()</code> will loop over each of the string formulas, running regressions for each one.</p></li>
<li><p>The outputs of this first <code>map()</code> are passed to a second <code>map()</code> command, which applies <code>tidy()</code> to the regression outputs.</p></li>
<li><p>Finally the output of the second <code>map()</code> (a list of tidied data frames) is condensed with <code>bind_rows()</code>, resulting in one data frame with all the univariate results.</p></li>
</ol>
<div class="sourceCode" id="cb982"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">models</span> <span class="op">&lt;-</span> <span class="va">explanatory_vars</span> <span class="op">%&gt;%</span>       <span class="co"># begin with variables of interest</span>
  <span class="fu">str_c</span><span class="op">(</span><span class="st">"outcome ~ "</span>, <span class="va">.</span><span class="op">)</span> <span class="op">%&gt;%</span>         <span class="co"># combine each variable into formula ("outcome ~ variable of interest")</span>
  
  <span class="co"># iterate through each univariate formula</span>
  <span class="fu">map</span><span class="op">(</span>                               
    .f <span class="op">=</span> <span class="op">~</span><span class="fu"><a href="https://rdrr.io/r/stats/glm.html">glm</a></span><span class="op">(</span>                       <span class="co"># pass the formulas one-by-one to glm()</span>
      formula <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/formula.html">as.formula</a></span><span class="op">(</span><span class="va">.x</span><span class="op">)</span>,      <span class="co"># within glm(), the string formula is .x</span>
      family <span class="op">=</span> <span class="st">"binomial"</span>,           <span class="co"># specify type of glm (logistic)</span>
      data <span class="op">=</span> <span class="va">linelist</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>          <span class="co"># dataset</span>
  
  <span class="co"># tidy up each of the glm regression outputs from above</span>
  <span class="fu">map</span><span class="op">(</span>
    .f <span class="op">=</span> <span class="op">~</span><span class="fu">tidy</span><span class="op">(</span>
      <span class="va">.x</span>, 
      exponentiate <span class="op">=</span> <span class="cn">TRUE</span>,           <span class="co"># exponentiate </span>
      conf.int <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>          <span class="co"># return confidence intervals</span>
  
  <span class="co"># collapse the list of regression outputs in to one data frame</span>
  <span class="fu">bind_rows</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  
  <span class="co"># round all numeric columns</span>
  <span class="fu">mutate</span><span class="op">(</span><span class="fu">across</span><span class="op">(</span><span class="fu">where</span><span class="op">(</span><span class="va">is.numeric</span><span class="op">)</span>, <span class="va">round</span>, digits <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>This time, the end object <code>models</code> is longer because it now represents the combined results of several univariate regressions. Click through to see all the rows of <code>model</code>.</p>
<div id="htmlwidget-00c618f6c94d4ebf812e" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-00c618f6c94d4ebf812e">{"x":{"filter":"none","data":[["(Intercept)","gender","(Intercept)","fever","(Intercept)","chills","(Intercept)","cough","(Intercept)","aches","(Intercept)","vomit","(Intercept)","age_cat5-9","age_cat10-14","age_cat15-19","age_cat20-29","age_cat30-49","age_cat50-69","age_cat70+"],[1.28,1,1.28,1,1.28,1.03,1.14,1.15,1.29,0.93,1.23,1.09,1.26,0.94,1.15,0.99,1.03,1.07,0.68,0.53],[0.04,0.06,0.07,0.08,0.03,0.08,0.08,0.09,0.03,0.1,0.04,0.06,0.07,0.1,0.11,0.11,0.1,0.11,0.26,0.92],[5.67,-0.04,3.44,0.01,6.98,0.41,1.53,1.58,7.78,-0.66,4.67,1.39,3.21,-0.62,1.29,-0.05,0.27,0.56,-1.5,-0.7],[0,0.97,0,0.99,0,0.68,0.13,0.11,0,0.51,0,0.17,0,0.54,0.2,0.96,0.79,0.58,0.13,0.49],[1.18,0.88,1.11,0.85,1.19,0.89,0.97,0.97,1.21,0.76,1.13,0.96,1.1,0.77,0.93,0.8,0.84,0.85,0.41,0.07],[1.4,1.13,1.48,1.17,1.37,1.21,1.34,1.37,1.38,1.14,1.34,1.23,1.46,1.15,1.42,1.24,1.26,1.33,1.13,3.2]],"container":"<table class=\"white-space: nowrap\">\n  <thead>\n    <tr>\n      <th>term<\/th>\n      <th>estimate<\/th>\n      <th>std.error<\/th>\n      <th>statistic<\/th>\n      <th>p.value<\/th>\n      <th>conf.low<\/th>\n      <th>conf.high<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"pageLength":5,"scrollX":true,"columnDefs":[{"className":"dt-right","targets":[1,2,3,4,5,6]}],"order":[],"autoWidth":false,"orderClasses":false,"lengthMenu":[5,10,25,50,100]}},"evals":[],"jsHooks":[]}</script><p>As before, we can create a counts table from the <code>linelist</code> for each explanatory variable, bind it to <code>models</code>, and make a nice table. We begin with the variables, and iterate through them with <code>map()</code>. We iterate through a user-defined function which involves creating a counts table with <strong>dplyr</strong> functions. Then the results are combined and bound with the <code>models</code> model results.</p>
<div class="sourceCode" id="cb983"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## for each explanatory variable</span>
<span class="va">univ_tab_base</span> <span class="op">&lt;-</span> <span class="va">explanatory_vars</span> <span class="op">%&gt;%</span> 
  <span class="fu">map</span><span class="op">(</span>.f <span class="op">=</span> 
    <span class="op">~</span><span class="op">{</span><span class="va">linelist</span> <span class="op">%&gt;%</span>                <span class="co">## begin with linelist</span>
        <span class="fu">group_by</span><span class="op">(</span><span class="va">outcome</span><span class="op">)</span> <span class="op">%&gt;%</span>     <span class="co">## group data set by outcome</span>
        <span class="fu">count</span><span class="op">(</span><span class="va">.data</span><span class="op">[[</span><span class="va">.x</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">%&gt;%</span>    <span class="co">## produce counts for variable of interest</span>
        <span class="fu">pivot_wider</span><span class="op">(</span>              <span class="co">## spread to wide format (as in cross-tabulation)</span>
          names_from <span class="op">=</span> <span class="va">outcome</span>,
          values_from <span class="op">=</span> <span class="va">n</span><span class="op">)</span> <span class="op">%&gt;%</span> 
        <span class="fu">drop_na</span><span class="op">(</span><span class="va">.data</span><span class="op">[[</span><span class="va">.x</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">%&gt;%</span>         <span class="co">## drop rows with missings</span>
        <span class="fu">rename</span><span class="op">(</span><span class="st">"variable"</span> <span class="op">=</span> <span class="va">.x</span><span class="op">)</span> <span class="op">%&gt;%</span>      <span class="co">## change variable of interest column to "variable"</span>
        <span class="fu">mutate</span><span class="op">(</span>variable <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">variable</span><span class="op">)</span><span class="op">)</span><span class="op">}</span> <span class="co">## convert to character, else non-dichotomous (categorical) variables come out as factor and cant be merged</span>
      <span class="op">)</span> <span class="op">%&gt;%</span> 
  
  <span class="co">## collapse the list of count outputs in to one data frame</span>
  <span class="fu">bind_rows</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  
  <span class="co">## merge with the outputs of the regression </span>
  <span class="fu">bind_cols</span><span class="op">(</span><span class="va">.</span>, <span class="va">models</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  
  <span class="co">## only keep columns interested in </span>
  <span class="fu">select</span><span class="op">(</span><span class="va">term</span>, <span class="fl">2</span><span class="op">:</span><span class="fl">3</span>, <span class="va">estimate</span>, <span class="va">conf.low</span>, <span class="va">conf.high</span>, <span class="va">p.value</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  
  <span class="co">## round decimal places</span>
  <span class="fu">mutate</span><span class="op">(</span><span class="fu">across</span><span class="op">(</span><span class="fu">where</span><span class="op">(</span><span class="va">is.numeric</span><span class="op">)</span>, <span class="va">round</span>, digits <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Below is what the data frame looks like. See the page on <a href="tables-for-presentation.html#tables-for-presentation">Tables for presentation</a> for ideas on how to further convert this table to pretty HTML output (e.g. with <strong>flextable</strong>).</p>
<div id="htmlwidget-caa10108358f36cea1b5" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-caa10108358f36cea1b5">{"x":{"filter":"none","data":[["(Intercept)","gender","(Intercept)","fever","(Intercept)","chills","(Intercept)","cough","(Intercept)","aches","(Intercept)","vomit","(Intercept)","age_cat5-9","age_cat10-14","age_cat15-19","age_cat20-29","age_cat30-49","age_cat50-69","age_cat70+"],[909,916,340,1485,1472,353,272,1553,1636,189,931,894,338,365,273,238,345,228,35,3],[1168,1174,436,1906,1877,465,309,2033,2114,228,1144,1198,427,433,396,299,448,307,30,2],[1.28,1,1.28,1,1.28,1.03,1.14,1.15,1.29,0.93,1.23,1.09,1.26,0.94,1.15,0.99,1.03,1.07,0.68,0.53],[1.18,0.88,1.11,0.85,1.19,0.89,0.97,0.97,1.21,0.76,1.13,0.96,1.1,0.77,0.93,0.8,0.84,0.85,0.41,0.07],[1.4,1.13,1.48,1.17,1.37,1.21,1.34,1.37,1.38,1.14,1.34,1.23,1.46,1.15,1.42,1.24,1.26,1.33,1.13,3.2],[0,0.97,0,0.99,0,0.68,0.13,0.11,0,0.51,0,0.17,0,0.54,0.2,0.96,0.79,0.58,0.13,0.49]],"container":"<table class=\"white-space: nowrap\">\n  <thead>\n    <tr>\n      <th>term<\/th>\n      <th>0<\/th>\n      <th>1<\/th>\n      <th>estimate<\/th>\n      <th>conf.low<\/th>\n      <th>conf.high<\/th>\n      <th>p.value<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"pageLength":5,"scrollX":true,"columnDefs":[{"className":"dt-right","targets":[1,2,3,4,5,6]}],"order":[],"autoWidth":false,"orderClasses":false,"lengthMenu":[5,10,25,50,100]}},"evals":[],"jsHooks":[]}</script><!-- ======================================================= -->
</div>
</div>
<div id="reg_gt_uni" class="section level3 unnumbered">
<h3>
<strong>gtsummary</strong> package<a class="anchor" aria-label="anchor" href="#reg_gt_uni"><i class="fas fa-link"></i></a>
</h3>
<p>Below we present the use of <code>tbl_uvregression()</code> from the <strong>gtsummary</strong> package. Just like in the page on <a href="https://epirhandbook.com/descriptive-tables.html">Descriptive tables</a>, <strong>gtsummary</strong> functions do a good job of running statistics <em>and</em> producing professional-looking outputs. This function produces a table of univariate regression results.</p>
<p>We select only the necessary columns from the <code>linelist</code> (explanatory variables and the outcome variable) and pipe them into <code>tbl_uvregression()</code>. We are going to run univariate regression on each of the columns we defined as <code>explanatory_vars</code> in the data Preparation section (gender, fever, chills, cough, aches, vomit, and age_cat).</p>
<p>Within the function itself, we provide the <code>method =</code> as <code>glm</code> (no quotes), the <code>y =</code> outcome column (<code>outcome</code>), specify to <code>method.args =</code> that we want to run logistic regression via <code>family = binomial</code>, and we tell it to exponentiate the results.</p>
<p>The output is HTML and contains the counts</p>
<div class="sourceCode" id="cb984"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">univ_tab</span> <span class="op">&lt;-</span> <span class="va">linelist</span> <span class="op">%&gt;%</span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">explanatory_vars</span>, <span class="va">outcome</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="co">## select variables of interest</span>

  <span class="fu">tbl_uvregression</span><span class="op">(</span>                         <span class="co">## produce univariate table</span>
    method <span class="op">=</span> <span class="va">glm</span>,                           <span class="co">## define regression want to run (generalised linear model)</span>
    y <span class="op">=</span> <span class="va">outcome</span>,                            <span class="co">## define outcome variable</span>
    method.args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>family <span class="op">=</span> <span class="va">binomial</span><span class="op">)</span>,  <span class="co">## define what type of glm want to run (logistic)</span>
    exponentiate <span class="op">=</span> <span class="cn">TRUE</span>                     <span class="co">## exponentiate to produce odds ratios (rather than log odds)</span>
  <span class="op">)</span>

<span class="co">## view univariate results table </span>
<span class="va">univ_tab</span></code></pre></div>
<div id="rxdrrhafki" style="overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
<style>html {
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Helvetica Neue', 'Fira Sans', 'Droid Sans', Arial, sans-serif;
}

#rxdrrhafki .gt_table {
  display: table;
  border-collapse: collapse;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#rxdrrhafki .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#rxdrrhafki .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#rxdrrhafki .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 0;
  padding-bottom: 4px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#rxdrrhafki .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#rxdrrhafki .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#rxdrrhafki .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#rxdrrhafki .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#rxdrrhafki .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#rxdrrhafki .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#rxdrrhafki .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#rxdrrhafki .gt_group_heading {
  padding: 8px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
}

#rxdrrhafki .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#rxdrrhafki .gt_from_md > :first-child {
  margin-top: 0;
}

#rxdrrhafki .gt_from_md > :last-child {
  margin-bottom: 0;
}

#rxdrrhafki .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#rxdrrhafki .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 12px;
}

#rxdrrhafki .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#rxdrrhafki .gt_first_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
}

#rxdrrhafki .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#rxdrrhafki .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#rxdrrhafki .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#rxdrrhafki .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#rxdrrhafki .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#rxdrrhafki .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding: 4px;
}

#rxdrrhafki .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#rxdrrhafki .gt_sourcenote {
  font-size: 90%;
  padding: 4px;
}

#rxdrrhafki .gt_left {
  text-align: left;
}

#rxdrrhafki .gt_center {
  text-align: center;
}

#rxdrrhafki .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#rxdrrhafki .gt_font_normal {
  font-weight: normal;
}

#rxdrrhafki .gt_font_bold {
  font-weight: bold;
}

#rxdrrhafki .gt_font_italic {
  font-style: italic;
}

#rxdrrhafki .gt_super {
  font-size: 65%;
}

#rxdrrhafki .gt_footnote_marks {
  font-style: italic;
  font-weight: normal;
  font-size: 65%;
}
</style>
<div class="inline-table"><table class="gt_table">
<thead class="gt_col_headings"><tr>
<th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1"><strong>Characteristic</strong></th>
      <th class="gt_col_heading gt_columns_bottom_border gt_center" rowspan="1" colspan="1"><strong>N</strong></th>
      <th class="gt_col_heading gt_columns_bottom_border gt_center" rowspan="1" colspan="1">
<strong>OR</strong><sup class="gt_footnote_marks">1</sup>
</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_center" rowspan="1" colspan="1">
<strong>95% CI</strong><sup class="gt_footnote_marks">1</sup>
</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_center" rowspan="1" colspan="1"><strong>p-value</strong></th>
    </tr></thead>
<tbody class="gt_table_body">
<tr>
<td class="gt_row gt_left">gender</td>
<td class="gt_row gt_center">4167</td>
<td class="gt_row gt_center">1.00</td>
<td class="gt_row gt_center">0.88, 1.13</td>
<td class="gt_row gt_center">&gt;0.9</td>
</tr>
<tr>
<td class="gt_row gt_left">fever</td>
<td class="gt_row gt_center">4167</td>
<td class="gt_row gt_center">1.00</td>
<td class="gt_row gt_center">0.85, 1.17</td>
<td class="gt_row gt_center">&gt;0.9</td>
</tr>
<tr>
<td class="gt_row gt_left">chills</td>
<td class="gt_row gt_center">4167</td>
<td class="gt_row gt_center">1.03</td>
<td class="gt_row gt_center">0.89, 1.21</td>
<td class="gt_row gt_center">0.7</td>
</tr>
<tr>
<td class="gt_row gt_left">cough</td>
<td class="gt_row gt_center">4167</td>
<td class="gt_row gt_center">1.15</td>
<td class="gt_row gt_center">0.97, 1.37</td>
<td class="gt_row gt_center">0.11</td>
</tr>
<tr>
<td class="gt_row gt_left">aches</td>
<td class="gt_row gt_center">4167</td>
<td class="gt_row gt_center">0.93</td>
<td class="gt_row gt_center">0.76, 1.14</td>
<td class="gt_row gt_center">0.5</td>
</tr>
<tr>
<td class="gt_row gt_left">vomit</td>
<td class="gt_row gt_center">4167</td>
<td class="gt_row gt_center">1.09</td>
<td class="gt_row gt_center">0.96, 1.23</td>
<td class="gt_row gt_center">0.2</td>
</tr>
<tr>
<td class="gt_row gt_left">age_cat</td>
<td class="gt_row gt_center">4167</td>
<td class="gt_row gt_center"></td>
<td class="gt_row gt_center"></td>
<td class="gt_row gt_center"></td>
</tr>
<tr>
<td class="gt_row gt_left" style="text-align: left; text-indent: 10px;">0-4</td>
<td class="gt_row gt_center"></td>
<td class="gt_row gt_center">—</td>
<td class="gt_row gt_center">—</td>
<td class="gt_row gt_center"></td>
</tr>
<tr>
<td class="gt_row gt_left" style="text-align: left; text-indent: 10px;">5-9</td>
<td class="gt_row gt_center"></td>
<td class="gt_row gt_center">0.94</td>
<td class="gt_row gt_center">0.77, 1.15</td>
<td class="gt_row gt_center">0.5</td>
</tr>
<tr>
<td class="gt_row gt_left" style="text-align: left; text-indent: 10px;">10-14</td>
<td class="gt_row gt_center"></td>
<td class="gt_row gt_center">1.15</td>
<td class="gt_row gt_center">0.93, 1.42</td>
<td class="gt_row gt_center">0.2</td>
</tr>
<tr>
<td class="gt_row gt_left" style="text-align: left; text-indent: 10px;">15-19</td>
<td class="gt_row gt_center"></td>
<td class="gt_row gt_center">0.99</td>
<td class="gt_row gt_center">0.80, 1.24</td>
<td class="gt_row gt_center">&gt;0.9</td>
</tr>
<tr>
<td class="gt_row gt_left" style="text-align: left; text-indent: 10px;">20-29</td>
<td class="gt_row gt_center"></td>
<td class="gt_row gt_center">1.03</td>
<td class="gt_row gt_center">0.84, 1.26</td>
<td class="gt_row gt_center">0.8</td>
</tr>
<tr>
<td class="gt_row gt_left" style="text-align: left; text-indent: 10px;">30-49</td>
<td class="gt_row gt_center"></td>
<td class="gt_row gt_center">1.07</td>
<td class="gt_row gt_center">0.85, 1.33</td>
<td class="gt_row gt_center">0.6</td>
</tr>
<tr>
<td class="gt_row gt_left" style="text-align: left; text-indent: 10px;">50-69</td>
<td class="gt_row gt_center"></td>
<td class="gt_row gt_center">0.68</td>
<td class="gt_row gt_center">0.41, 1.13</td>
<td class="gt_row gt_center">0.13</td>
</tr>
<tr>
<td class="gt_row gt_left" style="text-align: left; text-indent: 10px;">70+</td>
<td class="gt_row gt_center"></td>
<td class="gt_row gt_center">0.53</td>
<td class="gt_row gt_center">0.07, 3.20</td>
<td class="gt_row gt_center">0.5</td>
</tr>
</tbody>
<tfoot><tr class="gt_footnotes">
<td colspan="5">
        <p class="gt_footnote">
          <sup class="gt_footnote_marks"><em>1</em>
          </sup>
           
          OR = Odds Ratio, CI = Confidence Interval
          <br></p>
      </td>
    </tr></tfoot>
</table></div>
</div>
<p>There are many modifications you can make to this table output, such as adjusting the text labels, bolding rows by their p-value, etc. See tutorials <a href="http://www.danieldsjoberg.com/gtsummary/articles/tbl_regression.html">here</a> and elsewhere online.</p>
<!-- ======================================================= -->
</div>
</div>
<div id="stratified" class="section level2" number="19.3">
<h2>
<span class="header-section-number">19.3</span> Stratified<a class="anchor" aria-label="anchor" href="#stratified"><i class="fas fa-link"></i></a>
</h2>
<p>Stratified analysis is currently still being worked on for <strong>gtsummary</strong>,
this page will be updated in due course.</p>
</div>
<div id="multivariable" class="section level2" number="19.4">
<h2>
<span class="header-section-number">19.4</span> Multivariable<a class="anchor" aria-label="anchor" href="#multivariable"><i class="fas fa-link"></i></a>
</h2>
<p>For multivariable analysis, we again present two approaches:</p>
<ul>
<li>
<code><a href="https://rdrr.io/r/stats/glm.html">glm()</a></code> and <code>tidy()</code><br>
</li>
<li>
<strong>gtsummary</strong> package</li>
</ul>
<p>The workflow is similar for each and only the last step of pulling together a final table is different.</p>
<div id="conduct-multivariable" class="section level3 unnumbered">
<h3>Conduct multivariable<a class="anchor" aria-label="anchor" href="#conduct-multivariable"><i class="fas fa-link"></i></a>
</h3>
<p>Here we use <code><a href="https://rdrr.io/r/stats/glm.html">glm()</a></code> but add more variables to the right side of the equation, separated by plus symbols (<code><a href="https://rdrr.io/r/base/Arithmetic.html">+</a></code>).</p>
<p>To run the model with all of our explanatory variables we would run:</p>
<div class="sourceCode" id="cb985"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mv_reg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/glm.html">glm</a></span><span class="op">(</span><span class="va">outcome</span> <span class="op">~</span> <span class="va">gender</span> <span class="op">+</span> <span class="va">fever</span> <span class="op">+</span> <span class="va">chills</span> <span class="op">+</span> <span class="va">cough</span> <span class="op">+</span> <span class="va">aches</span> <span class="op">+</span> <span class="va">vomit</span> <span class="op">+</span> <span class="va">age_cat</span>, family <span class="op">=</span> <span class="st">"binomial"</span>, data <span class="op">=</span> <span class="va">linelist</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">mv_reg</span><span class="op">)</span></code></pre></div>
<pre><code>## 
## Call:
## glm(formula = outcome ~ gender + fever + chills + cough + aches + 
##     vomit + age_cat, family = "binomial", data = linelist)
## 
## Deviance Residuals: 
##    Min      1Q  Median      3Q     Max  
## -1.383  -1.279   1.029   1.078   1.346  
## 
## Coefficients:
##               Estimate Std. Error z value Pr(&gt;|z|)
## (Intercept)   0.069054   0.131726   0.524    0.600
## gender        0.002448   0.065133   0.038    0.970
## fever         0.004309   0.080522   0.054    0.957
## chills        0.034112   0.078924   0.432    0.666
## cough         0.138584   0.089909   1.541    0.123
## aches        -0.070705   0.104078  -0.679    0.497
## vomit         0.086098   0.062618   1.375    0.169
## age_cat5-9   -0.063562   0.101851  -0.624    0.533
## age_cat10-14  0.136372   0.107275   1.271    0.204
## age_cat15-19 -0.011074   0.113640  -0.097    0.922
## age_cat20-29  0.026552   0.102780   0.258    0.796
## age_cat30-49  0.059569   0.116402   0.512    0.609
## age_cat50-69 -0.388964   0.262384  -1.482    0.138
## age_cat70+   -0.647443   0.917375  -0.706    0.480
## 
## (Dispersion parameter for binomial family taken to be 1)
## 
##     Null deviance: 5712.4  on 4166  degrees of freedom
## Residual deviance: 5700.2  on 4153  degrees of freedom
## AIC: 5728.2
## 
## Number of Fisher Scoring iterations: 4</code></pre>
<p>If you want to include two variables and an interaction between them you can separate them with an asterisk <code><a href="https://rdrr.io/r/base/Arithmetic.html">*</a></code> instead of a <code><a href="https://rdrr.io/r/base/Arithmetic.html">+</a></code>. Separate them with a colon <code><a href="https://rdrr.io/r/base/Colon.html">:</a></code> if you are only specifying the interaction. For example:</p>
<div class="sourceCode" id="cb987"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/stats/glm.html">glm</a></span><span class="op">(</span><span class="va">outcome</span> <span class="op">~</span> <span class="va">gender</span> <span class="op">+</span> <span class="va">age_cat</span> <span class="op">*</span> <span class="va">fever</span>, family <span class="op">=</span> <span class="st">"binomial"</span>, data <span class="op">=</span> <span class="va">linelist</span><span class="op">)</span></code></pre></div>
<p><em>Optionally</em>, you can use this code to leverage the pre-defined vector of column names and re-create the above command using <code>str_c()</code>. This might be useful if your explanatory variable names are changing, or you don’t want to type them all out again.</p>
<div class="sourceCode" id="cb988"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## run a regression with all variables of interest </span>
<span class="va">mv_reg</span> <span class="op">&lt;-</span> <span class="va">explanatory_vars</span> <span class="op">%&gt;%</span>  <span class="co">## begin with vector of explanatory column names</span>
  <span class="fu">str_c</span><span class="op">(</span>collapse <span class="op">=</span> <span class="st">"+"</span><span class="op">)</span> <span class="op">%&gt;%</span>     <span class="co">## combine all names of the variables of interest separated by a plus</span>
  <span class="fu">str_c</span><span class="op">(</span><span class="st">"outcome ~ "</span>, <span class="va">.</span><span class="op">)</span> <span class="op">%&gt;%</span>    <span class="co">## combine the names of variables of interest with outcome in formula style</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/glm.html">glm</a></span><span class="op">(</span>family <span class="op">=</span> <span class="st">"binomial"</span>,      <span class="co">## define type of glm as logistic,</span>
      data <span class="op">=</span> <span class="va">linelist</span><span class="op">)</span>          <span class="co">## define your dataset</span></code></pre></div>
<div id="building-the-model" class="section level4 unnumbered">
<h4>Building the model<a class="anchor" aria-label="anchor" href="#building-the-model"><i class="fas fa-link"></i></a>
</h4>
<p>You can build your model step-by-step, saving various models that include certain explanatory variables. You can compare these models with likelihood-ratio tests using <code>lrtest()</code> from the package <strong>lmtest</strong>, as below:</p>
<p><span style="color: black;"><strong><em>NOTE:</em></strong> Using <strong>base</strong> <code>anova(model1, model2, test = "Chisq)</code> produces the same results </span></p>
<div class="sourceCode" id="cb989"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">model1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/glm.html">glm</a></span><span class="op">(</span><span class="va">outcome</span> <span class="op">~</span> <span class="va">age_cat</span>, family <span class="op">=</span> <span class="st">"binomial"</span>, data <span class="op">=</span> <span class="va">linelist</span><span class="op">)</span>
<span class="va">model2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/glm.html">glm</a></span><span class="op">(</span><span class="va">outcome</span> <span class="op">~</span> <span class="va">age_cat</span> <span class="op">+</span> <span class="va">gender</span>, family <span class="op">=</span> <span class="st">"binomial"</span>, data <span class="op">=</span> <span class="va">linelist</span><span class="op">)</span>

<span class="fu">lmtest</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/lmtest/man/lrtest.html">lrtest</a></span><span class="op">(</span><span class="va">model1</span>, <span class="va">model2</span><span class="op">)</span></code></pre></div>
<pre><code>## Likelihood ratio test
## 
## Model 1: outcome ~ age_cat
## Model 2: outcome ~ age_cat + gender
##   #Df  LogLik Df  Chisq Pr(&gt;Chisq)
## 1   8 -2852.6                     
## 2   9 -2852.6  1 0.0002     0.9883</code></pre>
<p>Another option is to take the model object and apply the <code><a href="https://rdrr.io/r/stats/step.html">step()</a></code> function from the <strong>stats</strong> package. Specify which variable selection direction you want use when building the model.</p>
<div class="sourceCode" id="cb991"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## choose a model using forward selection based on AIC</span>
<span class="co">## you can also do "backward" or "both" by adjusting the direction</span>
<span class="va">final_mv_reg</span> <span class="op">&lt;-</span> <span class="va">mv_reg</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/step.html">step</a></span><span class="op">(</span>direction <span class="op">=</span> <span class="st">"forward"</span>, trace <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<p>You can also turn off scientific notation in your R session, for clarity:</p>
<div class="sourceCode" id="cb992"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span><span class="op">(</span>scipen<span class="op">=</span><span class="fl">999</span><span class="op">)</span></code></pre></div>
<p>As described in the section on univariate analysis, pass the model output to <code>tidy()</code> to exponentiate the log odds and CIs. Finally we round all numeric columns to two decimal places. Scroll through to see all the rows.</p>
<div class="sourceCode" id="cb993"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mv_tab_base</span> <span class="op">&lt;-</span> <span class="va">final_mv_reg</span> <span class="op">%&gt;%</span> 
  <span class="fu">broom</span><span class="fu">::</span><span class="fu"><a href="https://generics.r-lib.org/reference/tidy.html">tidy</a></span><span class="op">(</span>exponentiate <span class="op">=</span> <span class="cn">TRUE</span>, conf.int <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">%&gt;%</span>  <span class="co">## get a tidy dataframe of estimates </span>
  <span class="fu">mutate</span><span class="op">(</span><span class="fu">across</span><span class="op">(</span><span class="fu">where</span><span class="op">(</span><span class="va">is.numeric</span><span class="op">)</span>, <span class="va">round</span>, digits <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span>          <span class="co">## round </span></code></pre></div>
<p>Here is what the resulting data frame looks like:</p>
<div id="htmlwidget-0ea43c400fa9b306cbf0" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-0ea43c400fa9b306cbf0">{"x":{"filter":"none","data":[["(Intercept)","gender","fever","chills","cough","aches","vomit","age_cat5-9","age_cat10-14","age_cat15-19","age_cat20-29","age_cat30-49","age_cat50-69","age_cat70+"],[1.07,1,1,1.03,1.15,0.93,1.09,0.94,1.15,0.99,1.03,1.06,0.68,0.52],[0.13,0.07,0.08,0.08,0.09,0.1,0.06,0.1,0.11,0.11,0.1,0.12,0.26,0.92],[0.52,0.04,0.05,0.43,1.54,-0.68,1.37,-0.62,1.27,-0.1,0.26,0.51,-1.48,-0.71],[0.6,0.97,0.96,0.67,0.12,0.5,0.17,0.53,0.2,0.92,0.8,0.61,0.14,0.48],[0.83,0.88,0.86,0.89,0.96,0.76,0.96,0.77,0.93,0.79,0.84,0.85,0.4,0.07],[1.39,1.14,1.18,1.21,1.37,1.14,1.23,1.15,1.41,1.24,1.26,1.33,1.13,3.19]],"container":"<table class=\"white-space: nowrap\">\n  <thead>\n    <tr>\n      <th>term<\/th>\n      <th>estimate<\/th>\n      <th>std.error<\/th>\n      <th>statistic<\/th>\n      <th>p.value<\/th>\n      <th>conf.low<\/th>\n      <th>conf.high<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"pageLength":10,"scrollX":true,"columnDefs":[{"className":"dt-right","targets":[1,2,3,4,5,6]}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><!-- ======================================================= -->
</div>
</div>
<div id="combine-univariate-and-multivariable" class="section level3 unnumbered">
<h3>Combine univariate and multivariable<a class="anchor" aria-label="anchor" href="#combine-univariate-and-multivariable"><i class="fas fa-link"></i></a>
</h3>
<div id="combine-with-gtsummary" class="section level4 unnumbered">
<h4>Combine with <strong>gtsummary</strong><a class="anchor" aria-label="anchor" href="#combine-with-gtsummary"><i class="fas fa-link"></i></a>
</h4>
<p>The <strong>gtsummary</strong> package provides the <code>tbl_regression()</code> function, which will
take the outputs from a regression (<code><a href="https://rdrr.io/r/stats/glm.html">glm()</a></code> in this case) and produce an nice
summary table.</p>
<div class="sourceCode" id="cb994"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## show results table of final regression </span>
<span class="va">mv_tab</span> <span class="op">&lt;-</span> <span class="fu">tbl_regression</span><span class="op">(</span><span class="va">final_mv_reg</span>, exponentiate <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>Let’s see the table:</p>
<div class="sourceCode" id="cb995"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mv_tab</span></code></pre></div>
<div id="qilbhgklhp" style="overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
<style>html {
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Helvetica Neue', 'Fira Sans', 'Droid Sans', Arial, sans-serif;
}

#qilbhgklhp .gt_table {
  display: table;
  border-collapse: collapse;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#qilbhgklhp .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#qilbhgklhp .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#qilbhgklhp .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 0;
  padding-bottom: 4px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#qilbhgklhp .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#qilbhgklhp .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#qilbhgklhp .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#qilbhgklhp .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#qilbhgklhp .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#qilbhgklhp .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#qilbhgklhp .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#qilbhgklhp .gt_group_heading {
  padding: 8px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
}

#qilbhgklhp .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#qilbhgklhp .gt_from_md > :first-child {
  margin-top: 0;
}

#qilbhgklhp .gt_from_md > :last-child {
  margin-bottom: 0;
}

#qilbhgklhp .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#qilbhgklhp .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 12px;
}

#qilbhgklhp .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#qilbhgklhp .gt_first_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
}

#qilbhgklhp .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#qilbhgklhp .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#qilbhgklhp .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#qilbhgklhp .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#qilbhgklhp .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#qilbhgklhp .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding: 4px;
}

#qilbhgklhp .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#qilbhgklhp .gt_sourcenote {
  font-size: 90%;
  padding: 4px;
}

#qilbhgklhp .gt_left {
  text-align: left;
}

#qilbhgklhp .gt_center {
  text-align: center;
}

#qilbhgklhp .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#qilbhgklhp .gt_font_normal {
  font-weight: normal;
}

#qilbhgklhp .gt_font_bold {
  font-weight: bold;
}

#qilbhgklhp .gt_font_italic {
  font-style: italic;
}

#qilbhgklhp .gt_super {
  font-size: 65%;
}

#qilbhgklhp .gt_footnote_marks {
  font-style: italic;
  font-weight: normal;
  font-size: 65%;
}
</style>
<div class="inline-table"><table class="gt_table">
<thead class="gt_col_headings"><tr>
<th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1"><strong>Characteristic</strong></th>
      <th class="gt_col_heading gt_columns_bottom_border gt_center" rowspan="1" colspan="1">
<strong>OR</strong><sup class="gt_footnote_marks">1</sup>
</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_center" rowspan="1" colspan="1">
<strong>95% CI</strong><sup class="gt_footnote_marks">1</sup>
</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_center" rowspan="1" colspan="1"><strong>p-value</strong></th>
    </tr></thead>
<tbody class="gt_table_body">
<tr>
<td class="gt_row gt_left">gender</td>
<td class="gt_row gt_center">1.00</td>
<td class="gt_row gt_center">0.88, 1.14</td>
<td class="gt_row gt_center">&gt;0.9</td>
</tr>
<tr>
<td class="gt_row gt_left">fever</td>
<td class="gt_row gt_center">1.00</td>
<td class="gt_row gt_center">0.86, 1.18</td>
<td class="gt_row gt_center">&gt;0.9</td>
</tr>
<tr>
<td class="gt_row gt_left">chills</td>
<td class="gt_row gt_center">1.03</td>
<td class="gt_row gt_center">0.89, 1.21</td>
<td class="gt_row gt_center">0.7</td>
</tr>
<tr>
<td class="gt_row gt_left">cough</td>
<td class="gt_row gt_center">1.15</td>
<td class="gt_row gt_center">0.96, 1.37</td>
<td class="gt_row gt_center">0.12</td>
</tr>
<tr>
<td class="gt_row gt_left">aches</td>
<td class="gt_row gt_center">0.93</td>
<td class="gt_row gt_center">0.76, 1.14</td>
<td class="gt_row gt_center">0.5</td>
</tr>
<tr>
<td class="gt_row gt_left">vomit</td>
<td class="gt_row gt_center">1.09</td>
<td class="gt_row gt_center">0.96, 1.23</td>
<td class="gt_row gt_center">0.2</td>
</tr>
<tr>
<td class="gt_row gt_left">age_cat</td>
<td class="gt_row gt_center"></td>
<td class="gt_row gt_center"></td>
<td class="gt_row gt_center"></td>
</tr>
<tr>
<td class="gt_row gt_left" style="text-align: left; text-indent: 10px;">0-4</td>
<td class="gt_row gt_center">—</td>
<td class="gt_row gt_center">—</td>
<td class="gt_row gt_center"></td>
</tr>
<tr>
<td class="gt_row gt_left" style="text-align: left; text-indent: 10px;">5-9</td>
<td class="gt_row gt_center">0.94</td>
<td class="gt_row gt_center">0.77, 1.15</td>
<td class="gt_row gt_center">0.5</td>
</tr>
<tr>
<td class="gt_row gt_left" style="text-align: left; text-indent: 10px;">10-14</td>
<td class="gt_row gt_center">1.15</td>
<td class="gt_row gt_center">0.93, 1.41</td>
<td class="gt_row gt_center">0.2</td>
</tr>
<tr>
<td class="gt_row gt_left" style="text-align: left; text-indent: 10px;">15-19</td>
<td class="gt_row gt_center">0.99</td>
<td class="gt_row gt_center">0.79, 1.24</td>
<td class="gt_row gt_center">&gt;0.9</td>
</tr>
<tr>
<td class="gt_row gt_left" style="text-align: left; text-indent: 10px;">20-29</td>
<td class="gt_row gt_center">1.03</td>
<td class="gt_row gt_center">0.84, 1.26</td>
<td class="gt_row gt_center">0.8</td>
</tr>
<tr>
<td class="gt_row gt_left" style="text-align: left; text-indent: 10px;">30-49</td>
<td class="gt_row gt_center">1.06</td>
<td class="gt_row gt_center">0.85, 1.33</td>
<td class="gt_row gt_center">0.6</td>
</tr>
<tr>
<td class="gt_row gt_left" style="text-align: left; text-indent: 10px;">50-69</td>
<td class="gt_row gt_center">0.68</td>
<td class="gt_row gt_center">0.40, 1.13</td>
<td class="gt_row gt_center">0.14</td>
</tr>
<tr>
<td class="gt_row gt_left" style="text-align: left; text-indent: 10px;">70+</td>
<td class="gt_row gt_center">0.52</td>
<td class="gt_row gt_center">0.07, 3.19</td>
<td class="gt_row gt_center">0.5</td>
</tr>
</tbody>
<tfoot><tr class="gt_footnotes">
<td colspan="4">
        <p class="gt_footnote">
          <sup class="gt_footnote_marks"><em>1</em>
          </sup>
           
          OR = Odds Ratio, CI = Confidence Interval
          <br></p>
      </td>
    </tr></tfoot>
</table></div>
</div>
<p>You can also combine several different output tables produced by <strong>gtsummary</strong> with
the <code>tbl_merge()</code> function. We now combine the multivariable results with the <strong>gtsummary</strong> <em>univariate</em> results that we created <a href="univariate-and-multivariable-regression.html#reg_gt_uni">above</a>:</p>
<div class="sourceCode" id="cb996"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## combine with univariate results </span>
<span class="fu">tbl_merge</span><span class="op">(</span>
  tbls <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">univ_tab</span>, <span class="va">mv_tab</span><span class="op">)</span>,                          <span class="co"># combine</span>
  tab_spanner <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"**Univariate**"</span>, <span class="st">"**Multivariable**"</span><span class="op">)</span><span class="op">)</span> <span class="co"># set header names</span></code></pre></div>
<div id="efafyomvns" style="overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
<style>html {
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Helvetica Neue', 'Fira Sans', 'Droid Sans', Arial, sans-serif;
}

#efafyomvns .gt_table {
  display: table;
  border-collapse: collapse;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#efafyomvns .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#efafyomvns .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#efafyomvns .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 0;
  padding-bottom: 4px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#efafyomvns .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#efafyomvns .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#efafyomvns .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#efafyomvns .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#efafyomvns .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#efafyomvns .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#efafyomvns .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#efafyomvns .gt_group_heading {
  padding: 8px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
}

#efafyomvns .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#efafyomvns .gt_from_md > :first-child {
  margin-top: 0;
}

#efafyomvns .gt_from_md > :last-child {
  margin-bottom: 0;
}

#efafyomvns .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#efafyomvns .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 12px;
}

#efafyomvns .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#efafyomvns .gt_first_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
}

#efafyomvns .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#efafyomvns .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#efafyomvns .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#efafyomvns .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#efafyomvns .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#efafyomvns .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding: 4px;
}

#efafyomvns .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#efafyomvns .gt_sourcenote {
  font-size: 90%;
  padding: 4px;
}

#efafyomvns .gt_left {
  text-align: left;
}

#efafyomvns .gt_center {
  text-align: center;
}

#efafyomvns .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#efafyomvns .gt_font_normal {
  font-weight: normal;
}

#efafyomvns .gt_font_bold {
  font-weight: bold;
}

#efafyomvns .gt_font_italic {
  font-style: italic;
}

#efafyomvns .gt_super {
  font-size: 65%;
}

#efafyomvns .gt_footnote_marks {
  font-style: italic;
  font-weight: normal;
  font-size: 65%;
}
</style>
<div class="inline-table"><table class="gt_table">
<thead class="gt_col_headings">
<tr>
<th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="2" colspan="1"><strong>Characteristic</strong></th>
      <th class="gt_center gt_columns_top_border gt_column_spanner_outer" rowspan="1" colspan="4">
        <span class="gt_column_spanner"><strong>Univariate</strong></span>
      </th>
      <th class="gt_center gt_columns_top_border gt_column_spanner_outer" rowspan="1" colspan="3">
        <span class="gt_column_spanner"><strong>Multivariable</strong></span>
      </th>
    </tr>
<tr>
<th class="gt_col_heading gt_columns_bottom_border gt_center" rowspan="1" colspan="1"><strong>N</strong></th>
      <th class="gt_col_heading gt_columns_bottom_border gt_center" rowspan="1" colspan="1">
<strong>OR</strong><sup class="gt_footnote_marks">1</sup>
</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_center" rowspan="1" colspan="1">
<strong>95% CI</strong><sup class="gt_footnote_marks">1</sup>
</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_center" rowspan="1" colspan="1"><strong>p-value</strong></th>
      <th class="gt_col_heading gt_columns_bottom_border gt_center" rowspan="1" colspan="1">
<strong>OR</strong><sup class="gt_footnote_marks">1</sup>
</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_center" rowspan="1" colspan="1">
<strong>95% CI</strong><sup class="gt_footnote_marks">1</sup>
</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_center" rowspan="1" colspan="1"><strong>p-value</strong></th>
    </tr>
</thead>
<tbody class="gt_table_body">
<tr>
<td class="gt_row gt_left">gender</td>
<td class="gt_row gt_center">4167</td>
<td class="gt_row gt_center">1.00</td>
<td class="gt_row gt_center">0.88, 1.13</td>
<td class="gt_row gt_center">&gt;0.9</td>
<td class="gt_row gt_center">1.00</td>
<td class="gt_row gt_center">0.88, 1.14</td>
<td class="gt_row gt_center">&gt;0.9</td>
</tr>
<tr>
<td class="gt_row gt_left">fever</td>
<td class="gt_row gt_center">4167</td>
<td class="gt_row gt_center">1.00</td>
<td class="gt_row gt_center">0.85, 1.17</td>
<td class="gt_row gt_center">&gt;0.9</td>
<td class="gt_row gt_center">1.00</td>
<td class="gt_row gt_center">0.86, 1.18</td>
<td class="gt_row gt_center">&gt;0.9</td>
</tr>
<tr>
<td class="gt_row gt_left">chills</td>
<td class="gt_row gt_center">4167</td>
<td class="gt_row gt_center">1.03</td>
<td class="gt_row gt_center">0.89, 1.21</td>
<td class="gt_row gt_center">0.7</td>
<td class="gt_row gt_center">1.03</td>
<td class="gt_row gt_center">0.89, 1.21</td>
<td class="gt_row gt_center">0.7</td>
</tr>
<tr>
<td class="gt_row gt_left">cough</td>
<td class="gt_row gt_center">4167</td>
<td class="gt_row gt_center">1.15</td>
<td class="gt_row gt_center">0.97, 1.37</td>
<td class="gt_row gt_center">0.11</td>
<td class="gt_row gt_center">1.15</td>
<td class="gt_row gt_center">0.96, 1.37</td>
<td class="gt_row gt_center">0.12</td>
</tr>
<tr>
<td class="gt_row gt_left">aches</td>
<td class="gt_row gt_center">4167</td>
<td class="gt_row gt_center">0.93</td>
<td class="gt_row gt_center">0.76, 1.14</td>
<td class="gt_row gt_center">0.5</td>
<td class="gt_row gt_center">0.93</td>
<td class="gt_row gt_center">0.76, 1.14</td>
<td class="gt_row gt_center">0.5</td>
</tr>
<tr>
<td class="gt_row gt_left">vomit</td>
<td class="gt_row gt_center">4167</td>
<td class="gt_row gt_center">1.09</td>
<td class="gt_row gt_center">0.96, 1.23</td>
<td class="gt_row gt_center">0.2</td>
<td class="gt_row gt_center">1.09</td>
<td class="gt_row gt_center">0.96, 1.23</td>
<td class="gt_row gt_center">0.2</td>
</tr>
<tr>
<td class="gt_row gt_left">age_cat</td>
<td class="gt_row gt_center">4167</td>
<td class="gt_row gt_center"></td>
<td class="gt_row gt_center"></td>
<td class="gt_row gt_center"></td>
<td class="gt_row gt_center"></td>
<td class="gt_row gt_center"></td>
<td class="gt_row gt_center"></td>
</tr>
<tr>
<td class="gt_row gt_left" style="text-align: left; text-indent: 10px;">0-4</td>
<td class="gt_row gt_center"></td>
<td class="gt_row gt_center">—</td>
<td class="gt_row gt_center">—</td>
<td class="gt_row gt_center"></td>
<td class="gt_row gt_center">—</td>
<td class="gt_row gt_center">—</td>
<td class="gt_row gt_center"></td>
</tr>
<tr>
<td class="gt_row gt_left" style="text-align: left; text-indent: 10px;">5-9</td>
<td class="gt_row gt_center"></td>
<td class="gt_row gt_center">0.94</td>
<td class="gt_row gt_center">0.77, 1.15</td>
<td class="gt_row gt_center">0.5</td>
<td class="gt_row gt_center">0.94</td>
<td class="gt_row gt_center">0.77, 1.15</td>
<td class="gt_row gt_center">0.5</td>
</tr>
<tr>
<td class="gt_row gt_left" style="text-align: left; text-indent: 10px;">10-14</td>
<td class="gt_row gt_center"></td>
<td class="gt_row gt_center">1.15</td>
<td class="gt_row gt_center">0.93, 1.42</td>
<td class="gt_row gt_center">0.2</td>
<td class="gt_row gt_center">1.15</td>
<td class="gt_row gt_center">0.93, 1.41</td>
<td class="gt_row gt_center">0.2</td>
</tr>
<tr>
<td class="gt_row gt_left" style="text-align: left; text-indent: 10px;">15-19</td>
<td class="gt_row gt_center"></td>
<td class="gt_row gt_center">0.99</td>
<td class="gt_row gt_center">0.80, 1.24</td>
<td class="gt_row gt_center">&gt;0.9</td>
<td class="gt_row gt_center">0.99</td>
<td class="gt_row gt_center">0.79, 1.24</td>
<td class="gt_row gt_center">&gt;0.9</td>
</tr>
<tr>
<td class="gt_row gt_left" style="text-align: left; text-indent: 10px;">20-29</td>
<td class="gt_row gt_center"></td>
<td class="gt_row gt_center">1.03</td>
<td class="gt_row gt_center">0.84, 1.26</td>
<td class="gt_row gt_center">0.8</td>
<td class="gt_row gt_center">1.03</td>
<td class="gt_row gt_center">0.84, 1.26</td>
<td class="gt_row gt_center">0.8</td>
</tr>
<tr>
<td class="gt_row gt_left" style="text-align: left; text-indent: 10px;">30-49</td>
<td class="gt_row gt_center"></td>
<td class="gt_row gt_center">1.07</td>
<td class="gt_row gt_center">0.85, 1.33</td>
<td class="gt_row gt_center">0.6</td>
<td class="gt_row gt_center">1.06</td>
<td class="gt_row gt_center">0.85, 1.33</td>
<td class="gt_row gt_center">0.6</td>
</tr>
<tr>
<td class="gt_row gt_left" style="text-align: left; text-indent: 10px;">50-69</td>
<td class="gt_row gt_center"></td>
<td class="gt_row gt_center">0.68</td>
<td class="gt_row gt_center">0.41, 1.13</td>
<td class="gt_row gt_center">0.13</td>
<td class="gt_row gt_center">0.68</td>
<td class="gt_row gt_center">0.40, 1.13</td>
<td class="gt_row gt_center">0.14</td>
</tr>
<tr>
<td class="gt_row gt_left" style="text-align: left; text-indent: 10px;">70+</td>
<td class="gt_row gt_center"></td>
<td class="gt_row gt_center">0.53</td>
<td class="gt_row gt_center">0.07, 3.20</td>
<td class="gt_row gt_center">0.5</td>
<td class="gt_row gt_center">0.52</td>
<td class="gt_row gt_center">0.07, 3.19</td>
<td class="gt_row gt_center">0.5</td>
</tr>
</tbody>
<tfoot><tr class="gt_footnotes">
<td colspan="8">
        <p class="gt_footnote">
          <sup class="gt_footnote_marks"><em>1</em>
          </sup>
           
          OR = Odds Ratio, CI = Confidence Interval
          <br></p>
      </td>
    </tr></tfoot>
</table></div>
</div>
</div>
<div id="combine-with-dplyr" class="section level4 unnumbered">
<h4>Combine with <strong>dplyr</strong><a class="anchor" aria-label="anchor" href="#combine-with-dplyr"><i class="fas fa-link"></i></a>
</h4>
<p>An alternative way of combining the <code><a href="https://rdrr.io/r/stats/glm.html">glm()</a></code>/<code>tidy()</code> univariate and multivariable outputs is with the <strong>dplyr</strong> join functions.</p>
<ul>
<li>Join the univariate results from earlier (<code>univ_tab_base</code>, which contains counts) with the tidied multivariable results <code>mv_tab_base</code><br>
</li>
<li>Use <code>select()</code> to keep only the columns we want, specify their order, and re-name them<br>
</li>
<li>Use <code><a href="https://rdrr.io/r/base/Round.html">round()</a></code> with two decimal places on all the column that are class Double</li>
</ul>
<div class="sourceCode" id="cb997"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## combine univariate and multivariable tables </span>
<span class="fu">left_join</span><span class="op">(</span><span class="va">univ_tab_base</span>, <span class="va">mv_tab_base</span>, by <span class="op">=</span> <span class="st">"term"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="co">## choose columns and rename them</span>
  <span class="fu">select</span><span class="op">(</span> <span class="co"># new name =  old name</span>
    <span class="st">"characteristic"</span> <span class="op">=</span> <span class="va">term</span>, 
    <span class="st">"recovered"</span>      <span class="op">=</span> <span class="st">"0"</span>, 
    <span class="st">"dead"</span>           <span class="op">=</span> <span class="st">"1"</span>, 
    <span class="st">"univ_or"</span>        <span class="op">=</span> <span class="va">estimate.x</span>, 
    <span class="st">"univ_ci_low"</span>    <span class="op">=</span> <span class="va">conf.low.x</span>, 
    <span class="st">"univ_ci_high"</span>   <span class="op">=</span> <span class="va">conf.high.x</span>,
    <span class="st">"univ_pval"</span>      <span class="op">=</span> <span class="va">p.value.x</span>, 
    <span class="st">"mv_or"</span>          <span class="op">=</span> <span class="va">estimate.y</span>, 
    <span class="st">"mvv_ci_low"</span>     <span class="op">=</span> <span class="va">conf.low.y</span>, 
    <span class="st">"mv_ci_high"</span>     <span class="op">=</span> <span class="va">conf.high.y</span>,
    <span class="st">"mv_pval"</span>        <span class="op">=</span> <span class="va">p.value.y</span> 
  <span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">mutate</span><span class="op">(</span><span class="fu">across</span><span class="op">(</span><span class="fu">where</span><span class="op">(</span><span class="va">is.double</span><span class="op">)</span>, <span class="va">round</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span>   </code></pre></div>
<pre><code>## # A tibble: 20 x 11
##    characteristic recovered  dead univ_or univ_ci_low univ_ci_high univ_pval mv_or mvv_ci_low mv_ci_high mv_pval
##    &lt;chr&gt;              &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;     &lt;dbl&gt; &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;   &lt;dbl&gt;
##  1 (Intercept)          909  1168    1.28        1.18         1.4       0     1.07       0.83       1.39    0.6 
##  2 gender               916  1174    1           0.88         1.13      0.97  1          0.88       1.14    0.97
##  3 (Intercept)          340   436    1.28        1.11         1.48      0     1.07       0.83       1.39    0.6 
##  4 fever               1485  1906    1           0.85         1.17      0.99  1          0.86       1.18    0.96
##  5 (Intercept)         1472  1877    1.28        1.19         1.37      0     1.07       0.83       1.39    0.6 
##  6 chills               353   465    1.03        0.89         1.21      0.68  1.03       0.89       1.21    0.67
##  7 (Intercept)          272   309    1.14        0.97         1.34      0.13  1.07       0.83       1.39    0.6 
##  8 cough               1553  2033    1.15        0.97         1.37      0.11  1.15       0.96       1.37    0.12
##  9 (Intercept)         1636  2114    1.29        1.21         1.38      0     1.07       0.83       1.39    0.6 
## 10 aches                189   228    0.93        0.76         1.14      0.51  0.93       0.76       1.14    0.5 
## 11 (Intercept)          931  1144    1.23        1.13         1.34      0     1.07       0.83       1.39    0.6 
## 12 vomit                894  1198    1.09        0.96         1.23      0.17  1.09       0.96       1.23    0.17
## 13 (Intercept)          338   427    1.26        1.1          1.46      0     1.07       0.83       1.39    0.6 
## 14 age_cat5-9           365   433    0.94        0.77         1.15      0.54  0.94       0.77       1.15    0.53
## 15 age_cat10-14         273   396    1.15        0.93         1.42      0.2   1.15       0.93       1.41    0.2 
## 16 age_cat15-19         238   299    0.99        0.8          1.24      0.96  0.99       0.79       1.24    0.92
## 17 age_cat20-29         345   448    1.03        0.84         1.26      0.79  1.03       0.84       1.26    0.8 
## 18 age_cat30-49         228   307    1.07        0.85         1.33      0.58  1.06       0.85       1.33    0.61
## 19 age_cat50-69          35    30    0.68        0.41         1.13      0.13  0.68       0.4        1.13    0.14
## 20 age_cat70+             3     2    0.53        0.07         3.2       0.49  0.52       0.07       3.19    0.48</code></pre>
<!-- ======================================================= -->
</div>
</div>
</div>
<div id="forest-plot" class="section level2" number="19.5">
<h2>
<span class="header-section-number">19.5</span> Forest plot<a class="anchor" aria-label="anchor" href="#forest-plot"><i class="fas fa-link"></i></a>
</h2>
<p>This section shows how to produce a plot with the outputs of your regression.
There are two options, you can build a plot yourself using <strong>ggplot2</strong> or use a
meta-package called <strong>easystats</strong> (a package that includes many packages).</p>
<p>See the page on <a href="ggplot-basics.html#ggplot-basics">ggplot basics</a> if you are unfamiliar with the <strong>ggplot2</strong> plotting package.</p>
<!-- ======================================================= -->
<div id="ggplot2-package" class="section level3 unnumbered">
<h3>
<strong>ggplot2</strong> package<a class="anchor" aria-label="anchor" href="#ggplot2-package"><i class="fas fa-link"></i></a>
</h3>
<p>You can build a forest plot with <code>ggplot()</code> by plotting elements of the multivariable regression results. Add the layers of the plots using these “geoms”:</p>
<ul>
<li>estimates with <code>geom_point()</code><br>
</li>
<li>confidence intervals with <code>geom_errorbar()</code><br>
</li>
<li>a vertical line at OR = 1 with <code>geom_vline()</code>
</li>
</ul>
<p>Before plotting, you may want to use <code>fct_relevel()</code> from the <strong>forcats</strong> package to set the order of the variables/levels on the y-axis. <code>ggplot()</code> may display them in alpha-numeric order which would not work well for these age category values (“30” would appear before “5”). See the page on <a href="factors.html#factors">Factors</a> for more details.</p>
<div class="sourceCode" id="cb999"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## remove the intercept term from your multivariable results</span>
<span class="va">mv_tab_base</span> <span class="op">%&gt;%</span> 
  
  <span class="co">#set order of levels to appear along y-axis</span>
  <span class="fu">mutate</span><span class="op">(</span>term <span class="op">=</span> <span class="fu">fct_relevel</span><span class="op">(</span>
    <span class="va">term</span>,
    <span class="st">"vomit"</span>, <span class="st">"gender"</span>, <span class="st">"fever"</span>, <span class="st">"cough"</span>, <span class="st">"chills"</span>, <span class="st">"aches"</span>,
    <span class="st">"age_cat5-9"</span>, <span class="st">"age_cat10-14"</span>, <span class="st">"age_cat15-19"</span>, <span class="st">"age_cat20-29"</span>,
    <span class="st">"age_cat30-49"</span>, <span class="st">"age_cat50-69"</span>, <span class="st">"age_cat70+"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="co"># remove "intercept" row from plot</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">term</span> <span class="op">!=</span> <span class="st">"(Intercept)"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  
  <span class="co">## plot with variable on the y axis and estimate (OR) on the x axis</span>
  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">estimate</span>, y <span class="op">=</span> <span class="va">term</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  
  <span class="co">## show the estimate as a point</span>
  <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  
  <span class="co">## add in an error bar for the confidence intervals</span>
  <span class="fu">geom_errorbar</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>xmin <span class="op">=</span> <span class="va">conf.low</span>, xmax <span class="op">=</span> <span class="va">conf.high</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  
  <span class="co">## show where OR = 1 is for reference as a dashed line</span>
  <span class="fu">geom_vline</span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">1</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/ggplot_forest-1.png" width="672"></div>
<!-- ======================================================= -->
</div>
<div id="easystats-packages" class="section level3 unnumbered">
<h3>
<strong>easystats</strong> packages<a class="anchor" aria-label="anchor" href="#easystats-packages"><i class="fas fa-link"></i></a>
</h3>
<p>An alternative, if you do not want to the fine level of control that <strong>ggplot2</strong> provides, is to use a combination of <strong>easystats</strong> packages.</p>
<p>The function <code>model_parameters()</code> from the <strong>parameters</strong> package does the equivalent
of the <strong>broom</strong> package function <code>tidy()</code>. The <strong>see</strong> package then accepts those outputs
and creates a default forest plot as a <code>ggplot()</code> object.</p>
<div class="sourceCode" id="cb1000"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">pacman</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pacman/man/p_load.html">p_load</a></span><span class="op">(</span><span class="va">easystats</span><span class="op">)</span></code></pre></div>
<pre><code>## Installing package into 'C:/Users/neale/OneDrive - Neale Batra/Documents/Analytics-LAPTOP-RS5P2IBO/R/Projects/R handbook/epiRhandbook_eng/renv/library/R-4.1/x86_64-w64-mingw32'
## (as 'lib' is unspecified)</code></pre>
<pre><code>## Warning: package 'easystats' is not available for this version of R
## 
## A version of this package for your version of R might be available elsewhere,
## see the ideas at
## https://cran.r-project.org/doc/manuals/r-patched/R-admin.html#Installing-packages</code></pre>
<pre><code>## Warning: unable to access index for repository http://www.stats.ox.ac.uk/pub/RWin/bin/windows/contrib/4.1:
##   cannot open URL 'http://www.stats.ox.ac.uk/pub/RWin/bin/windows/contrib/4.1/PACKAGES'</code></pre>
<pre><code>## Warning in p_install(package, character.only = TRUE, ...):</code></pre>
<pre><code>## Warning in library(package, lib.loc = lib.loc, character.only = TRUE, logical.return = TRUE, : there is no package called 'easystats'</code></pre>
<pre><code>## Warning in pacman::p_load(easystats): Failed to install/load:
## easystats</code></pre>
<div class="sourceCode" id="cb1007"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## remove the intercept term from your multivariable results</span>
<span class="va">final_mv_reg</span> <span class="op">%&gt;%</span> 
  <span class="fu">model_parameters</span><span class="op">(</span>exponentiate <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/easystats_forest-1.png" width="672"></div>
<!-- ======================================================= -->
</div>
</div>
<div id="resources-12" class="section level2" number="19.6">
<h2>
<span class="header-section-number">19.6</span> Resources<a class="anchor" aria-label="anchor" href="#resources-12"><i class="fas fa-link"></i></a>
</h2>
<p>The content of this page was informed by these resources and vignettes online:</p>
<p><a href="https://www.datacamp.com/community/tutorials/linear-regression-R">Linear regression in R</a></p>
<p><a href="http://www.danieldsjoberg.com/gtsummary/articles/tbl_regression.html">gtsummary</a></p>
<p><a href="https://stats.idre.ucla.edu/other/dae/">UCLA stats page</a></p>
<p><a href="http://www.sthda.com/english/articles/36-classification-methods-essentials/150-stepwise-logistic-regression-essentials-in-r/">sthda stepwise regression</a></p>

</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="simple-statistical-tests.html"><span class="header-section-number">18</span> Simple statistical tests</a></div>
<div class="next"><a href="missing-data.html"><span class="header-section-number">20</span> Missing data</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#univariate-and-multivariable-regression"><span class="header-section-number">19</span> Univariate and multivariable regression</a></li>
<li>
<a class="nav-link" href="#preparation-10"><span class="header-section-number">19.1</span> Preparation</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#load-packages-12">Load packages</a></li>
<li><a class="nav-link" href="#import-data-12">Import data</a></li>
<li><a class="nav-link" href="#clean-data">Clean data</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#univariate"><span class="header-section-number">19.2</span> Univariate</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#base-r-3">base R</a></li>
<li><a class="nav-link" href="#reg_gt_uni">gtsummary package</a></li>
</ul>
</li>
<li><a class="nav-link" href="#stratified"><span class="header-section-number">19.3</span> Stratified</a></li>
<li>
<a class="nav-link" href="#multivariable"><span class="header-section-number">19.4</span> Multivariable</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#conduct-multivariable">Conduct multivariable</a></li>
<li><a class="nav-link" href="#combine-univariate-and-multivariable">Combine univariate and multivariable</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#forest-plot"><span class="header-section-number">19.5</span> Forest plot</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#ggplot2-package">ggplot2 package</a></li>
<li><a class="nav-link" href="#easystats-packages">easystats packages</a></li>
</ul>
</li>
<li><a class="nav-link" href="#resources-12"><span class="header-section-number">19.6</span> Resources</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
          
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>The Epidemiologist R Handbook</strong>" was written by the handbook team. It was last built on 2021-11-17.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>
</html>
